{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This a jupyter notebook guide on domain analysis\n",
    "\n",
    "by Pu Zheng and Bogdan Bintu\n",
    "\n",
    "2020.06.06\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import sys, os, glob, time, copy\n",
    "import numpy as np\n",
    "import scipy\n",
    "import pickle\n",
    "\n",
    "sys.path.append(os.path.abspath(r\"..\\.\"))\n",
    "\n",
    "import source as ia\n",
    "\n",
    "from scipy.signal import find_peaks\n",
    "from scipy.spatial.distance import cdist,pdist,squareform\n",
    "\n",
    "print(os.getpid())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required plotting setting\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rc('font', family='serif')\n",
    "plt.rc('font', serif='Arial')\n",
    "_font_size = 7.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required plotting parameters\n",
    "from source.figure_tools import _dpi,_single_col_width,_double_col_width,_single_row_height,_ref_bar_length, _ticklabel_size,_ticklabel_width,_font_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure folder\n",
    "parent_figure_folder = r'\\\\10.245.74.158\\Chromatin_NAS_4\\Chromatin_Share\\final_figures'\n",
    "figure_folder = os.path.join(parent_figure_folder, 'Chr21_domain_figures')\n",
    "print(figure_folder)\n",
    "if not os.path.exists(figure_folder):\n",
    "    os.makedirs(figure_folder)\n",
    "    print(\"generating this folder\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = r'E:\\Users\\puzheng\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\DataForReviewers'\n",
    "data_folder = r'C:\\Users\\Bogdan\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\DataForReviewers'\n",
    "rep1_filename = os.path.join(data_folder, 'chromosome21.tsv')\n",
    "rep2_filename = os.path.join(data_folder, 'chromosome21-cell_cycle.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.1 load chr21 (replicate 1 - without cell cycle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load from file and extract info\n",
    "import csv\n",
    "rep1_info_dict = {}\n",
    "with open(rep1_filename, 'r') as _handle:\n",
    "    _reader = csv.reader(_handle, delimiter='\\t', quotechar='|')\n",
    "    _headers = next(_reader)\n",
    "    print(_headers)\n",
    "    # create keys for each header\n",
    "    for _h in _headers:\n",
    "        rep1_info_dict[_h] = []\n",
    "    # loop through content\n",
    "    for _contents in _reader:\n",
    "        for _h, _info in zip(_headers,_contents):\n",
    "            rep1_info_dict[_h].append(_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "# clean up info\n",
    "data_rep1 = {'params':{}}\n",
    "\n",
    "# clean up genomic coordiantes\n",
    "region_names = np.unique(rep1_info_dict['Genomic coordinate'])\n",
    "region_starts = np.array([int(_n.split(':')[1].split('-')[0]) for _n in region_names])\n",
    "region_ends = np.array([int(_n.split(':')[1].split('-')[1]) for _n in region_names])[np.argsort(region_starts)]\n",
    "region_starts = np.sort(region_starts)\n",
    "\n",
    "mid_positions = ((region_starts + region_ends)/2).astype(np.int)\n",
    "mid_positions_Mb = np.round(mid_positions / 1e6, 2) \n",
    "\n",
    "# clean up chrom copy number\n",
    "chr_nums = np.array([int(_info) for _info in rep1_info_dict['Chromosome copy number']])\n",
    "chr_ids, region_cts = np.unique(chr_nums, return_counts=True)\n",
    "dna_zxys_list = [[[] for _start in region_starts] for _id in chr_ids]\n",
    "\n",
    "# clean up zxy\n",
    "for _z,_x,_y,_reg_info, _cid in tqdm(zip(rep1_info_dict['Z(nm)'],rep1_info_dict['X(nm)'],\\\n",
    "                                         rep1_info_dict['Y(nm)'],rep1_info_dict['Genomic coordinate'],\\\n",
    "                                         rep1_info_dict['Chromosome copy number'])):\n",
    "    # get chromosome inds\n",
    "    _cid = int(_cid)\n",
    "    _cind = np.where(chr_ids == _cid)[0][0]\n",
    "    \n",
    "    # get region indices\n",
    "    _start = int(_reg_info.split(':')[1].split('-')[0])\n",
    "    _rind = np.where(region_starts==_start)[0][0]\n",
    "    \n",
    "    dna_zxys_list[_cind][_rind] = np.array([float(_z),float(_x), float(_y)])\n",
    "\n",
    "# merge together\n",
    "dna_zxys_list = np.array(dna_zxys_list)\n",
    "data_rep1['chrom_ids'] = chr_ids\n",
    "data_rep1['mid_position_Mb'] = mid_positions_Mb\n",
    "data_rep1['dna_zxys'] = dna_zxys_list\n",
    "\n",
    "# clean up tss and transcription\n",
    "if 'Gene names' in rep1_info_dict:\n",
    "    import re\n",
    "    # first extract number of genes\n",
    "    gene_names = []\n",
    "    for _gene_info, _trans_info, _tss_coord in zip(rep1_info_dict['Gene names'],\n",
    "                                                   rep1_info_dict['Transcription'],\n",
    "                                                   rep1_info_dict['TSS ZXY(nm)']):\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            for _gene in _genes:\n",
    "                if _gene not in gene_names:\n",
    "                    gene_names.append(_gene)\n",
    "    print(f\"{len(gene_names)} genes exist in this dataset.\")\n",
    "    \n",
    "    # initialize gene and transcription\n",
    "    tss_zxys_list = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    transcription_profiles = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    # loop through to get info\n",
    "    for _cid, _gene_info, _trans_info, _tss_locations in tqdm(zip(rep1_info_dict['Chromosome copy number'],\n",
    "                                                                  rep1_info_dict['Gene names'],\n",
    "                                                                  rep1_info_dict['Transcription'],\n",
    "                                                                  rep1_info_dict['TSS ZXY(nm)'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        # process if there are genes in this region:\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            _transcribes = _trans_info.split(';')[:-1]\n",
    "            _tss_zxys = _tss_locations.split(';')[:-1]\n",
    "            for _gene, _transcribe, _tss_zxy in zip(_genes, _transcribes, _tss_zxys):\n",
    "                # get gene index\n",
    "                _gind = gene_names.index(_gene)\n",
    "                # get transcription profile\n",
    "                if _transcribe == 'on':\n",
    "                    transcription_profiles[_cind][_gind] = True\n",
    "                else:\n",
    "                    transcription_profiles[_cind][_gind] = False\n",
    "                # get coordinates\n",
    "                _tss_zxy = np.array([np.float(_c) for _c in re.split(r'\\s+', _tss_zxy.split('[')[1].split(']')[0]) if _c != ''])\n",
    "                tss_zxys_list[_cind][_gind] = _tss_zxy\n",
    "                \n",
    "    tss_zxys_list = np.array(tss_zxys_list)\n",
    "    transcription_profiles = np.array(transcription_profiles)\n",
    "    data_rep1['gene_names'] = gene_names\n",
    "    data_rep1['tss_zxys'] = tss_zxys_list\n",
    "    data_rep1['trans_pfs'] = transcription_profiles\n",
    "\n",
    "# clean up cell_cycle states\n",
    "if 'Cell cycle state' in rep1_info_dict:\n",
    "    cell_cycle_types = np.unique(rep1_info_dict['Cell cycle state'])\n",
    "    cell_cycle_flag_dict = {_k:[[] for _id in chr_ids] for _k in cell_cycle_types if _k != 'ND'}\n",
    "    for _cid, _state in tqdm(zip(rep1_info_dict['Chromosome copy number'],rep1_info_dict['Cell cycle state'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        if np.array([_v[_cind]==[] for _k,_v in cell_cycle_flag_dict.items()]).any():\n",
    "            for _k,_v in cell_cycle_flag_dict.items():\n",
    "                if _k == _state:\n",
    "                    _v[_cind] = True\n",
    "                else:\n",
    "                    _v[_cind] = False\n",
    "    # append to data\n",
    "    for _k, _v in cell_cycle_flag_dict.items():\n",
    "        data_rep1[f'{_k}_flags'] = np.array(_v)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_rep1.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. population averaged description of chr21"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 (FigS1F) population average maps:\n",
    "\n",
    "median distance map, proximity frequency map, corresponding Hi-C map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zxys_rep1_list = np.array(data_rep1['dna_zxys'])\n",
    "distmap_rep1_list = np.array([squareform(pdist(_zxy)) for _zxy in tqdm(zxys_rep1_list)])\n",
    "\n",
    "# generate median distance map\n",
    "median_distance_map_rep1 = np.nanmedian(distmap_rep1_list, axis = 0)\n",
    "# generate contact map\n",
    "contact_th = 500\n",
    "contact_map_rep1 = np.nanmean(distmap_rep1_list < contact_th, axis=0)\n",
    "# load Hi-C\n",
    "hic_file = r'C:\\Users\\Bogdan\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\DataForReviewers\\population-averaged\\hi-c_contacts_chromosome21.tsv'\n",
    "hic_txt = np.array([ln[:-1].split('\\t')for ln in open(hic_file,'r') if len(ln)>1])\n",
    "hic_raw_map = np.array(hic_txt[1:,1:],dtype=np.float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import LogNorm\n",
    "\n",
    "median_limits = [0, 2000]\n",
    "median_cmap = matplotlib.cm.get_cmap('seismic_r')\n",
    "median_cmap.set_bad(color=[0.,0.,0.,1])\n",
    "\n",
    "contact_limits = [0.05, 0.6]\n",
    "contact_norm = LogNorm(vmin=np.min(contact_limits), \n",
    "                       vmax=np.max(contact_limits))\n",
    "contact_cmap = matplotlib.cm.get_cmap('seismic')\n",
    "contact_cmap.set_bad(color=[0.,0.,0.,1])\n",
    "\n",
    "hic_limits = [1, 400]\n",
    "hic_norm = LogNorm(vmin=np.min(hic_limits), \n",
    "                   vmax=np.max(hic_limits))\n",
    "hic_cmap = matplotlib.cm.get_cmap('seismic')\n",
    "hic_cmap.set_bad(color=[0.,0.,0.,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from source.figure_tools.distmap import plot_distance_map\n",
    "print(figure_folder)\n",
    "%matplotlib inline\n",
    "distmap_ax = plot_distance_map(median_distance_map_rep1, \n",
    "                               cmap=median_cmap,\n",
    "                               color_limits=median_limits,\n",
    "                               tick_labels=mid_positions_Mb, \n",
    "                               ax_label='Genomic positions (Mb)', \n",
    "                               colorbar_labels='Distances (nm)',\n",
    "                               save=True, save_folder=figure_folder, \n",
    "                               save_basename=f'FigS1F1_median_distmap_rep1.pdf',\n",
    "                               font_size=5)\n",
    "contact_ax = plot_distance_map(contact_map_rep1, \n",
    "                               cmap=contact_cmap,\n",
    "                               color_limits=contact_limits,\n",
    "                               color_norm=contact_norm,\n",
    "                               tick_labels=mid_positions_Mb, \n",
    "                               ax_label='Genomic positions (Mb)', \n",
    "                               colorbar_labels='Proximity frequency',\n",
    "                               save=True, save_folder=figure_folder, \n",
    "                               save_basename=f'FigS1F2_contact_map_rep1.pdf',\n",
    "                               font_size=5)\n",
    "hic_ax = plot_distance_map(hic_raw_map, \n",
    "                           cmap=hic_cmap,\n",
    "                           color_limits=hic_limits,\n",
    "                           color_norm=hic_norm,\n",
    "                           tick_labels=mid_positions_Mb, \n",
    "                           ax_label='Genomic positions (Mb)', \n",
    "                           colorbar_labels='Hi-C count',\n",
    "                           save=True, save_folder=figure_folder, \n",
    "                           save_basename=f'FigS1F3_hic_map.pdf',\n",
    "                           font_size=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 (S1G) correlation between median-distance vs. Hi-C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_spot_flags = np.isnan(np.array(zxys_rep1_list)).sum(2)==0\n",
    "\n",
    "failure_rates = 1 - np.mean(good_spot_flags, axis=0)\n",
    "\n",
    "good_regions_rep1 = np.where(failure_rates < 0.25)[0]\n",
    "print(len(good_regions_rep1))\n",
    "\n",
    "kept_median_rep1 = median_distance_map_rep1[good_regions_rep1][:,good_regions_rep1]\n",
    "kept_hic_rep1 = hic_raw_map[good_regions_rep1][:,good_regions_rep1]\n",
    "\n",
    "wt_median_entries_rep1 = kept_median_rep1[np.triu_indices(len(kept_median_rep1),1)]\n",
    "hic_contact_entries_rep1 = kept_hic_rep1[np.triu_indices(len(kept_hic_rep1),1)]\n",
    "\n",
    "kept = (wt_median_entries_rep1>0) * (hic_contact_entries_rep1>0)\n",
    "\n",
    "from scipy.stats import linregress, pearsonr\n",
    "lr_rep1 = linregress(np.log(wt_median_entries_rep1[kept]),\n",
    "                np.log(hic_contact_entries_rep1[kept]))\n",
    "print(lr_rep1)\n",
    "print('pearson correlation:', np.abs(lr_rep1.rvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xticks = np.round(2**np.linspace(-2,1,4)*1000,0).astype(np.int)\n",
    "yticks = np.logspace(0, 4, 3).astype(np.int)\n",
    "xlim = [200,2200]\n",
    "# draw scatter plot\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=_dpi)\n",
    "ax.plot(wt_median_entries_rep1[kept], hic_contact_entries_rep1[kept], '.', color='gray', alpha=0.3, markersize=1, )\n",
    "\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "ax.set_xticks(xticks, minor=False)\n",
    "ax.set_xticklabels(xticks)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "               width=_ticklabel_width, length=_ticklabel_size,\n",
    "               pad=1)\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.set_xlabel('Median distances (nm)', labelpad=2, fontsize=_font_size+1)\n",
    "ax.set_ylabel('Hi-C counts', labelpad=2, fontsize=_font_size+1)\n",
    "\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_ylim([0.9,10000])\n",
    "\n",
    "#ax.set_yticks(yticks, minor=True)\n",
    "#ax.set_yticklabels(yticks)\n",
    "\n",
    "\n",
    "reg_x = np.linspace(250, 2000, 100)\n",
    "reg_y = np.exp( lr_rep1.slope * np.log(reg_x) + lr_rep1.intercept)\n",
    "\n",
    "ax.plot(reg_x, reg_y, 'r', label=f'slope = {lr_rep1.slope:.2f}\\n\\u03C1 = {lr_rep1.rvalue:.2f}')\n",
    "\n",
    "plt.legend(loc='upper right', fontsize=_font_size-1)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, 'FigS1G_scatter_median_hic_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zoomed in correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "limits = [325, 390]\n",
    "crop = slice(limits[0],limits[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_crop_regions_rep1 = np.array([_r for _r in good_regions_rep1 if _r in np.arange(limits[0], limits[1])], dtype=np.int)\n",
    "\n",
    "kept_crop_median_rep1 = median_distance_map_rep1[good_crop_regions_rep1][:,good_crop_regions_rep1]\n",
    "kept_crop_hic_rep1 = hic_raw_map[good_crop_regions_rep1][:,good_crop_regions_rep1]\n",
    "\n",
    "wt_crop_median_entries_rep1 = kept_crop_median_rep1[np.triu_indices(len(kept_crop_median_rep1),1)]\n",
    "hic_crop_contact_entries_rep1 = kept_crop_hic_rep1[np.triu_indices(len(kept_crop_hic_rep1),1)]\n",
    "\n",
    "kept_crop_rep1 = (wt_crop_median_entries_rep1>0) * (hic_crop_contact_entries_rep1>0)\n",
    "\n",
    "from scipy.stats import linregress, pearsonr\n",
    "lr_crop_rep1 = linregress(np.log(wt_crop_median_entries_rep1[kept_crop_rep1]),\n",
    "                np.log(hic_crop_contact_entries_rep1[kept_crop_rep1]))\n",
    "print(lr_crop_rep1)\n",
    "print('pearson correlation:', np.abs(lr_crop_rep1.rvalue))\n",
    "\n",
    "# Plot\n",
    "xticks = np.round(2**np.linspace(-2,1,4)*1000,0).astype(np.int)\n",
    "yticks = np.logspace(0, 4, 3).astype(np.int)\n",
    "xlim = [160,1700]\n",
    "# draw scatter plot\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=_dpi)\n",
    "ax.plot(wt_crop_median_entries_rep1[kept_crop_rep1], \n",
    "        hic_crop_contact_entries_rep1[kept_crop_rep1], '.', color='gray', alpha=0.3, markersize=1, )\n",
    "ax.set_xscale('log')\n",
    "ax.set_yscale('log')\n",
    "ax.set_xticks(xticks, minor=False)\n",
    "ax.set_xticklabels(xticks)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "               width=_ticklabel_width, length=_ticklabel_size,\n",
    "               pad=1)\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "ax.set_xlabel('Median distances (nm)', labelpad=2, fontsize=_font_size+1)\n",
    "ax.set_ylabel('Hi-C counts', labelpad=2, fontsize=_font_size+1)\n",
    "\n",
    "reg_x = np.linspace(250, 2000, 100)\n",
    "reg_y = np.exp( lr_crop_rep1.slope * np.log(reg_x) + lr_crop_rep1.intercept)\n",
    "\n",
    "ax.set_xlim(xlim)\n",
    "ax.set_ylim([0.9,10000])\n",
    "\n",
    "ax.plot(reg_x, reg_y, 'r', label=f'slope = {lr_crop_rep1.slope:.2f}\\n\\u03C1 = {lr_crop_rep1.rvalue:.2f}')\n",
    "\n",
    "plt.legend(loc='upper right', fontsize=_font_size-1)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'FigS1I_scatter_median_hic_{limits}.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Determine best correlation bewteen contact and Hi-C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# generate contact maps\n",
    "contact_map_dict_rep1 = {}\n",
    "thr_list = [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000]\n",
    "\n",
    "for _thr in thr_list:\n",
    "    print(_thr)\n",
    "    # new way to calculate contact\n",
    "    contact_map_dict_rep1[_thr] = np.sum(distmap_rep1_list<_thr, axis=0) / np.sum(np.isnan(distmap_rep1_list)==False, axis=0)\n",
    "# calculate pearson correlation with Hi-C\n",
    "pearson_corr_list_rep1 = []\n",
    "for _thr in thr_list:\n",
    "    _contact_map = contact_map_dict_rep1[_thr]\n",
    "    good_spot_flags = np.isnan(np.array(zxys_rep1_list)).sum(2)==0\n",
    "\n",
    "    failure_rates = 1 - np.mean(good_spot_flags, axis=0)\n",
    "\n",
    "    good_regions = np.where(failure_rates < 0.25)[0]\n",
    "    #print(len(good_regions))\n",
    "\n",
    "    kept_contacts = _contact_map[good_regions][:,good_regions]\n",
    "    kept_hic = hic_raw_map[good_regions][:,good_regions]\n",
    "\n",
    "    wt_contact_entries = kept_contacts[np.triu_indices(len(kept_contacts),1)]\n",
    "    hic_contact_entries = kept_hic[np.triu_indices(len(kept_hic),1)]\n",
    "\n",
    "    kept = (wt_contact_entries>0) * (hic_contact_entries>0)\n",
    "\n",
    "    from scipy.stats import linregress, pearsonr\n",
    "    lr = linregress(np.log(wt_contact_entries[kept]),\n",
    "                    np.log(hic_contact_entries[kept]))\n",
    "    print(_thr, 'nm; pearson correlation:', np.abs(lr.rvalue))\n",
    "    \n",
    "    pearson_corr_list_rep1.append(lr.rvalue)\n",
    "    \n",
    "\n",
    "%matplotlib inline\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=600)\n",
    "ax.plot(thr_list, pearson_corr_list_rep1, linewidth=1, alpha=0.7, marker ='.')\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.set_ylim([0.725,0.9])\n",
    "ax.set_xlim([0,1050])\n",
    "ax.set_xticks(np.arange(0,1001,200))\n",
    "ax.set_yticks(np.arange(0.7,0.91,0.05))\n",
    "\n",
    "ax.set_xlabel(\"Cutoff threshold (nm)\", fontsize=_font_size, labelpad=1)\n",
    "ax.set_ylabel(\"Pearson correlation with Hi-C\", fontsize=_font_size, labelpad=1)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.16)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'FigS1J_chr21_proximity_hic_pearson_with_thresholds_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Analysis for single-cell domains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1.1 find single-cell domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import source.domain_tools.DomainAnalysis as da\n",
    "import multiprocessing as mp\n",
    "\n",
    "num_threads=32\n",
    "domain_corr_cutoff = 0.75 \n",
    "domain_dist_cutoff = 500 # nm\n",
    "\n",
    "_domain_args = [(_zxys, 4, 1000, domain_corr_cutoff, domain_dist_cutoff) \n",
    "                     for _zxys in data_rep1['dna_zxys']]\n",
    "_domain_time = time.time()\n",
    "\n",
    "print(f\"Multiprocessing call domain starts\", end=' ')\n",
    "if 'domain_starts' not in data_rep1:\n",
    "    with mp.Pool(num_threads) as domain_pool:\n",
    "        domain_results = domain_pool.starmap(da.get_dom_starts_cor, _domain_args)\n",
    "        domain_pool.close()\n",
    "        domain_pool.join()\n",
    "        domain_pool.terminate()\n",
    "    # save\n",
    "    data_rep1['domain_starts'] = [np.array(_r[-1]) for _r in domain_results]\n",
    "    data_rep1['params']['domain_corr_cutoff'] = domain_corr_cutoff\n",
    "    data_rep1['params']['domain_dist_cutoff'] = domain_dist_cutoff\n",
    "    \n",
    "print(f\"in {time.time()-_domain_time:.3f}s.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Add noise of 100nm and re-find single-cell domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "data_noise = deepcopy(data_rep1)\n",
    "del(data_noise['domain_starts'])\n",
    "data_noise['dna_zxys']+=np.random.normal(scale=100/1.6,size=data_noise['dna_zxys'].shape)\n",
    "dist_dif = np.linalg.norm(data_rep1['dna_zxys']-data_noise['dna_zxys'],axis=-1)\n",
    "print(\"Displacement error:\",np.nanmean(dist_dif))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import source.domain_tools.DomainAnalysis as da\n",
    "import multiprocessing as mp\n",
    "\n",
    "num_threads=32\n",
    "domain_corr_cutoff = 0.75 \n",
    "domain_dist_cutoff = 500 # nm\n",
    "\n",
    "_domain_args = [(_zxys, 4, 1000, domain_corr_cutoff, domain_dist_cutoff) \n",
    "                     for _zxys in data_noise['dna_zxys']]\n",
    "_domain_time = time.time()\n",
    "\n",
    "print(f\"Multiprocessing call domain starts\", end=' ')\n",
    "if 'domain_starts' not in data_noise:\n",
    "    with mp.Pool(num_threads) as domain_pool:\n",
    "        domain_results = domain_pool.starmap(da.get_dom_starts_cor, _domain_args)\n",
    "        domain_pool.close()\n",
    "        domain_pool.join()\n",
    "        domain_pool.terminate()\n",
    "    # save\n",
    "    data_noise['domain_starts'] = [np.array(_r[-1]) for _r in domain_results]\n",
    "    data_noise['params']['domain_corr_cutoff'] = domain_corr_cutoff\n",
    "    data_noise['params']['domain_dist_cutoff'] = domain_dist_cutoff\n",
    "    \n",
    "print(f\"in {time.time()-_domain_time:.3f}s.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Genomic size and radius of gyration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genomic sizes\n",
    "region_size = 0.05 # Mb\n",
    "rep1_sz_list = []\n",
    "\n",
    "for _zxys, _dm_starts in zip(data_rep1['dna_zxys'],data_rep1['domain_starts']):\n",
    "    _starts = _dm_starts[:-1]\n",
    "    _ends = _dm_starts[1:]\n",
    "    # sizes\n",
    "    _sizes = (_dm_starts[1:] - _dm_starts[:-1]) * region_size\n",
    "    # append \n",
    "    rep1_sz_list.append(_sizes)\n",
    "\n",
    "noise_sz_list = []\n",
    "for _zxys, _dm_starts in zip(data_noise['dna_zxys'],data_noise['domain_starts']):\n",
    "    _starts = _dm_starts[:-1]\n",
    "    _ends = _dm_starts[1:]\n",
    "    # sizes\n",
    "    _sizes = (_dm_starts[1:] - _dm_starts[:-1]) * region_size\n",
    "    # append \n",
    "    noise_sz_list.append(_sizes)\n",
    "    \n",
    "    \n",
    "%matplotlib inline\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=600)\n",
    "ax.hist(np.concatenate(noise_sz_list), 100, range=(0,5), \n",
    "        density=True, color='k', alpha=1, label=f'median={np.nanmedian(np.concatenate(rep1_sz_list)):.2f}Mb')\n",
    "ax.hist(np.concatenate(rep1_sz_list), 100, range=(0,5), \n",
    "        density=True, color='g', alpha=0.5, label=f'median={np.nanmedian(np.concatenate(rep1_sz_list)):.2f}Mb')\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "#ax.legend(fontsize=_font_size-1)\n",
    "\n",
    "ax.set_xlabel(\"Genomic size (Mb)\", labelpad=1, fontsize=_font_size)\n",
    "ax.set_ylabel(\"Probability density\", labelpad=1, fontsize=_font_size)\n",
    "ax.set_title(\"Chr21 domain genomic size\", pad=2, fontsize=_font_size)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.16)\n",
    "save_file = os.path.join(figure_folder, f'Fig1I_chr21_domain_gsize_hist_rep1.pdf')\n",
    "plt.savefig(save_file, transparent=True)\n",
    "print(save_file)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rg_mean(zxy):\n",
    "    \"\"\"computes radius of gyration\"\"\"\n",
    "    zxy_ = np.array(zxy)\n",
    "    zxy_ = zxy_[~np.isnan(zxy_[:,0])]\n",
    "    zxy_ = zxy_ - np.mean(zxy_,0)\n",
    "    return np.sqrt(np.mean(np.sum(zxy_**2,axis=-1)))\n",
    "# radius of gyrations\n",
    "rep1_rg_list = []\n",
    "for _zxys, _dm_starts in zip(data_rep1['dna_zxys'],data_rep1['domain_starts']):\n",
    "    _starts = _dm_starts[:-1]\n",
    "    _ends = _dm_starts[1:]\n",
    "    # rgs\n",
    "    _rgs = np.array([rg_mean(_zxys[_s:_e]) for _s, _e in zip(_starts, _ends)])\n",
    "    # append \n",
    "    rep1_rg_list.append(_rgs)\n",
    "# radius of gyration for noise\n",
    "noise_rg_list = []\n",
    "for _zxys, _dm_starts in zip(data_noise['dna_zxys'],data_noise['domain_starts']):\n",
    "    _starts = _dm_starts[:-1]\n",
    "    _ends = _dm_starts[1:]\n",
    "    # rgs\n",
    "    _rgs = np.array([rg_mean(_zxys[_s:_e]) for _s, _e in zip(_starts, _ends)])\n",
    "    # append \n",
    "    noise_rg_list.append(_rgs)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=600)\n",
    "ax.hist(np.concatenate(noise_rg_list), 100, range=(0,1500), \n",
    "        density=True, color='k', alpha=1, label=f'median={np.nanmedian(np.concatenate(rep1_rg_list)):.0f}nm')\n",
    "ax.hist(np.concatenate(rep1_rg_list), 100, range=(0,1500), \n",
    "        density=True, color='g', alpha=0.6, label=f'median={np.nanmedian(np.concatenate(rep1_rg_list)):.0f}nm')\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size-1, \n",
    "                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "#ax.legend(fontsize=_font_size-1)\n",
    "\n",
    "ax.set_xlabel(\"Radius of gyration (nm)\", labelpad=1, fontsize=_font_size)\n",
    "ax.set_ylabel(\"Probability density\", labelpad=1, fontsize=_font_size)\n",
    "ax.set_title(\"Chr21 domain radius of gyration\", pad=2, fontsize=_font_size)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.16)\n",
    "save_file = os.path.join(figure_folder, f'Fig1J_chr21_domain_RG_hist_rep1.pdf')\n",
    "plt.savefig(save_file, transparent=True)\n",
    "print(save_file)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Single-cell boundary probability, insulation and alignment with CTCF/TADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_starts_fs = data_rep1['domain_starts']\n",
    "zxys = data_rep1['dna_zxys']\n",
    "pts= zxys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### calculate boundary probability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_all = np.array([dom for doms in dom_starts_fs[::] for dom in doms[1:-1]])\n",
    "unk_,cts_=np.unique(dom_all,return_counts=True)\n",
    "cts = np.zeros(len(pts[0]))\n",
    "cts[unk_]=cts_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### boundary probability for a zoom-in example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pylab as plt\n",
    "import numpy as np\n",
    "import pickle,os\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from scipy.spatial.distance import pdist,cdist,squareform\n",
    "\n",
    "\n",
    "####### You will need cv2. If you do not have it, run: pip install opencv-python\n",
    "import cv2\n",
    "\n",
    "\n",
    "from matplotlib import cm\n",
    "\n",
    "\n",
    "def resize(im__,scale_percent = 100):\n",
    "    width = int(im__.shape[1] * scale_percent / 100)\n",
    "    height = int(im__.shape[0] * scale_percent / 100)\n",
    "    dim = (width, height)\n",
    "    resized = cv2.resize(im__, dim, interpolation = cv2.INTER_NEAREST)\n",
    "    return resized\n",
    "\n",
    "def rotate_bound(image, angle):\n",
    "    # grab the dimensions of the image and then determine the\n",
    "    # center\n",
    "    (h, w) = image.shape[:2]\n",
    "    (cX, cY) = (w // 2, h // 2)\n",
    " \n",
    "    # grab the rotation matrix (applying the negative of the\n",
    "    # angle to rotate clockwise), then grab the sine and cosine\n",
    "    # (i.e., the rotation components of the matrix)\n",
    "    M = cv2.getRotationMatrix2D((cX, cY), -angle, 1.0)\n",
    "    cos = np.abs(M[0, 0])\n",
    "    sin = np.abs(M[0, 1])\n",
    " \n",
    "    # compute the new bounding dimensions of the image\n",
    "    nW = int((h * sin) + (w * cos))\n",
    "    nH = int((h * cos) + (w * sin))\n",
    " \n",
    "    # adjust the rotation matrix to take into account translation\n",
    "    M[0, 2] += (nW / 2) - cX\n",
    "    M[1, 2] += (nH / 2) - cY\n",
    " \n",
    "    # perform the actual rotation and return the image\n",
    "    return cv2.warpAffine(image, M, (nW, nH),cv2.INTER_NEAREST)\n",
    "\n",
    "\n",
    "def interp1dnan(A):\n",
    "    A_=np.array(A)\n",
    "    ok = np.isnan(A)==False\n",
    "    xp = ok.nonzero()[0]\n",
    "    fp = A[ok]\n",
    "    x  = np.isnan(A).nonzero()[0]\n",
    "    A_[np.isnan(A)] = np.interp(x, xp, fp)\n",
    "    return A_\n",
    "\n",
    "def interpolate_chr(_chr):\n",
    "    \"\"\"linear interpolate chromosome coordinates\"\"\"\n",
    "    _new_chr = np.array(_chr)\n",
    "    for i in range(_new_chr.shape[-1]):\n",
    "        _new_chr[:,i]=interp1dnan(_new_chr[:,i])\n",
    "    return _new_chr\n",
    "from mpl_toolkits.axes_grid1 import ImageGrid\n",
    "from matplotlib import cm\n",
    "fig = plt.figure(figsize=(20,20))\n",
    "grid = ImageGrid(fig, 111, nrows_ncols=(4, 1),axes_pad=0.)\n",
    "\n",
    "\n",
    "mat_ = np.log(contact_map_rep1)\n",
    "pad=0\n",
    "min_val,max_val = -2,None   # the minimum and maximum distance in nanometers. this sets the threshold of the image\n",
    "if max_val is None: max_val = np.nanmax(mat_)\n",
    "if min_val is None: min_val = np.nanmin(mat_)\n",
    "#This colors the image\n",
    "im_ = (np.clip(mat_,min_val,max_val)-min_val)/(max_val-min_val)\n",
    "im__ = np.array(cm.seismic(im_)[:,:,:3]*255,dtype=np.uint8)\n",
    "\n",
    "# resize image 10x to get good resolution\n",
    "resc = 10############\n",
    "resized = resize(im__,resc*100)\n",
    "\n",
    "# Rotate 45 degs\n",
    "resized = rotate_bound(resized,-45)\n",
    "start = int(pad*np.sqrt(2)*resc)\n",
    "center = int(resized.shape[1]/2)\n",
    "\n",
    "#Clip it to the desired size\n",
    "padup=30##### how much of the matrix to keep in the up direction\n",
    "resized = resized[center-resc*padup:center+resc*padup]\n",
    "\n",
    "\n",
    "\n",
    "#List of positions of CTCF and rad21 in chr21\n",
    "#ctcf \n",
    "ctcf = [  9,  21,  33,  67,  73,  78, 139, 226, 231, 235, 242, 253, 256,\n",
    "       273, 284, 292, 301, 307, 339, 350, 355, 363, 366, 370, 373, 376,\n",
    "       381, 385, 390, 396, 402, 405, 410, 436, 440, 446, 456, 469, 472,\n",
    "       476, 482, 485, 488, 492, 500, 505, 508, 512, 520, 540, 543, 550,\n",
    "       554, 560, 565, 576, 580, 585, 589, 592, 595, 599, 602, 606, 615,\n",
    "       619, 622, 625, 628, 633, 636, 639]\n",
    "# rad21\n",
    "rad21=[ 21,  33,  67,  73, 139, 226, 231, 236, 242, 256, 273, 284, 292,\n",
    "       301, 305, 339, 350, 355, 363, 366, 370, 381, 386, 390, 396, 405,\n",
    "       410, 415, 436, 440, 446, 456, 469, 472, 482, 485, 492, 500, 505,\n",
    "       508, 512, 543, 550, 554, 560, 576, 581, 585, 589, 593, 596, 599,\n",
    "       602, 615, 619, 622, 625, 628, 633, 636]\n",
    "\n",
    "start = 0\n",
    "min__ = 0\n",
    "cts_perc = 1.*cts/len(pts)*100*resc\n",
    "x_vals = (np.arange(len(cts_perc))-min__)*resc*np.sqrt(2)-start\n",
    "\n",
    "\n",
    "#grid[1].imshow(A_.T,cmap='bwr')\n",
    "grid[1].plot(x_vals,cts_perc,'ko-')\n",
    "grid[0].imshow(resized)\n",
    "grid[2].plot(x_vals[ctcf],[0]*len(ctcf),'^',color='orange',mec='k')\n",
    "grid[3].plot(x_vals[rad21],[0]*len(rad21),'^',color='yellow',mec='k')\n",
    "ypad=20\n",
    "grid[2].set_ylim([-ypad,ypad])\n",
    "grid[3].set_ylim([-ypad,ypad])\n",
    "grid[2].set_yticks([])\n",
    "grid[3].set_yticks([])\n",
    "#grid[1].set_yticks([])\n",
    "#grid[1].set_ylabel('AB  ',rotation='horizontal')\n",
    "grid[2].set_ylabel('CTCF        ',rotation='horizontal')\n",
    "grid[3].set_ylabel('RAD21        ',rotation='horizontal')\n",
    "min_,max_ = (282, 480)\n",
    "grid[0].set_xlim([min_*resc*np.sqrt(2),max_*resc*np.sqrt(2)])\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, \n",
    "                         f'Fig1C_chr21_sc-domain_prob_rep1.pdf'), transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate TADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#median_distance_map_rep1\n",
    "#contact_map_dict_rep1\n",
    "\n",
    "zxys_noise_list = np.array(data_noise['dna_zxys'])\n",
    "distmap_noise_list = np.array([squareform(pdist(_zxy)) for _zxy in tqdm(zxys_noise_list)])\n",
    "\n",
    "# generate median distance map\n",
    "median_distance_map_noise = np.nanmedian(distmap_noise_list, axis = 0)\n",
    "# generate contact map\n",
    "contact_th = 500\n",
    "contact_map_noise = np.nanmean(distmap_noise_list < contact_th, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from source.domain_tools.distance import _sliding_window_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.signal import find_peaks\n",
    "distance_wd_dists = _sliding_window_dist(median_distance_map_rep1, _wd=8, \n",
    "                                         _dist_metric='normed_insulation')\n",
    "distance_wd_dists_ = _sliding_window_dist(median_distance_map_noise, _wd=8, \n",
    "                                         _dist_metric='normed_insulation')\n",
    "distance_peaks = find_peaks(-distance_wd_dists_, distance=5, prominence=0.013, width=3)\n",
    "\n",
    "fig = plt.figure(figsize=(9,2),dpi=200)\n",
    "plt.plot(-distance_wd_dists_, color='k', label='simulation', linewidth=1)\n",
    "for _p in distance_peaks[0]:\n",
    "    plt.vlines(_p, 0, 1, linewidth=0.5, linestyles=\"dotted\",color='k' )\n",
    "    \n",
    "distance_peaks = find_peaks(-distance_wd_dists, distance=5, prominence=0.013, width=3)\n",
    "\n",
    "plt.plot(-distance_wd_dists, color=[1,0.4,0], label='data', linewidth=1)\n",
    "for _p in distance_peaks[0]:\n",
    "    plt.vlines(_p, 0, 1, linewidth=0.5, linestyles=\"dotted\",color='k' )\n",
    "    \n",
    "    \n",
    "#plt.legend()\n",
    "plt.ylim([0,0.5])\n",
    "plt.xlim([0,651])\n",
    "plt.xlabel('Genomic coordinate')\n",
    "plt.ylabel('Insulation')\n",
    "plt.show()\n",
    "\n",
    "#folder_ = r'C:\\Users\\Bogdan\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\FinalFigures\\Figure S1\\100nmDisplacement_simmulation_newAnalysis'\n",
    "#fig.savefig(folder_+os.sep+'TADsInsulation_medianDistance.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contact_wd_dists = _sliding_window_dist(contact_map_rep1, _wd=8, _dist_metric='normed_insulation')\n",
    "contact_wd_dists_ = _sliding_window_dist(contact_map_noise, _wd=8, _dist_metric='normed_insulation')\n",
    "from scipy.signal import find_peaks\n",
    "fig = plt.figure(figsize=(9,2),dpi=200)\n",
    "\n",
    "plt.plot(contact_wd_dists_,color='k',linewidth=2,label='100nm displaced loci')\n",
    "plt.plot(contact_wd_dists,linewidth=1,color='g',label='original data')\n",
    "contact_peaks = find_peaks(contact_wd_dists, distance=5, prominence=0.022, width=3)\n",
    "for _p in contact_peaks[0]:\n",
    "    plt.vlines(_p, 0, 1, linewidth=0.5, linestyles=\"dotted\", color='k')\n",
    "\n",
    "\n",
    "contact_peaks_ = find_peaks(contact_wd_dists_, distance=5, prominence=0.022, width=3)\n",
    "for _p in contact_peaks_[0]:\n",
    "    plt.vlines(_p, 0, 1, linewidth=0.5, linestyles=\"dotted\", color='k')\n",
    "\n",
    "plt.legend()\n",
    "plt.ylim([0,0.5])\n",
    "plt.xlim([0,651])\n",
    "plt.xlabel('Genomic coordinate')\n",
    "plt.ylabel('Insulation')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TADs = contact_peaks[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hic_wd_dists = _sliding_window_dist(hic_raw_map, _wd=8, _dist_metric='normed_insulation')\n",
    "from scipy.signal import find_peaks\n",
    "fig = plt.figure(figsize=(9,2),dpi=200)\n",
    "\n",
    "plt.plot(hic_wd_dists,linewidth=1,color='r',label='original data')\n",
    "hic_peaks = find_peaks(hic_wd_dists, distance=5, prominence=0.08, width=3)\n",
    "for _p in hic_peaks[0]:\n",
    "    plt.vlines(_p, 0, 1, linewidth=0.5, linestyles=\"dotted\", color='k')\n",
    "\n",
    "\n",
    "plt.legend()\n",
    "plt.ylim([0,1])\n",
    "plt.xlim([0,651])\n",
    "plt.xlabel('Genomic coordinate')\n",
    "plt.ylabel('Insulation')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of positions of CTCF and rad21 in chr21\n",
    "#ctcf \n",
    "ctcf = [  9,  21,  33,  67,  73,  78, 139, 226, 231, 235, 242, 253, 256,\n",
    "       273, 284, 292, 301, 307, 339, 350, 355, 363, 366, 370, 373, 376,\n",
    "       381, 385, 390, 396, 402, 405, 410, 436, 440, 446, 456, 469, 472,\n",
    "       476, 482, 485, 488, 492, 500, 505, 508, 512, 520, 540, 543, 550,\n",
    "       554, 560, 565, 576, 580, 585, 589, 592, 595, 599, 602, 606, 615,\n",
    "       619, 622, 625, 628, 633, 636, 639]\n",
    "# rad21\n",
    "rad21=[ 21,  33,  67,  73, 139, 226, 231, 236, 242, 256, 273, 284, 292,\n",
    "       301, 305, 339, 350, 355, 363, 366, 370, 381, 386, 390, 396, 405,\n",
    "       410, 415, 436, 440, 446, 456, 469, 472, 482, 485, 492, 500, 505,\n",
    "       508, 512, 543, 550, 554, 560, 576, 581, 585, 589, 593, 596, 599,\n",
    "       602, 615, 619, 622, 625, 628, 633, 636]\n",
    "\n",
    "#A = [255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449, 450, 451, 452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463, 464, 465, 466, 467, 468, 469, 494, 495, 496, 497, 498, 499, 500, 575, 576, 577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619, 620, 621, 622, 623, 624, 625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650]\n",
    "#A_ = np.zeros(len(zxys[0])+2)\n",
    "#A_[np.array(A)+1]=1\n",
    "#AB_bds = np.abs(np.diff(A_))\n",
    "#AB_bds = np.where(AB_bds)[0]\n",
    "pts = data_rep1['dna_zxys']\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['font.size']=15\n",
    "matplotlib.rcParams['font.family']='Arial'\n",
    "fig = plt.figure(figsize=(3,5))\n",
    "bds_avg_ = TADs #from jun-han - 500nm#[20, 35, 52, 67, 80, 113, 139, 159, 179, 198, 213, 227, 254, 273, 298, 317, 340, 351, 365, 373, 388, 411, 439, 460, 471, 486, 507, 540, 550, 561, 575, 592, 604, 613, 627, 636, 644]\n",
    "dmat = np.abs(np.array([[bd2-bd1 for bd1 in bds_avg_] for bd2 in np.arange(len(pts[0]))],dtype=int))\n",
    "dmat = np.array([[bd2-bd1 for bd1 in bds_avg_] for bd2 in np.arange(len(pts[0]))],dtype=int)\n",
    "range_ = range(-15,15)\n",
    "yvec = np.array([np.median(cts[np.where(dmat==i)[0]]) for i in range_])\n",
    "plt.plot(np.array(range_)*50,yvec/len(pts)*100,'o-',color=[0.6]*3,label='Domains')\n",
    "#dmat = np.abs(np.array([[bd2-bd1 for bd1 in AB_bds] for bd2 in np.arange(len(pts[0]))],dtype=int))\n",
    "#dmat = np.array([[bd2-bd1 for bd1 in AB_bds] for bd2 in np.arange(len(pts[0]))],dtype=int)\n",
    "#yvec = np.array([np.median(cts[np.where(dmat==i)[0]]) for i in range_])\n",
    "#plt.plot(np.array(range_)*50,yvec/len(pts)*100,'ko-',label='A/B compartments')\n",
    "ctcf_rad21 = np.intersect1d(rad21,ctcf)\n",
    "dmat = np.array([[bd2-bd1 for bd1 in ctcf_rad21] for bd2 in np.arange(len(pts[0]))],dtype=int)\n",
    "yvec = np.array([np.median(cts[np.where(dmat==i)[0]]) for i in range_])\n",
    "plt.plot(np.array(range_)*50,yvec/len(pts)*100,'o-',color='orange',label='CTCF&RAD21')\n",
    "\n",
    "plt.yticks([5,7.5,10])\n",
    "plt.ylim([4,14])\n",
    "plt.xlabel('Genomic distance from boundary (kb)')\n",
    "plt.ylabel('Single-cell \\nboundary probability(%)')\n",
    "plt.legend()\n",
    "#folder_ = r'C:\\Users\\Bogdan\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\Bogdan_Figures\\Figure1\\base_images'\n",
    "save_file = os.path.join(figure_folder, f'Fig1D_chr21_sc-domain_prob_ctcf_rep1.pdf')\n",
    "print(save_file)\n",
    "plt.savefig(save_file, transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### insulation scores w/wo ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ichr=0\n",
    "ins = []\n",
    "bdr_ins = []\n",
    "\n",
    "dom_starts_fs = data_rep1['domain_starts']\n",
    "zxys = data_rep1['dna_zxys']\n",
    "for dom_ in tqdm(dom_starts_fs):\n",
    "    zxy_ = zxys[ichr]\n",
    "    for idom in range(1,len(dom_)-3):\n",
    "        a,b,c = dom_[idom],dom_[idom+1],dom_[idom+2]\n",
    "        #a,b,c = dom_[idom+1]-5,dom_[idom+1],dom_[idom+1]+5\n",
    "        zxy1 = zxy_[a:b]\n",
    "        zxy2 = zxy_[b:c]\n",
    "        med_in = np.nanmedian(np.concatenate([pdist(zxy1),pdist(zxy2)]))\n",
    "        med_out = np.nanmedian(cdist(zxy1,zxy2))\n",
    "        ins_ = med_out/med_in\n",
    "        ins.append(ins_)\n",
    "        bdr_ins.append(b)\n",
    "    ichr+=1    \n",
    "\n",
    "bdr_ins=np.array(bdr_ins)\n",
    "ins = np.array(ins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nonctcf = np.ones(len(zxys[0]))\n",
    "nonctcf[ctcf_rad21]=0\n",
    "nonctcf = np.nonzero(nonctcf)[0]\n",
    "\n",
    "fig = plt.figure()#figsize=(10,7))\n",
    "bins=np.linspace(0,4,75)\n",
    "plt.hist(ins[np.in1d(bdr_ins,nonctcf)],alpha=0.5,normed=True,bins=bins,color='gray',label = 'non-CTCF/RAD21')\n",
    "plt.hist(ins[np.in1d(bdr_ins,ctcf_rad21)],alpha=0.5,normed=True,bins=bins,color='orange',label = 'CTCF/RAD21')\n",
    "plt.xlabel('Boundary insulation score')\n",
    "plt.ylabel('Probability density function')\n",
    "plt.legend()\n",
    "plt.savefig(os.path.join(figure_folder, f'Fig1K_chr21_sc-domain_insulation_ctcf_rep1.pdf'), transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.5 CTCF end-end distance and radii of gyration of CTCF bound domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "dic_ctcf = {}\n",
    "dic_nonctcf = {}\n",
    "\n",
    "def rg_med(zxy):\n",
    "    \"\"\"computes radius of gyration\"\"\"\n",
    "    zxy_ = np.array(zxy)\n",
    "    zxy_ = zxy_[~np.isnan(zxy_[:,0])]\n",
    "    zxy_ = zxy_ - np.mean(zxy_,0)\n",
    "    return np.sqrt(np.mean(np.sum(zxy_**2,axis=-1)))\n",
    "\n",
    "def inD(zxy): \n",
    "    \"\"\"computest mean interdistance\"\"\"\n",
    "    return np.nanmean(pdist(zxy))\n",
    "\n",
    "dic_rgctcf = {}\n",
    "dic_rgnonctcf = {}\n",
    "\n",
    "dic_inDctcf = {}\n",
    "dic_inDnonctcf = {}\n",
    "\n",
    "dic_withinDctcf = {}\n",
    "dic_withinDnonctcf = {}\n",
    "###################### This does not save each ctcf pair in its own key but already groups by genomic distance\n",
    "ctcf_or_rad21 = np.union1d(ctcf,rad21)\n",
    "for ichr in tqdm(range(len(zxys))):\n",
    "    doms=dom_starts_fs[ichr][1:-1]\n",
    "    zxy = zxys[ichr]\n",
    "    for i in range(len(doms)-1):\n",
    "        e1,e2 = doms[i],doms[i+1]-1\n",
    "        dist_ = np.linalg.norm(zxy[e1]-zxy[e2])\n",
    "        gen_dist = e2-e1\n",
    "        rg_ = rg_med(zxy[e1:e2])\n",
    "        inD_ = inD(zxy[e1:e2])\n",
    "        if (e1 in ctcf_or_rad21) and (e2 in ctcf_or_rad21):\n",
    "            dic_ctcf[gen_dist]=dic_ctcf.get(gen_dist,[])+[dist_]\n",
    "            dic_rgctcf[gen_dist]=dic_rgctcf.get(gen_dist,[])+[rg_]\n",
    "            dic_inDctcf[gen_dist]=dic_inDctcf.get(gen_dist,[])+[inD_]\n",
    "        if not np.any(np.in1d([e1,e1+1,e1-1,e2,e2+1,e2-1],ctcf_or_rad21)):\n",
    "        #if not np.any(np.in1d([e1,e2],ctcf_or_rad21)):\n",
    "            dic_nonctcf[gen_dist]=dic_nonctcf.get(gen_dist,[])+[dist_]\n",
    "            dic_rgnonctcf[gen_dist]=dic_rgnonctcf.get(gen_dist,[])+[rg_]\n",
    "\n",
    "        e1p = np.random.randint(e1+1,e2-1)\n",
    "        e2p = np.random.randint(e1p,e2-1)\n",
    "        if not np.any(np.in1d([e1p,e1p+1,e1p-1,e2p,e2p+1,e2p-1],ctcf_or_rad21)):\n",
    "            gen_dist__ = abs(e1p-e2p)\n",
    "            dist__ = np.linalg.norm(zxy[e1p]-zxy[e2p])\n",
    "            dic_withinDnonctcf[gen_dist__]=dic_withinDnonctcf.get(gen_dist__,[])+[dist__]\n",
    "        for e1p in range(e1+1,e2-1):\n",
    "            for e2p in range(e1p,e2-1):\n",
    "                if (e1p in ctcf_or_rad21) and (e2p in ctcf_or_rad21):\n",
    "                    gen_dist__ = abs(e1p-e2p)\n",
    "                    dist__ = np.linalg.norm(zxy[e1p]-zxy[e2p])\n",
    "                    dic_withinDctcf[gen_dist__]=dic_withinDctcf.get(gen_dist__,[])+[dist__]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pickle.dump([dic_ctcf,dic_nonctcf,dic_rgctcf,dic_rgnonctcf,dic_inDctcf,dic_inDnonctcf,dic_withinDctcf,dic_withinDnonctcf],\n",
    "           open(r'C:\\Users\\Bogdan\\Dropbox\\Chromosome21_online\\rg_and_edge-edge_distance_v2_repeat_testPu','wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen_dists = np.sort(list(dic_ctcf.keys()))\n",
    "gen_dists = gen_dists[gen_dists<=28]\n",
    "gen_dists = gen_dists[gen_dists>=4]\n",
    "print([len(dic_nonctcf.get(gn,[])) for gn in gen_dists])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def boostrap_err2(x_,y_,func,N=1000,perc_min=5,perc_max=95):\n",
    "    elems = []\n",
    "    for istrap in range(N):\n",
    "        x__ = np.random.choice(x_,[len(x_)])\n",
    "        y__ = np.random.choice(y_,[len(y_)])\n",
    "        elems.append(func(x__,y__))\n",
    "    return (np.nanpercentile(elems,perc_min),np.nanpercentile(elems,perc_max))\n",
    "\n",
    "gen_dists = np.sort(list(dic_ctcf.keys()))\n",
    "gen_dists = gen_dists[gen_dists<=28]\n",
    "gen_dists = gen_dists[gen_dists>=4]\n",
    "def func(x,y): return np.nanmedian(x)/np.nanmedian(y)\n",
    "xelems = gen_dists*50\n",
    "meds_ctcf = [func(dic_ctcf.get(gn,[]),dic_withinDctcf.get(gn,[])) for gn in gen_dists]\n",
    "errs_ctcf = np.abs(np.array([boostrap_err2(dic_ctcf.get(gn,[]),dic_withinDctcf.get(gn,[]),func) \n",
    "                             for gn in tqdm(gen_dists)]).T-meds_ctcf)\n",
    "xelems = gen_dists*50\n",
    "meds_non = [func(dic_nonctcf.get(gn,[]),dic_withinDnonctcf.get(gn,[])) for gn in gen_dists]\n",
    "errs_non = np.abs(np.array([boostrap_err2(dic_nonctcf.get(gn,[]),dic_withinDnonctcf.get(gn,[]),func) \n",
    "                             for gn in tqdm(gen_dists)]).T-meds_non)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure()\n",
    "xelems = gen_dists*50\n",
    "plt.errorbar(xelems,meds_ctcf,\n",
    "             yerr=errs_ctcf,\n",
    "             color='orange',mec='k',label='CTCF/cohesin domains',marker='o')\n",
    "\n",
    "plt.errorbar(xelems,meds_non,\n",
    "             yerr=errs_non,\n",
    "             color='gray',mec='k',label='non-CTCF/cohesin domains',marker='o')\n",
    "plt.legend()\n",
    "plt.ylim([0.5,1.75])\n",
    "plt.ylabel('Median edge distance/ \\nMedian distance within domains')\n",
    "plt.xlabel('Genomic distance (kb)')\n",
    "#folder_ = r'C:\\Users\\Bogdan\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\FinalFigures\\Figure 1\\subpanels'\n",
    "#fig.savefig(folder_+os.sep+r'Fig1L_new.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Radius of gyration\n",
    "def boostrap_err(x_,func,N=1000,perc_min=5,perc_max=95):\n",
    "    elems = []\n",
    "    for istrap in range(N):\n",
    "        elems.append(func(np.random.choice(x_,[len(x_)])))\n",
    "    return (np.nanpercentile(elems,perc_min),np.nanpercentile(elems,perc_max))\n",
    "\n",
    "gen_dists = np.sort(list(dic_rgctcf.keys()))\n",
    "gen_dists = gen_dists[gen_dists<=28]\n",
    "gen_dists = gen_dists[gen_dists>=4]\n",
    "\n",
    "\n",
    "func = np.nanmedian\n",
    "\n",
    "xelems = gen_dists*50\n",
    "meds_ctcf_rg = [func(dic_rgctcf[gn]) for gn in gen_dists]\n",
    "errs_ctcf_rg = np.abs(np.array([boostrap_err(dic_rgctcf[gn],func) for gn in gen_dists]).T-meds_ctcf_rg)\n",
    "xelems = gen_dists*50\n",
    "meds_non_rg = [func(dic_rgnonctcf[gn]) for gn in gen_dists]\n",
    "errs_non_rg = np.abs(np.array([boostrap_err(dic_rgnonctcf[gn],func) for gn in gen_dists]).T-meds_non_rg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(5,5))#figsize=(8,3))\n",
    "\n",
    "plt.errorbar(xelems,meds_ctcf_rg,\n",
    "             yerr=errs_ctcf_rg,\n",
    "             color='orange',mec='k',label='CTCF/cohesin domains',marker='o')\n",
    "\n",
    "plt.errorbar(xelems,meds_non_rg,\n",
    "             yerr=errs_non_rg,\n",
    "             color='gray',mec='k',label='non-CTCF/cohesin domains',marker='o')\n",
    "plt.ylabel('Radius of gyration(nm)')\n",
    "plt.xlabel('Genomic distance (kb)')\n",
    "plt.legend()\n",
    "save_file = os.path.join(figure_folder, f'Fig1M_chr21_domain_rg_ctcf_rep1.pdf')\n",
    "plt.savefig(save_file, transparent=True)\n",
    "print(save_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Radius of gyration vs genomic distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dom_starts_fs = data_rep1['domain_starts']\n",
    "zxys = data_rep1['dna_zxys'][:,:,1:]\n",
    "pts=zxys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rg_med(zxy):\n",
    "    \"\"\"computes radius of gyration\"\"\"\n",
    "    zxy_ = np.array(zxy)\n",
    "    zxy_ = zxy_[~np.isnan(zxy_[:,0])]\n",
    "    zxy_ = zxy_ - np.mean(zxy_,0)\n",
    "    return np.sqrt(np.mean(np.sum(zxy_**2,axis=-1)))\n",
    "\n",
    "dic_rg = {}\n",
    "for ichr in tqdm(range(len(pts))):\n",
    "    doms=dom_starts_fs[ichr][1:-1]\n",
    "    zxy = pts[ichr]\n",
    "    for i in range(len(doms)-1):\n",
    "        e1,e2 = doms[i],doms[i+1]-1\n",
    "        start = e1\n",
    "        end=e2#-1\n",
    "        rg_ = rg_med(zxy[start:end])\n",
    "        key = end-start\n",
    "        dic_rg[key] = dic_rg.get(key,[])+[rg_]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10,5))\n",
    "keys = np.sort(list(dic_rg.keys()))\n",
    "keys = keys[keys>=4]\n",
    "plt.boxplot([dic_rg[gn] for gn in keys][:100-4], notch=True, showfliers=False,whis = [10, 90]);\n",
    "xlab = np.arange(6)\n",
    "plt.xticks((xlab-0.2)*1000/50,xlab);\n",
    "\n",
    "plt.ylabel('Radius of gyration (nm)')\n",
    "plt.xlabel('Genomic size of single-cell domains (Mb)')\n",
    "\n",
    "save_file = os.path.join(figure_folder, f'FigS1N_chr21_domain_rgs_vs_genomic.pdf')\n",
    "plt.savefig(save_file, transparent=True)\n",
    "print(save_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Characterization domain and compartment in G1/G2-S cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data rep2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load from file and extract info\n",
    "import csv\n",
    "rep2_info_dict = {}\n",
    "with open(rep2_filename, 'r') as _handle:\n",
    "    _reader = csv.reader(_handle, delimiter='\\t', quotechar='|')\n",
    "    _headers = next(_reader)\n",
    "    print(_headers)\n",
    "    # create keys for each header\n",
    "    for _h in _headers:\n",
    "        rep2_info_dict[_h] = []\n",
    "    # loop through content\n",
    "    for _contents in _reader:\n",
    "        for _h, _info in zip(_headers,_contents):\n",
    "            rep2_info_dict[_h].append(_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "# clean up infoa\n",
    "data_rep2 = {'params':{}}\n",
    "\n",
    "# clean up genomic coordiantes\n",
    "region_names = np.array([_n for _n in sorted(region_names, key=lambda s:int(s.split(':')[1].split('-')[0]))])\n",
    "region_starts = np.array([int(_n.split(':')[1].split('-')[0]) for _n in region_names])\n",
    "region_ends = np.array([int(_n.split(':')[1].split('-')[1]) for _n in region_names])[np.argsort(region_starts)]\n",
    "region_starts = np.sort(region_starts)\n",
    "\n",
    "mid_positions = ((region_starts + region_ends)/2).astype(np.int)\n",
    "mid_positions_Mb = np.round(mid_positions / 1e6, 2) \n",
    "\n",
    "# clean up chrom copy number\n",
    "chr_nums = np.array([int(_info) for _info in rep2_info_dict['Chromosome copy number']])\n",
    "chr_ids, region_cts = np.unique(chr_nums, return_counts=True)\n",
    "dna_zxys_list = [[[] for _start in region_starts] for _id in chr_ids]\n",
    "\n",
    "# clean up zxy\n",
    "for _z,_x,_y,_reg_info, _cid in tqdm(zip(rep2_info_dict['Z(nm)'],rep2_info_dict['X(nm)'],\\\n",
    "                                         rep2_info_dict['Y(nm)'],rep2_info_dict['Genomic coordinate'],\\\n",
    "                                         rep2_info_dict['Chromosome copy number'])):\n",
    "    # get chromosome inds\n",
    "    _cid = int(_cid)\n",
    "    _cind = np.where(chr_ids == _cid)[0][0]\n",
    "    \n",
    "    # get region indices\n",
    "    _start = int(_reg_info.split(':')[1].split('-')[0])\n",
    "    _rind = np.where(region_starts==_start)[0][0]\n",
    "    \n",
    "    dna_zxys_list[_cind][_rind] = np.array([float(_z),float(_x), float(_y)])\n",
    "\n",
    "# merge together\n",
    "dna_zxys_list = np.array(dna_zxys_list)\n",
    "data_rep2['chrom_ids'] = chr_ids\n",
    "data_rep2['region_names'] = region_names\n",
    "data_rep2['mid_position_Mb'] = mid_positions_Mb\n",
    "data_rep2['dna_zxys'] = dna_zxys_list\n",
    "\n",
    "# clean up tss and transcription\n",
    "if 'Gene names' in rep2_info_dict:\n",
    "    import re\n",
    "    # first extract number of genes\n",
    "    gene_names = []\n",
    "    for _gene_info, _trans_info, _tss_coord in zip(rep2_info_dict['Gene names'],\n",
    "                                                   rep2_info_dict['Transcription'],\n",
    "                                                   rep2_info_dict['TSS ZXY(nm)']):\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            for _gene in _genes:\n",
    "                if _gene not in gene_names:\n",
    "                    gene_names.append(_gene)\n",
    "    print(f\"{len(gene_names)} genes exist in this dataset.\")\n",
    "    \n",
    "    # initialize gene and transcription\n",
    "    tss_zxys_list = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    transcription_profiles = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    # loop through to get info\n",
    "    for _cid, _gene_info, _trans_info, _tss_locations in tqdm(zip(rep2_info_dict['Chromosome copy number'],\n",
    "                                                                  rep2_info_dict['Gene names'],\n",
    "                                                                  rep2_info_dict['Transcription'],\n",
    "                                                                  rep2_info_dict['TSS ZXY(nm)'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        # process if there are genes in this region:\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            _transcribes = _trans_info.split(';')[:-1]\n",
    "            _tss_zxys = _tss_locations.split(';')[:-1]\n",
    "            for _gene, _transcribe, _tss_zxy in zip(_genes, _transcribes, _tss_zxys):\n",
    "                # get gene index\n",
    "                _gind = gene_names.index(_gene)\n",
    "                # get transcription profile\n",
    "                if _transcribe == 'on':\n",
    "                    transcription_profiles[_cind][_gind] = True\n",
    "                else:\n",
    "                    transcription_profiles[_cind][_gind] = False\n",
    "                # get coordinates\n",
    "                _tss_zxy = np.array([np.float(_c) for _c in re.split(r'\\s+', _tss_zxy.split('[')[1].split(']')[0]) if _c != ''])\n",
    "                tss_zxys_list[_cind][_gind] = _tss_zxy\n",
    "                \n",
    "    tss_zxys_list = np.array(tss_zxys_list)\n",
    "    transcription_profiles = np.array(transcription_profiles)\n",
    "    data_rep2['gene_names'] = gene_names\n",
    "    data_rep2['tss_zxys'] = tss_zxys_list\n",
    "    data_rep2['trans_pfs'] = transcription_profiles\n",
    "\n",
    "# clean up cell_cycle states\n",
    "if 'Cell cycle state' in rep2_info_dict:\n",
    "    cell_cycle_types = np.unique(rep2_info_dict['Cell cycle state'])\n",
    "    cell_cycle_flag_dict = {_k:[[] for _id in chr_ids] for _k in cell_cycle_types if _k != 'ND'}\n",
    "    for _cid, _state in tqdm(zip(rep2_info_dict['Chromosome copy number'],rep2_info_dict['Cell cycle state'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        if np.array([_v[_cind]==[] for _k,_v in cell_cycle_flag_dict.items()]).any():\n",
    "            for _k,_v in cell_cycle_flag_dict.items():\n",
    "                if _k == _state:\n",
    "                    _v[_cind] = True\n",
    "                else:\n",
    "                    _v[_cind] = False\n",
    "    # append to data\n",
    "    for _k, _v in cell_cycle_flag_dict.items():\n",
    "        data_rep2[f'{_k}_flags'] = np.array(_v)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## call domains for rep2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import source.domain_tools.DomainAnalysis as da\n",
    "import multiprocessing as mp\n",
    "\n",
    "num_threads=32\n",
    "domain_corr_cutoff = 0.75 \n",
    "domain_dist_cutoff = 500 # nm\n",
    "\n",
    "_domain_args = [(_zxys, 4, 1000, domain_corr_cutoff, domain_dist_cutoff) \n",
    "                     for _zxys in data_rep2['dna_zxys']]\n",
    "_domain_time = time.time()\n",
    "\n",
    "print(f\"Multiprocessing call domain starts\", end=' ')\n",
    "if 'domain_starts' not in data_rep2:\n",
    "    with mp.Pool(num_threads) as domain_pool:\n",
    "        domain_results = domain_pool.starmap(da.get_dom_starts_cor, _domain_args)\n",
    "        domain_pool.close()\n",
    "        domain_pool.join()\n",
    "        domain_pool.terminate()\n",
    "    # save\n",
    "    data_rep2['domain_starts'] = [np.array(_r[-1]) for _r in domain_results]\n",
    "    data_rep2['params']['domain_corr_cutoff'] = domain_corr_cutoff\n",
    "    data_rep2['params']['domain_dist_cutoff'] = domain_dist_cutoff\n",
    "    \n",
    "print(f\"in {time.time()-_domain_time:.3f}s.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "def rg_mean(zxy):\n",
    "    \"\"\"computes radius of gyration\"\"\"\n",
    "    zxy_ = np.array(zxy)\n",
    "    zxy_ = zxy_[~np.isnan(zxy_[:,0])]\n",
    "    zxy_ = zxy_ - np.mean(zxy_,0)\n",
    "    return np.sqrt(np.mean(np.sum(zxy_**2,axis=-1)))\n",
    "\n",
    "g1_rgs = []\n",
    "g2_rgs = []\n",
    "\n",
    "for _i, (pt_,doms_) in tqdm(enumerate(zip(data_rep2['dna_zxys'],data_rep2['domain_starts']))):\n",
    "    for i1,i2 in zip(doms_[1:-2],doms_[2:-1]):\n",
    "        if data_rep2['G1_flags'][_i]:\n",
    "            g1_rgs.append(rg_mean(pt_[i1:i2]))\n",
    "        elif data_rep2['G2/S_flags'][_i]:\n",
    "            g2_rgs.append(rg_mean(pt_[i1:i2]))\n",
    "g1_rgs = np.array(g1_rgs)\n",
    "g2_rgs = np.array(g2_rgs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "rg_limits = [0,1500]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width),dpi=600)\n",
    "\n",
    "ax.hist(g1_rgs, 50, range=(min(rg_limits),max(rg_limits)),\n",
    "        density=True, alpha=0.5, \n",
    "        color=[0.2,0.5,0.5], label=f'G1, median={np.nanmedian(g1_rgs):.0f}nm')\n",
    "ax.hist(g2_rgs, 50, range=(min(rg_limits),max(rg_limits)),\n",
    "        density=True, alpha=0.5, \n",
    "        color=[1,0.2,0.2], label=f'G2/S, median={np.nanmedian(g2_rgs):.0f}nm')\n",
    "\n",
    "ax.legend(fontsize=_font_size-1, loc='upper right')\n",
    "\n",
    "ax.set_xlabel(\"Radius of gyration (nm)\", fontsize=_font_size, labelpad=1)\n",
    "ax.set_ylabel(\"Probability density\", fontsize=_font_size, labelpad=1)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for a_ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(rg_limits)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'LFig5A_chr21-repeat_radius_of_gyration_G1-G2.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "resolution = 0.05 # Mb\n",
    "\n",
    "g1_gsizes = []\n",
    "g2_gsizes = []\n",
    "\n",
    "for _i, (pt_,doms_) in tqdm(enumerate(zip(data_rep2['dna_hzxys'][:,:,1:],data_rep2['domain_starts']))):\n",
    "    for i1,i2 in zip(doms_[1:-2],doms_[2:-1]):\n",
    "        if data_rep2['G1_flags'][_i]:\n",
    "            g1_gsizes.append((i2-i1)*resolution)\n",
    "        elif data_rep2['G2_flags'][_i]:\n",
    "            g2_gsizes.append((i2-i1)*resolution)\n",
    "g1_gsizes = np.array(g1_gsizes)\n",
    "g2_gsizes = np.array(g2_gsizes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "rg_limits = [0,4]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width),dpi=600)\n",
    "\n",
    "ax.hist(g1_gsizes, 40, range=(min(rg_limits),max(rg_limits)),\n",
    "        density=True, alpha=0.5, \n",
    "        color=[0.2,0.5,0.5], label=f'G1, median={np.nanmedian(g1_gsizes):.2f}Mb')\n",
    "ax.hist(g2_gsizes, 40, range=(min(rg_limits),max(rg_limits)),\n",
    "        density=True, alpha=0.5, \n",
    "        color=[1,0.2,0.2], label=f'G2/S, median={np.nanmedian(g2_gsizes):.2f}Mb')\n",
    "\n",
    "ax.legend(fontsize=_font_size-1, loc='upper right')\n",
    "\n",
    "ax.set_xlabel(\"Genomic size (Mb)\", fontsize=_font_size, labelpad=1)\n",
    "ax.set_ylabel(\"Probability density\", fontsize=_font_size, labelpad=1)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for a_ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(rg_limits)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'LFig5B_chr21-repeat_domain_size_G1-G2.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
