{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Library design for CTP-05, Exonic RNA\n",
    "\n",
    "\n",
    "by Pu Zheng\n",
    "\n",
    "This library design is based on target regions designed by Stephen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-16T23:32:23.430571Z",
     "start_time": "2019-06-16T23:32:21.112844Z"
    }
   },
   "outputs": [],
   "source": [
    "%run \"E:\\Users\\puzheng\\Documents\\Startup_py3.py\"\n",
    "sys.path.append(r\"E:\\Users\\puzheng\\Documents\")\n",
    "\n",
    "import ImageAnalysis3\n",
    "from ImageAnalysis3 import get_img_info, visual_tools, corrections, library_tools\n",
    "\n",
    "from ImageAnalysis3.library_tools import LibraryDesigner as ld\n",
    "from ImageAnalysis3.library_tools import LibraryTools as lt\n",
    "\n",
    "%matplotlib notebook\n",
    "print(os.getpid())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Some Global used folder names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# folder for genomic info\n",
    "genome_folder = r'E:\\Genomes\\hg38'\n",
    "\n",
    "# Library directories\n",
    "library_folder = r'X:\\Libraries\\CTP-06'\n",
    "# folder for sub-pool\n",
    "pool_folder = os.path.join(library_folder, 'chr21_promoters')\n",
    "# folder for fasta sequences\n",
    "sequence_folder = os.path.join(pool_folder, 'sequences')\n",
    "\n",
    "# folder to save result probes\n",
    "save_folder = os.path.join(pool_folder, 'reports')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Design probes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## probe_designer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-18T18:49:43.987075Z",
     "start_time": "2019-06-18T18:49:43.968061Z"
    }
   },
   "source": [
    "## Construct count table with all the 17-mers in the genome\n",
    "\n",
    "Only do this if you don't have pre-built 17-mer\n",
    "\n",
    "However you can do almost the same thing for your own library during quality check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genome\n",
    "names,seqs = lt.fastaread(os.path.join(genome_folder, 'human_genome.fasta'))\n",
    "ct = ld.countTable(word=17,save_file=os.path.join(genome_folder, 'hg38_genome_17w.npy'), sparse=False)\n",
    "for iseq in tqdm(range(len(seqs))):\n",
    "    print(names[iseq], end=' ')\n",
    "    ct.consume(seqs[iseq],verbose=True)\n",
    "ct.complete(verbose=True)\n",
    "ct.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeats\n",
    "names,seqs = lt.fastaread(os.path.join(genome_folder, 'HumanRepeats.fasta'))\n",
    "ct = ld.countTable(word=17,save_file=os.path.join(genome_folder, 'hg38_genome_repeats_17w.npy'), sparse=False)\n",
    "for iseq in tqdm(range(len(seqs))):\n",
    "    print(names[iseq], end=' ')\n",
    "    ct.consume(seqs[iseq],verbose=True)\n",
    "ct.complete(verbose=True)\n",
    "ct.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run probe_designer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Construct SI15 probes-whole_chr21\n",
    "#import time,os,sys\n",
    "#reload(ld)\n",
    "#import matplotlib.pyplot as plt\n",
    "#plt.switch_backend('agg')\n",
    "reload(ld)\n",
    "# folder for genomic info\n",
    "genome_folder = r'E:\\Genomes\\hg38'\n",
    "# Indeces\n",
    "genmoe_index = os.path.join(genome_folder, r'hg38_genome_17w.npy')\n",
    "repeat_index = os.path.join(genome_folder, r'hg38_genome_repeats_17w.npy')\n",
    "transcriptome_index = os.path.join(genome_folder, r'hg38_transcriptome_17w.npy')\n",
    "\n",
    "# Library directories\n",
    "library_folder = r'X:\\Libraries\\CTP-06'\n",
    "# folder for sub-pool\n",
    "pool_folder = os.path.join(library_folder, 'chr21_promoters')\n",
    "# folder for fasta sequences\n",
    "sequence_folder = os.path.join(pool_folder, 'sequences')\n",
    "# get input files \n",
    "input_files = glob.glob(os.path.join(sequence_folder, '*.fasta'))\n",
    "# folder to save result probes\n",
    "save_folder = os.path.join(pool_folder, 'reports')\n",
    "\n",
    "overwrite = True\n",
    "\n",
    "if not os.path.exists(save_folder):\n",
    "    os.makedirs(save_folder);\n",
    "# Loop through all input files\n",
    "for in_file in input_files:\n",
    "    #print in_file\n",
    "    save_file = os.path.join(save_folder, os.path.basename(in_file).replace('.fasta','.pbr'))#'/pb_reports_reg_'+str(-i-1)+'.pbr'\n",
    "    if not os.path.exists(save_file) or overwrite: # if probe report file doesnt exist\n",
    "        \n",
    "        local_genome_fl = in_file\n",
    "        print(in_file)\n",
    "        pb_designer = ld.pb_reports_class(\n",
    "            sequence_dic={'file':in_file,'use_revc':False,'use_kmer':True},\n",
    "            map_dic={'genome':{'file':genmoe_index,'use_revc':True,'use_kmer':True},\n",
    "                  'rep_genome':{'file':repeat_index,'use_revc':True,'use_kmer':True},\n",
    "                  'local_genome':{'file':local_genome_fl,'force_list':True,'use_revc':True,'use_kmer':True}},\n",
    "            save_file=save_file,\n",
    "            params_dic={'word_size':17,'pb_len':42,'buffer_len':0,'max_count':2**16-1,'check_on_go':False,'auto':False},\n",
    "            dic_check={('genome','local_genome'):75,'rep_genome':0,'gc':[0.25,0.85],'tm':70, })\n",
    "\n",
    "        pb_designer.computeOTmaps()\n",
    "        pb_designer.compute_pb_report()\n",
    "        pb_designer.perform_check_end()\n",
    "        pb_designer.plots()\n",
    "        pb_designer.save_csv()\n",
    "        print(f\"Number of probes kept: {len(pb_designer.pb_reports_keep)}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Screen probes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## screen probes against an existing fasta (previous library)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-13T01:48:25.177478Z",
     "start_time": "2019-06-13T01:48:25.162466Z"
    },
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "library_folder = r'X:\\Libraries\\CTP-06\\chr21_promoters'\n",
    "sequence_folder = os.path.join(library_folder, 'sequences')\n",
    "report_folder = os.path.join(library_folder, 'reports')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-13T19:08:45.570045Z",
     "start_time": "2019-06-13T18:46:55.947624Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Screen probes against given fasta file:X:\\Libraries\\CTP-04\\chr21\\filtered_blast_centered_probes.fasta\n",
      "-- start reading 86 probe files\n",
      "-- constructing reference table for fasta file\n",
      "Mapping no. of seqs: 231435\n",
      "--- 112 / 112 probes kept for ADAMTS1\n",
      "--- 32 / 108 probes kept for ADAMTS5\n",
      "--- 36 / 109 probes kept for ADARB1\n",
      "--- 21 / 85 probes kept for AGPAT3\n",
      "--- 62 / 64 probes kept for ANKRD20A11P\n",
      "--- 25 / 97 probes kept for APP\n",
      "--- 32 / 101 probes kept for ATP5PF\n",
      "--- 35 / 109 probes kept for BACE2\n",
      "--- 82 / 82 probes kept for BACH1\n",
      "--- 26 / 68 probes kept for BRWD1\n",
      "--- 52 / 96 probes kept for BTG3\n",
      "--- 33 / 94 probes kept for C21orf91\n",
      "--- 96 / 96 probes kept for C2CD2\n",
      "--- 16 / 75 probes kept for CBR3\n",
      "--- 28 / 87 probes kept for CCT8\n",
      "--- 29 / 76 probes kept for CHAF1B\n",
      "--- 109 / 109 probes kept for COL18A1\n",
      "--- 32 / 90 probes kept for COL6A1\n",
      "--- 105 / 105 probes kept for COL6A2\n",
      "--- 29 / 99 probes kept for CRYZL1\n",
      "--- 34 / 85 probes kept for CXADR\n",
      "--- 56 / 56 probes kept for CYP4F29P\n",
      "--- 45 / 84 probes kept for DIP2A\n",
      "--- 41 / 87 probes kept for DONSON\n",
      "--- 78 / 90 probes kept for DSCR9\n",
      "--- 31 / 97 probes kept for DYRK1A\n",
      "--- 109 / 109 probes kept for ETS2\n",
      "--- 20 / 106 probes kept for EVA1C\n",
      "--- 94 / 97 probes kept for FAM207A\n",
      "--- 39 / 84 probes kept for HLCS\n",
      "--- 96 / 96 probes kept for HMGN1\n",
      "--- 29 / 103 probes kept for HSPA13\n",
      "--- 34 / 105 probes kept for HUNK\n",
      "--- 32 / 73 probes kept for IFNAR1\n",
      "--- 21 / 83 probes kept for IFNAR2\n",
      "--- 29 / 91 probes kept for IL10RB\n",
      "--- 35 / 112 probes kept for ITGB2\n",
      "--- 30 / 97 probes kept for JAM2\n",
      "--- 94 / 94 probes kept for LCA5L\n",
      "--- 35 / 106 probes kept for LSS\n",
      "--- 37 / 82 probes kept for LTN1\n",
      "--- 20 / 83 probes kept for MCM3AP\n",
      "--- 27 / 86 probes kept for MIS18A\n",
      "--- 15 / 61 probes kept for MORC3\n",
      "--- 79 / 102 probes kept for MRPL39\n",
      "--- 34 / 99 probes kept for MRPS6\n",
      "--- 30 / 70 probes kept for MX1\n",
      "--- 79 / 100 probes kept for MX2\n",
      "--- 28 / 84 probes kept for N6AMT1\n",
      "--- 103 / 103 probes kept for NCAM2\n",
      "--- 36 / 112 probes kept for NRIP1\n",
      "--- 46 / 80 probes kept for PAXBP1\n",
      "--- 107 / 107 probes kept for PCBP3\n",
      "--- 30 / 102 probes kept for PCNT\n",
      "--- 107 / 107 probes kept for PDE9A\n",
      "--- 24 / 106 probes kept for PDXK\n",
      "--- 91 / 91 probes kept for PFKL\n",
      "--- 28 / 85 probes kept for PKNOX1\n",
      "--- 35 / 102 probes kept for PRDM15\n",
      "--- 34 / 79 probes kept for PRMT2\n",
      "--- 35 / 97 probes kept for PSMG1\n",
      "--- 22 / 109 probes kept for PTTG1IP\n",
      "--- 73 / 96 probes kept for PWP2\n",
      "--- 32 / 90 probes kept for RCAN1\n",
      "--- 32 / 70 probes kept for RRP1\n",
      "--- 62 / 96 probes kept for RRP1B\n",
      "--- 72 / 72 probes kept for RUNX1\n",
      "--- 40 / 107 probes kept for SCAF4\n",
      "--- 25 / 86 probes kept for SETD4\n",
      "--- 53 / 93 probes kept for SIK1\n",
      "--- 107 / 107 probes kept for SIM2\n",
      "--- 41 / 73 probes kept for SLC19A1\n",
      "--- 95 / 95 probes kept for SLC37A1\n",
      "--- 91 / 91 probes kept for SOD1\n",
      "--- 56 / 101 probes kept for SON\n",
      "--- 36 / 101 probes kept for SUMO3\n",
      "--- 32 / 96 probes kept for SYNJ1\n",
      "--- 30 / 104 probes kept for TIAM1\n",
      "--- 37 / 86 probes kept for TRAPPC10\n",
      "--- 41 / 103 probes kept for TTC3\n",
      "--- 100 / 100 probes kept for UBE2G2\n",
      "--- 87 / 87 probes kept for URB1\n",
      "--- 19 / 76 probes kept for USP16\n",
      "--- 31 / 94 probes kept for USP25\n",
      "--- 18 / 79 probes kept for WDR4\n",
      "--- 22 / 73 probes kept for WRB\n"
     ]
    }
   ],
   "source": [
    "reload(library_tools)\n",
    "pb_dict = library_tools.Screen_probe_against_fasta(report_folder, \n",
    "                                                   r'X:\\Libraries\\CTP-04\\chr21\\filtered_blast_centered_probes.fasta',\n",
    "                                                   overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-17T15:11:43.097070Z",
     "start_time": "2019-06-17T15:11:25.811195Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Check raedouts->adaptors against fasta\n",
      "-- 252 readout loaded\n",
      "Mapping no. of seqs: 198\n",
      "Mapping no. of seqs: 1125\n",
      "Mapping no. of seqs: 198\n",
      "Mapping no. of seqs: 1125\n",
      "Mapping no. of seqs: 198\n",
      "Mapping no. of seqs: 1125\n",
      "Mapping no. of seqs: 198\n",
      "Mapping no. of seqs: 1125\n",
      "Mapping no. of seqs: 198\n",
      "Mapping no. of seqs: 1125\n",
      "-- 199 readous are kept.\n"
     ]
    }
   ],
   "source": [
    "from ImageAnalysis3.library_tools import readouts\n",
    "readout_cand_file = r'E:\\Users\\puzheng\\Documents\\Adaptors\\NDBs_new.fasta'\n",
    "adaptor_site_file = r'E:\\Users\\puzheng\\Documents\\Adaptors\\Adaptor_sites.fasta'\n",
    "barcode_dir = r'W:\\Pu\\Readouts'\n",
    "existing_readout_files = [os.path.join(barcode_dir, 'Stvs.fasta'),\n",
    "                          os.path.join(barcode_dir, 'NDBs.fasta')]\n",
    "saved_readouts = readouts.Check_adaptors_against_fasta(readout_cand_file, adaptor_site_file, existing_readout_files,\n",
    "                                                       save=True, save_adaptors=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:55:01.774402Z",
     "start_time": "2019-06-14T01:55:01.755389Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Splitting 199 readouts into 3 channels\n",
      "-- saving 67 readouts into file:\\\\SMIRNOV\\Chromatin_NAS_3\\Pu\\Readouts\\NDB_new_0.fasta\n",
      "-- saving 66 readouts into file:\\\\SMIRNOV\\Chromatin_NAS_3\\Pu\\Readouts\\NDB_new_1.fasta\n",
      "-- saving 66 readouts into file:\\\\SMIRNOV\\Chromatin_NAS_3\\Pu\\Readouts\\NDB_new_2.fasta\n"
     ]
    }
   ],
   "source": [
    "reload(library_tools)\n",
    "splitted_readouts = library_tools.Split_readouts_into_channels(saved_readouts, num_channels=3, save_name='NDB_new')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-13T01:39:55.011501Z",
     "start_time": "2019-06-13T01:39:54.968468Z"
    }
   },
   "source": [
    "## Patch Barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:56:07.945686Z",
     "start_time": "2019-06-14T01:56:07.852617Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Picked primer: ID: W1A03_primer_2\n",
      "Name: W1A03_primer_2\n",
      "Description: W1A03_primer_2\n",
      "Number of features: 0\n",
      "Seq('CCCGCAATGGCTGACAACCG', SingleLetterAlphabet())\n",
      "- Picked primer: ID: W1A10_primer_9\n",
      "Name: W1A10_primer_9\n",
      "Description: W1A10_primer_9\n",
      "Number of features: 0\n",
      "Seq('TAATACGACTCACTATAGGGATTGCCGCATGGTTTCCG', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "reload(library_tools)\n",
    "\n",
    "# folder for this library\n",
    "library_folder = r'X:\\Libraries\\CTP-06\\chr21_promoters'\n",
    "\n",
    "# extract primers and readouts\n",
    "primers = library_tools.load_primers([2,9])\n",
    "unique_readouts = library_tools.load_readouts(len(pb_dict), 'NDB_New', _num_colors=3, _start_id=0)\n",
    "\n",
    "readout_dict = {'u': unique_readouts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 402,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:57:44.608682Z",
     "start_time": "2019-06-14T01:57:42.673243Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Assemble probes by given target sequences, readouts and primers.\n",
      "-- included readout types: ['u']\n",
      "--- assemblying 112 probes in region: ADAMTS1\n",
      "--- assemblying 32 probes in region: ADAMTS5\n",
      "--- assemblying 36 probes in region: ADARB1\n",
      "--- assemblying 21 probes in region: AGPAT3\n",
      "--- assemblying 62 probes in region: ANKRD20A11P\n",
      "--- assemblying 25 probes in region: APP\n",
      "--- assemblying 32 probes in region: ATP5PF\n",
      "--- assemblying 35 probes in region: BACE2\n",
      "--- assemblying 82 probes in region: BACH1\n",
      "--- assemblying 26 probes in region: BRWD1\n",
      "--- assemblying 52 probes in region: BTG3\n",
      "--- assemblying 33 probes in region: C21orf91\n",
      "--- assemblying 96 probes in region: C2CD2\n",
      "--- assemblying 16 probes in region: CBR3\n",
      "--- assemblying 28 probes in region: CCT8\n",
      "--- assemblying 29 probes in region: CHAF1B\n",
      "--- assemblying 109 probes in region: COL18A1\n",
      "--- assemblying 32 probes in region: COL6A1\n",
      "--- assemblying 105 probes in region: COL6A2\n",
      "--- assemblying 29 probes in region: CRYZL1\n",
      "--- assemblying 34 probes in region: CXADR\n",
      "--- assemblying 56 probes in region: CYP4F29P\n",
      "--- assemblying 45 probes in region: DIP2A\n",
      "--- assemblying 41 probes in region: DONSON\n",
      "--- assemblying 78 probes in region: DSCR9\n",
      "--- assemblying 31 probes in region: DYRK1A\n",
      "--- assemblying 109 probes in region: ETS2\n",
      "--- assemblying 20 probes in region: EVA1C\n",
      "--- assemblying 94 probes in region: FAM207A\n",
      "--- assemblying 39 probes in region: HLCS\n",
      "--- assemblying 96 probes in region: HMGN1\n",
      "--- assemblying 29 probes in region: HSPA13\n",
      "--- assemblying 34 probes in region: HUNK\n",
      "--- assemblying 32 probes in region: IFNAR1\n",
      "--- assemblying 21 probes in region: IFNAR2\n",
      "--- assemblying 29 probes in region: IL10RB\n",
      "--- assemblying 35 probes in region: ITGB2\n",
      "--- assemblying 30 probes in region: JAM2\n",
      "--- assemblying 94 probes in region: LCA5L\n",
      "--- assemblying 35 probes in region: LSS\n",
      "--- assemblying 37 probes in region: LTN1\n",
      "--- assemblying 20 probes in region: MCM3AP\n",
      "--- assemblying 27 probes in region: MIS18A\n",
      "--- assemblying 15 probes in region: MORC3\n",
      "--- assemblying 79 probes in region: MRPL39\n",
      "--- assemblying 34 probes in region: MRPS6\n",
      "--- assemblying 30 probes in region: MX1\n",
      "--- assemblying 79 probes in region: MX2\n",
      "--- assemblying 28 probes in region: N6AMT1\n",
      "--- assemblying 103 probes in region: NCAM2\n",
      "--- assemblying 36 probes in region: NRIP1\n",
      "--- assemblying 46 probes in region: PAXBP1\n",
      "--- assemblying 107 probes in region: PCBP3\n",
      "--- assemblying 30 probes in region: PCNT\n",
      "--- assemblying 107 probes in region: PDE9A\n",
      "--- assemblying 24 probes in region: PDXK\n",
      "--- assemblying 91 probes in region: PFKL\n",
      "--- assemblying 28 probes in region: PKNOX1\n",
      "--- assemblying 35 probes in region: PRDM15\n",
      "--- assemblying 34 probes in region: PRMT2\n",
      "--- assemblying 35 probes in region: PSMG1\n",
      "--- assemblying 22 probes in region: PTTG1IP\n",
      "--- assemblying 73 probes in region: PWP2\n",
      "--- assemblying 32 probes in region: RCAN1\n",
      "--- assemblying 32 probes in region: RRP1\n",
      "--- assemblying 62 probes in region: RRP1B\n",
      "--- assemblying 72 probes in region: RUNX1\n",
      "--- assemblying 40 probes in region: SCAF4\n",
      "--- assemblying 25 probes in region: SETD4\n",
      "--- assemblying 53 probes in region: SIK1\n",
      "--- assemblying 107 probes in region: SIM2\n",
      "--- assemblying 41 probes in region: SLC19A1\n",
      "--- assemblying 95 probes in region: SLC37A1\n",
      "--- assemblying 91 probes in region: SOD1\n",
      "--- assemblying 56 probes in region: SON\n",
      "--- assemblying 36 probes in region: SUMO3\n",
      "--- assemblying 32 probes in region: SYNJ1\n",
      "--- assemblying 30 probes in region: TIAM1\n",
      "--- assemblying 37 probes in region: TRAPPC10\n",
      "--- assemblying 41 probes in region: TTC3\n",
      "--- assemblying 100 probes in region: UBE2G2\n",
      "--- assemblying 87 probes in region: URB1\n",
      "--- assemblying 19 probes in region: USP16\n",
      "--- assemblying 31 probes in region: USP25\n",
      "--- assemblying 18 probes in region: WDR4\n",
      "--- assemblying 22 probes in region: WRB\n",
      "-- 4253 probes assembled in total.\n",
      "-- saving 4253 probes into file:X:\\Libraries\\CTP-06\\chr21_promoters\\candidate_probes.fasta\n",
      "-- saving readout_summary into file:X:\\Libraries\\CTP-06\\chr21_promoters\\readout_summary.pkl\n"
     ]
    }
   ],
   "source": [
    "reload(library_tools)\n",
    "from ImageAnalysis3.library_tools import _assemble_single_probe, _assemble_single_probename\n",
    "cand_probes, readout_summary = Assemble_probes(library_folder, pb_dict, gene_readout_dict, readout_dict, primers, \n",
    "                                               rc_targets=True, overwrite=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Check quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 405,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:58:00.718683Z",
     "start_time": "2019-06-14T01:58:00.712678Z"
    }
   },
   "outputs": [],
   "source": [
    "# biopython for SeqRecord\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "# blast\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:58:03.059425Z",
     "start_time": "2019-06-14T01:58:02.965354Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total probe loaded: 4253\n"
     ]
    }
   ],
   "source": [
    "# folder for this library\n",
    "# candidate full-length probe filename\n",
    "candidate_full_name = 'candidate_probes.fasta'\n",
    "# load full probes\n",
    "full_records = []\n",
    "with open(os.path.join(library_folder, candidate_full_name), 'r') as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        full_records.append(record)\n",
    "\n",
    "print(f\"Total probe loaded: {len(full_records)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:58:12.547490Z",
     "start_time": "2019-06-14T01:58:05.211027Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Checking primer usage, total probes: 4253\n",
      "gene: ADAMTS1 -> 112\n",
      "gene: ADAMTS5 -> 32\n",
      "gene: ADARB1 -> 36\n",
      "gene: AGPAT3 -> 21\n",
      "gene: ANKRD20A11P -> 62\n",
      "gene: APP -> 25\n",
      "gene: ATP5PF -> 32\n",
      "gene: BACE2 -> 35\n",
      "gene: BACH1 -> 82\n",
      "gene: BRWD1 -> 26\n",
      "gene: BTG3 -> 52\n",
      "gene: C21orf91 -> 33\n",
      "gene: C2CD2 -> 96\n",
      "gene: CBR3 -> 16\n",
      "gene: CCT8 -> 28\n",
      "gene: CHAF1B -> 29\n",
      "gene: COL18A1 -> 109\n",
      "gene: COL6A1 -> 32\n",
      "gene: COL6A2 -> 105\n",
      "gene: CRYZL1 -> 29\n",
      "gene: CXADR -> 34\n",
      "gene: CYP4F29P -> 56\n",
      "gene: DIP2A -> 45\n",
      "gene: DONSON -> 41\n",
      "gene: DSCR9 -> 78\n",
      "gene: DYRK1A -> 31\n",
      "gene: ETS2 -> 109\n",
      "gene: EVA1C -> 20\n",
      "gene: FAM207A -> 94\n",
      "gene: HLCS -> 39\n",
      "gene: HMGN1 -> 96\n",
      "gene: HSPA13 -> 29\n",
      "gene: HUNK -> 34\n",
      "gene: IFNAR1 -> 32\n",
      "gene: IFNAR2 -> 21\n",
      "gene: IL10RB -> 29\n",
      "gene: ITGB2 -> 35\n",
      "gene: JAM2 -> 30\n",
      "gene: LCA5L -> 94\n",
      "gene: LSS -> 35\n",
      "gene: LTN1 -> 37\n",
      "gene: MCM3AP -> 20\n",
      "gene: MIS18A -> 27\n",
      "gene: MORC3 -> 15\n",
      "gene: MRPL39 -> 79\n",
      "gene: MRPS6 -> 34\n",
      "gene: MX1 -> 30\n",
      "gene: MX2 -> 79\n",
      "gene: N6AMT1 -> 28\n",
      "gene: NCAM2 -> 103\n",
      "gene: NRIP1 -> 36\n",
      "gene: PAXBP1 -> 46\n",
      "gene: PCBP3 -> 107\n",
      "gene: PCNT -> 30\n",
      "gene: PDE9A -> 107\n",
      "gene: PDXK -> 24\n",
      "gene: PFKL -> 91\n",
      "gene: PKNOX1 -> 28\n",
      "gene: PRDM15 -> 35\n",
      "gene: PRMT2 -> 34\n",
      "gene: PSMG1 -> 35\n",
      "gene: PTTG1IP -> 22\n",
      "gene: PWP2 -> 73\n",
      "gene: RCAN1 -> 32\n",
      "gene: RRP1 -> 32\n",
      "gene: RRP1B -> 62\n",
      "gene: RUNX1 -> 72\n",
      "gene: SCAF4 -> 40\n",
      "gene: SETD4 -> 25\n",
      "gene: SIK1 -> 53\n",
      "gene: SIM2 -> 107\n",
      "gene: SLC19A1 -> 41\n",
      "gene: SLC37A1 -> 95\n",
      "gene: SOD1 -> 91\n",
      "gene: SON -> 56\n",
      "gene: SUMO3 -> 36\n",
      "gene: SYNJ1 -> 32\n",
      "gene: TIAM1 -> 30\n",
      "gene: TRAPPC10 -> 37\n",
      "gene: TTC3 -> 41\n",
      "gene: UBE2G2 -> 100\n",
      "gene: URB1 -> 87\n",
      "gene: USP16 -> 19\n",
      "gene: USP25 -> 31\n",
      "gene: WDR4 -> 18\n",
      "gene: WRB -> 22\n",
      "ADAMTS1 -> ['NDB_1147', 'NDB_1147', 'NDB_1147']\n",
      "ADAMTS5 -> ['NDB_1130', 'NDB_1130', 'NDB_1130']\n",
      "ADARB1 -> ['NDB_1157', 'NDB_1157', 'NDB_1157']\n",
      "AGPAT3 -> ['NDB_1176', 'NDB_1176', 'NDB_1176']\n",
      "ANKRD20A11P -> ['NDB_1209', 'NDB_1209', 'NDB_1209']\n",
      "APP -> ['NDB_1146', 'NDB_1146', 'NDB_1146']\n",
      "ATP5PF -> ['NDB_1136', 'NDB_1136', 'NDB_1136']\n",
      "BACE2 -> ['NDB_1137', 'NDB_1137', 'NDB_1137']\n",
      "BACH1 -> ['NDB_1175', 'NDB_1175', 'NDB_1175']\n",
      "BRWD1 -> ['NDB_1170', 'NDB_1170', 'NDB_1170']\n",
      "BTG3 -> ['NDB_1163', 'NDB_1163', 'NDB_1163']\n",
      "C21orf91 -> ['NDB_1132', 'NDB_1132', 'NDB_1132']\n",
      "C2CD2 -> ['NDB_1131', 'NDB_1131', 'NDB_1131']\n",
      "CBR3 -> ['NDB_1126', 'NDB_1126', 'NDB_1126']\n",
      "CCT8 -> ['NDB_1127', 'NDB_1127', 'NDB_1127']\n",
      "CHAF1B -> ['NDB_1174', 'NDB_1174', 'NDB_1174']\n",
      "COL18A1 -> ['NDB_1166', 'NDB_1166', 'NDB_1166']\n",
      "COL6A1 -> ['NDB_1208', 'NDB_1208', 'NDB_1208']\n",
      "COL6A2 -> ['NDB_1140', 'NDB_1140', 'NDB_1140']\n",
      "CRYZL1 -> ['NDB_1134', 'NDB_1134', 'NDB_1134']\n",
      "CXADR -> ['NDB_1203', 'NDB_1203', 'NDB_1203']\n",
      "CYP4F29P -> ['NDB_1199', 'NDB_1199', 'NDB_1199']\n",
      "DIP2A -> ['NDB_1193', 'NDB_1193', 'NDB_1193']\n",
      "DONSON -> ['NDB_1158', 'NDB_1158', 'NDB_1158']\n",
      "DSCR9 -> ['NDB_1168', 'NDB_1168', 'NDB_1168']\n",
      "DYRK1A -> ['NDB_1207', 'NDB_1207', 'NDB_1207']\n",
      "ETS2 -> ['NDB_1153', 'NDB_1153', 'NDB_1153']\n",
      "EVA1C -> ['NDB_1202', 'NDB_1202', 'NDB_1202']\n",
      "FAM207A -> ['NDB_1177', 'NDB_1177', 'NDB_1177']\n",
      "HLCS -> ['NDB_1160', 'NDB_1160', 'NDB_1160']\n",
      "HMGN1 -> ['NDB_1161', 'NDB_1161', 'NDB_1161']\n",
      "HSPA13 -> ['NDB_1159', 'NDB_1159', 'NDB_1159']\n",
      "HUNK -> ['NDB_1200', 'NDB_1200', 'NDB_1200']\n",
      "IFNAR1 -> ['NDB_1149', 'NDB_1149', 'NDB_1149']\n",
      "IFNAR2 -> ['NDB_1145', 'NDB_1145', 'NDB_1145']\n",
      "IL10RB -> ['NDB_1171', 'NDB_1171', 'NDB_1171']\n",
      "ITGB2 -> ['NDB_1189', 'NDB_1189', 'NDB_1189']\n",
      "JAM2 -> ['NDB_1192', 'NDB_1192', 'NDB_1192']\n",
      "LCA5L -> ['NDB_1194', 'NDB_1194', 'NDB_1194']\n",
      "LSS -> ['NDB_1139', 'NDB_1139', 'NDB_1139']\n",
      "LTN1 -> ['NDB_1169', 'NDB_1169', 'NDB_1169']\n",
      "MCM3AP -> ['NDB_1141', 'NDB_1141', 'NDB_1141']\n",
      "MIS18A -> ['NDB_1196', 'NDB_1196', 'NDB_1196']\n",
      "MORC3 -> ['NDB_1155', 'NDB_1155', 'NDB_1155']\n",
      "MRPL39 -> ['NDB_1205', 'NDB_1205', 'NDB_1205']\n",
      "MRPS6 -> ['NDB_1150', 'NDB_1150', 'NDB_1150']\n",
      "MX1 -> ['NDB_1151', 'NDB_1151', 'NDB_1151']\n",
      "MX2 -> ['NDB_1152', 'NDB_1152', 'NDB_1152']\n",
      "N6AMT1 -> ['NDB_1186', 'NDB_1186', 'NDB_1186']\n",
      "NCAM2 -> ['NDB_1135', 'NDB_1135', 'NDB_1135']\n",
      "NRIP1 -> ['NDB_1156', 'NDB_1156', 'NDB_1156']\n",
      "PAXBP1 -> ['NDB_1211', 'NDB_1211', 'NDB_1211']\n",
      "PCBP3 -> ['NDB_1164', 'NDB_1164', 'NDB_1164']\n",
      "PCNT -> ['NDB_1133', 'NDB_1133', 'NDB_1133']\n",
      "PDE9A -> ['NDB_1173', 'NDB_1173', 'NDB_1173']\n",
      "PDXK -> ['NDB_1172', 'NDB_1172', 'NDB_1172']\n",
      "PFKL -> ['NDB_1184', 'NDB_1184', 'NDB_1184']\n",
      "PKNOX1 -> ['NDB_1138', 'NDB_1138', 'NDB_1138']\n",
      "PRDM15 -> ['NDB_1198', 'NDB_1198', 'NDB_1198']\n",
      "PRMT2 -> ['NDB_1180', 'NDB_1180', 'NDB_1180']\n",
      "PSMG1 -> ['NDB_1210', 'NDB_1210', 'NDB_1210']\n",
      "PTTG1IP -> ['NDB_1183', 'NDB_1183', 'NDB_1183']\n",
      "PWP2 -> ['NDB_1195', 'NDB_1195', 'NDB_1195']\n",
      "RCAN1 -> ['NDB_1191', 'NDB_1191', 'NDB_1191']\n",
      "RRP1 -> ['NDB_1181', 'NDB_1181', 'NDB_1181']\n",
      "RRP1B -> ['NDB_1190', 'NDB_1190', 'NDB_1190']\n",
      "RUNX1 -> ['NDB_1129', 'NDB_1129', 'NDB_1129']\n",
      "SCAF4 -> ['NDB_1204', 'NDB_1204', 'NDB_1204']\n",
      "SETD4 -> ['NDB_1188', 'NDB_1188', 'NDB_1188']\n",
      "SIK1 -> ['NDB_1182', 'NDB_1182', 'NDB_1182']\n",
      "SIM2 -> ['NDB_1206', 'NDB_1206', 'NDB_1206']\n",
      "SLC19A1 -> ['NDB_1201', 'NDB_1201', 'NDB_1201']\n",
      "SLC37A1 -> ['NDB_1187', 'NDB_1187', 'NDB_1187']\n",
      "SOD1 -> ['NDB_1162', 'NDB_1162', 'NDB_1162']\n",
      "SON -> ['NDB_1154', 'NDB_1154', 'NDB_1154']\n",
      "SUMO3 -> ['NDB_1128', 'NDB_1128', 'NDB_1128']\n",
      "SYNJ1 -> ['NDB_1142', 'NDB_1142', 'NDB_1142']\n",
      "TIAM1 -> ['NDB_1197', 'NDB_1197', 'NDB_1197']\n",
      "TRAPPC10 -> ['NDB_1178', 'NDB_1178', 'NDB_1178']\n",
      "TTC3 -> ['NDB_1143', 'NDB_1143', 'NDB_1143']\n",
      "UBE2G2 -> ['NDB_1144', 'NDB_1144', 'NDB_1144']\n",
      "URB1 -> ['NDB_1165', 'NDB_1165', 'NDB_1165']\n",
      "USP16 -> ['NDB_1167', 'NDB_1167', 'NDB_1167']\n",
      "USP25 -> ['NDB_1179', 'NDB_1179', 'NDB_1179']\n",
      "WDR4 -> ['NDB_1148', 'NDB_1148', 'NDB_1148']\n",
      "WRB -> ['NDB_1185', 'NDB_1185', 'NDB_1185']\n",
      "NDB_1126 -> ['CBR3', 'CBR3', 'CBR3']\n",
      "NDB_1127 -> ['CCT8', 'CCT8', 'CCT8']\n",
      "NDB_1128 -> ['SUMO3', 'SUMO3', 'SUMO3']\n",
      "NDB_1129 -> ['RUNX1', 'RUNX1', 'RUNX1']\n",
      "NDB_1130 -> ['ADAMTS5', 'ADAMTS5', 'ADAMTS5']\n",
      "NDB_1131 -> ['C2CD2', 'C2CD2', 'C2CD2']\n",
      "NDB_1132 -> ['C21orf91', 'C21orf91', 'C21orf91']\n",
      "NDB_1133 -> ['PCNT', 'PCNT', 'PCNT']\n",
      "NDB_1134 -> ['CRYZL1', 'CRYZL1', 'CRYZL1']\n",
      "NDB_1135 -> ['NCAM2', 'NCAM2', 'NCAM2']\n",
      "NDB_1136 -> ['ATP5PF', 'ATP5PF', 'ATP5PF']\n",
      "NDB_1137 -> ['BACE2', 'BACE2', 'BACE2']\n",
      "NDB_1138 -> ['PKNOX1', 'PKNOX1', 'PKNOX1']\n",
      "NDB_1139 -> ['LSS', 'LSS', 'LSS']\n",
      "NDB_1140 -> ['COL6A2', 'COL6A2', 'COL6A2']\n",
      "NDB_1141 -> ['MCM3AP', 'MCM3AP', 'MCM3AP']\n",
      "NDB_1142 -> ['SYNJ1', 'SYNJ1', 'SYNJ1']\n",
      "NDB_1143 -> ['TTC3', 'TTC3', 'TTC3']\n",
      "NDB_1144 -> ['UBE2G2', 'UBE2G2', 'UBE2G2']\n",
      "NDB_1145 -> ['IFNAR2', 'IFNAR2', 'IFNAR2']\n",
      "NDB_1146 -> ['APP', 'APP', 'APP']\n",
      "NDB_1147 -> ['ADAMTS1', 'ADAMTS1', 'ADAMTS1']\n",
      "NDB_1148 -> ['WDR4', 'WDR4', 'WDR4']\n",
      "NDB_1149 -> ['IFNAR1', 'IFNAR1', 'IFNAR1']\n",
      "NDB_1150 -> ['MRPS6', 'MRPS6', 'MRPS6']\n",
      "NDB_1151 -> ['MX1', 'MX1', 'MX1']\n",
      "NDB_1152 -> ['MX2', 'MX2', 'MX2']\n",
      "NDB_1153 -> ['ETS2', 'ETS2', 'ETS2']\n",
      "NDB_1154 -> ['SON', 'SON', 'SON']\n",
      "NDB_1155 -> ['MORC3', 'MORC3', 'MORC3']\n",
      "NDB_1156 -> ['NRIP1', 'NRIP1', 'NRIP1']\n",
      "NDB_1157 -> ['ADARB1', 'ADARB1', 'ADARB1']\n",
      "NDB_1158 -> ['DONSON', 'DONSON', 'DONSON']\n",
      "NDB_1159 -> ['HSPA13', 'HSPA13', 'HSPA13']\n",
      "NDB_1160 -> ['HLCS', 'HLCS', 'HLCS']\n",
      "NDB_1161 -> ['HMGN1', 'HMGN1', 'HMGN1']\n",
      "NDB_1162 -> ['SOD1', 'SOD1', 'SOD1']\n",
      "NDB_1163 -> ['BTG3', 'BTG3', 'BTG3']\n",
      "NDB_1164 -> ['PCBP3', 'PCBP3', 'PCBP3']\n",
      "NDB_1165 -> ['URB1', 'URB1', 'URB1']\n",
      "NDB_1166 -> ['COL18A1', 'COL18A1', 'COL18A1']\n",
      "NDB_1167 -> ['USP16', 'USP16', 'USP16']\n",
      "NDB_1168 -> ['DSCR9', 'DSCR9', 'DSCR9']\n",
      "NDB_1169 -> ['LTN1', 'LTN1', 'LTN1']\n",
      "NDB_1170 -> ['BRWD1', 'BRWD1', 'BRWD1']\n",
      "NDB_1171 -> ['IL10RB', 'IL10RB', 'IL10RB']\n",
      "NDB_1172 -> ['PDXK', 'PDXK', 'PDXK']\n",
      "NDB_1173 -> ['PDE9A', 'PDE9A', 'PDE9A']\n",
      "NDB_1174 -> ['CHAF1B', 'CHAF1B', 'CHAF1B']\n",
      "NDB_1175 -> ['BACH1', 'BACH1', 'BACH1']\n",
      "NDB_1176 -> ['AGPAT3', 'AGPAT3', 'AGPAT3']\n",
      "NDB_1177 -> ['FAM207A', 'FAM207A', 'FAM207A']\n",
      "NDB_1178 -> ['TRAPPC10', 'TRAPPC10', 'TRAPPC10']\n",
      "NDB_1179 -> ['USP25', 'USP25', 'USP25']\n",
      "NDB_1180 -> ['PRMT2', 'PRMT2', 'PRMT2']\n",
      "NDB_1181 -> ['RRP1', 'RRP1', 'RRP1']\n",
      "NDB_1182 -> ['SIK1', 'SIK1', 'SIK1']\n",
      "NDB_1183 -> ['PTTG1IP', 'PTTG1IP', 'PTTG1IP']\n",
      "NDB_1184 -> ['PFKL', 'PFKL', 'PFKL']\n",
      "NDB_1185 -> ['WRB', 'WRB', 'WRB']\n",
      "NDB_1186 -> ['N6AMT1', 'N6AMT1', 'N6AMT1']\n",
      "NDB_1187 -> ['SLC37A1', 'SLC37A1', 'SLC37A1']\n",
      "NDB_1188 -> ['SETD4', 'SETD4', 'SETD4']\n",
      "NDB_1189 -> ['ITGB2', 'ITGB2', 'ITGB2']\n",
      "NDB_1190 -> ['RRP1B', 'RRP1B', 'RRP1B']\n",
      "NDB_1191 -> ['RCAN1', 'RCAN1', 'RCAN1']\n",
      "NDB_1192 -> ['JAM2', 'JAM2', 'JAM2']\n",
      "NDB_1193 -> ['DIP2A', 'DIP2A', 'DIP2A']\n",
      "NDB_1194 -> ['LCA5L', 'LCA5L', 'LCA5L']\n",
      "NDB_1195 -> ['PWP2', 'PWP2', 'PWP2']\n",
      "NDB_1196 -> ['MIS18A', 'MIS18A', 'MIS18A']\n",
      "NDB_1197 -> ['TIAM1', 'TIAM1', 'TIAM1']\n",
      "NDB_1198 -> ['PRDM15', 'PRDM15', 'PRDM15']\n",
      "NDB_1199 -> ['CYP4F29P', 'CYP4F29P', 'CYP4F29P']\n",
      "NDB_1200 -> ['HUNK', 'HUNK', 'HUNK']\n",
      "NDB_1201 -> ['SLC19A1', 'SLC19A1', 'SLC19A1']\n",
      "NDB_1202 -> ['EVA1C', 'EVA1C', 'EVA1C']\n",
      "NDB_1203 -> ['CXADR', 'CXADR', 'CXADR']\n",
      "NDB_1204 -> ['SCAF4', 'SCAF4', 'SCAF4']\n",
      "NDB_1205 -> ['MRPL39', 'MRPL39', 'MRPL39']\n",
      "NDB_1206 -> ['SIM2', 'SIM2', 'SIM2']\n",
      "NDB_1207 -> ['DYRK1A', 'DYRK1A', 'DYRK1A']\n",
      "NDB_1208 -> ['COL6A1', 'COL6A1', 'COL6A1']\n",
      "NDB_1209 -> ['ANKRD20A11P', 'ANKRD20A11P', 'ANKRD20A11P']\n",
      "NDB_1210 -> ['PSMG1', 'PSMG1', 'PSMG1']\n",
      "NDB_1211 -> ['PAXBP1', 'PAXBP1', 'PAXBP1']\n"
     ]
    }
   ],
   "source": [
    "reload(check)\n",
    "\n",
    "primer_check = check._check_primer_usage(full_records, primers[0], primers[1])\n",
    "\n",
    "reg_size_dic, len_check = check._check_region_size(full_records, min_size=15)\n",
    "\n",
    "reg_readout_dic, reg2readout_check = check._check_region_to_readouts(full_records, readout_dict)\n",
    "\n",
    "readout_reg_dic, readout2reg_check = check._check_readout_to_region(reg_readout_dic, full_records, readout_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T01:58:23.865921Z",
     "start_time": "2019-06-14T01:58:14.957286Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- saving internal 17-mer map to file:X:\\Libraries\\CTP-06\\chr21_promoters\\probe_table_17.npz\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_4_pos_1441_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:52, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_5_pos_1501_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:62, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_9_pos_1749_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:87, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_10_pos_1791_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:92, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_11_pos_1833_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:59, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_13_pos_1930_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:59, dumped\n",
      "--- Sequence:chr21:33321943-33326943_gene_IFNAR1_pb_14_pos_1998_readouts_[NDB_1149_u,NDB_1149_u,NDB_1149_u] got hits:75, dumped\n",
      "-- total probes removed by internal screening: 7\n"
     ]
    }
   ],
   "source": [
    "reload(check)\n",
    "reload(ld)\n",
    "int_map = check._construct_internal_map(full_records, library_folder)\n",
    "\n",
    "readout_count_dic, readout_count_check = check._check_readout_in_probes(readout_reg_dic, reg_size_dic, int_map, readout_dict)\n",
    "\n",
    "kept_records, removed_count = check._check_between_probes(full_records, int_map)\n",
    "# save kept records\n",
    "with open(os.path.join(library_folder, 'filtered_full_probes.fasta'), 'w') as output_handle:\n",
    "    SeqIO.write(kept_records, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-17T01:08:42.503545Z",
     "start_time": "2019-06-17T01:08:42.481528Z"
    }
   },
   "outputs": [],
   "source": [
    "reload(ImageAnalysis3)\n",
    "from ImageAnalysis3.library_tools import quality_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-17T01:07:44.291187Z",
     "start_time": "2019-06-17T01:07:44.273175Z"
    }
   },
   "outputs": [],
   "source": [
    "quality_check.Blast_probes(kept_records, library_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T02:31:53.374630Z",
     "start_time": "2019-06-14T02:23:50.071656Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Number of probes loaded: 4246\n",
      "- Number of regions in this library: 86\n",
      "-- checking probes in region: ADAMTS1\n",
      "--- number of probes: 112 , kept by blast: 112 , if remove dups: 112\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: ADAMTS5\n",
      "--- gene=ADAMTS5, id=28 removed by soft count = 13627\n",
      "--- number of probes: 32 , kept by blast: 31 , if remove dups: 31\n",
      "-- number of probes kept for this region: 31\n",
      "-- checking probes in region: ADARB1\n",
      "--- gene=ADARB1, id=16 removed by soft count = 1018\n",
      "--- gene=ADARB1, id=24 removed by soft count = 80\n",
      "--- number of probes: 36 , kept by blast: 34 , if remove dups: 34\n",
      "-- number of probes kept for this region: 34\n",
      "-- checking probes in region: AGPAT3\n",
      "--- gene=AGPAT3, id=8 removed by hard count = 0\n",
      "--- number of probes: 21 , kept by blast: 20 , if remove dups: 20\n",
      "-- number of probes kept for this region: 20\n",
      "-- checking probes in region: ANKRD20A11P\n",
      "--- gene=ANKRD20A11P, id=1 removed by soft count = 1695\n",
      "--- gene=ANKRD20A11P, id=6 removed by soft count = 352\n",
      "--- gene=ANKRD20A11P, id=10 removed by hard count = 3\n",
      "--- gene=ANKRD20A11P, id=12 removed by hard count = 3\n",
      "--- gene=ANKRD20A11P, id=19 removed by soft count = 63\n",
      "--- gene=ANKRD20A11P, id=21 removed by hard count = 5\n",
      "--- gene=ANKRD20A11P, id=30 removed by soft count = 33\n",
      "--- gene=ANKRD20A11P, id=34 removed by hard count = 4\n",
      "--- gene=ANKRD20A11P, id=45 removed by soft count = 42\n",
      "--- gene=ANKRD20A11P, id=46 removed by soft count = 41\n",
      "--- gene=ANKRD20A11P, id=47 removed by hard count = 3\n",
      "--- gene=ANKRD20A11P, id=51 removed by soft count = 37\n",
      "--- gene=ANKRD20A11P, id=52 removed by hard count = 3\n",
      "--- gene=ANKRD20A11P, id=56 removed by soft count = 35\n",
      "--- gene=ANKRD20A11P, id=57 removed by soft count = 120\n",
      "--- gene=ANKRD20A11P, id=58 removed by soft count = 192\n",
      "--- gene=ANKRD20A11P, id=59 removed by hard count = 3\n",
      "--- gene=ANKRD20A11P, id=60 removed by soft count = 1593\n",
      "--- gene=ANKRD20A11P, id=61 removed by soft count = 47\n",
      "--- number of probes: 62 , kept by blast: 43 , if remove dups: 23\n",
      "-- number of probes kept for this region: 43\n",
      "-- checking probes in region: APP\n",
      "--- gene=APP, id=22 removed by soft count = 255\n",
      "--- number of probes: 25 , kept by blast: 24 , if remove dups: 24\n",
      "-- number of probes kept for this region: 24\n",
      "-- checking probes in region: ATP5PF\n",
      "--- gene=ATP5PF, id=1 removed by soft count = 2498\n",
      "--- gene=ATP5PF, id=7 removed by hard count = 0\n",
      "--- gene=ATP5PF, id=30 removed by soft count = 844\n",
      "--- number of probes: 32 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: BACE2\n",
      "--- number of probes: 35 , kept by blast: 35 , if remove dups: 35\n",
      "-- number of probes kept for this region: 35\n",
      "-- checking probes in region: BACH1\n",
      "--- gene=BACH1, id=21 removed by soft count = 91\n",
      "--- gene=BACH1, id=22 removed by soft count = 49\n",
      "--- gene=BACH1, id=23 removed by soft count = 48\n",
      "--- gene=BACH1, id=25 removed by soft count = 79\n",
      "--- gene=BACH1, id=26 removed by soft count = 73\n",
      "--- gene=BACH1, id=27 removed by soft count = 49\n",
      "--- gene=BACH1, id=28 removed by soft count = 575\n",
      "--- gene=BACH1, id=29 removed by soft count = 81\n",
      "--- gene=BACH1, id=39 removed by soft count = 31\n",
      "--- gene=BACH1, id=50 removed by soft count = 34\n",
      "--- gene=BACH1, id=51 removed by soft count = 39\n",
      "--- gene=BACH1, id=53 removed by soft count = 53\n",
      "--- gene=BACH1, id=54 removed by soft count = 98\n",
      "--- gene=BACH1, id=55 removed by soft count = 132\n",
      "--- gene=BACH1, id=74 removed by soft count = 34\n",
      "--- number of probes: 82 , kept by blast: 67 , if remove dups: 67\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 67\n",
      "-- checking probes in region: BRWD1\n",
      "--- gene=BRWD1, id=11 removed by soft count = 320\n",
      "--- number of probes: 26 , kept by blast: 25 , if remove dups: 25\n",
      "-- number of probes kept for this region: 25\n",
      "-- checking probes in region: BTG3\n",
      "--- gene=BTG3, id=47 removed by soft count = 104\n",
      "--- gene=BTG3, id=51 removed by soft count = 205\n",
      "--- number of probes: 52 , kept by blast: 50 , if remove dups: 0\n",
      "-- number of probes kept for this region: 50\n",
      "-- checking probes in region: C21orf91\n",
      "--- gene=C21orf91, id=2 removed by soft count = 157\n",
      "--- gene=C21orf91, id=27 removed by soft count = 138\n",
      "--- gene=C21orf91, id=29 removed by soft count = 199\n",
      "--- gene=C21orf91, id=31 removed by soft count = 431\n",
      "--- gene=C21orf91, id=32 removed by soft count = 1403\n",
      "--- number of probes: 33 , kept by blast: 28 , if remove dups: 28\n",
      "-- number of probes kept for this region: 28\n",
      "-- checking probes in region: C2CD2\n",
      "--- gene=C2CD2, id=28 removed by soft count = 68\n",
      "--- gene=C2CD2, id=66 removed by hard count = 0\n",
      "--- gene=C2CD2, id=67 removed by hard count = 0\n",
      "--- gene=C2CD2, id=68 removed by hard count = 0\n",
      "--- gene=C2CD2, id=71 removed by hard count = 0\n",
      "--- gene=C2CD2, id=72 removed by soft count = 132\n",
      "--- gene=C2CD2, id=73 removed by hard count = 0\n",
      "--- gene=C2CD2, id=74 removed by soft count = 36\n",
      "--- gene=C2CD2, id=75 removed by hard count = 0\n",
      "--- gene=C2CD2, id=76 removed by hard count = 0\n",
      "--- number of probes: 96 , kept by blast: 86 , if remove dups: 85\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: CBR3\n",
      "--- number of probes: 16 , kept by blast: 16 , if remove dups: 16\n",
      "-- number of probes kept for this region: 16\n",
      "-- checking probes in region: CCT8\n",
      "--- gene=CCT8, id=20 removed by soft count = 34\n",
      "--- gene=CCT8, id=22 removed by soft count = 81\n",
      "--- gene=CCT8, id=24 removed by soft count = 98\n",
      "--- gene=CCT8, id=25 removed by soft count = 92\n",
      "--- gene=CCT8, id=26 removed by soft count = 165\n",
      "--- gene=CCT8, id=27 removed by soft count = 440\n",
      "--- number of probes: 28 , kept by blast: 22 , if remove dups: 22\n",
      "-- number of probes kept for this region: 22\n",
      "-- checking probes in region: CHAF1B\n",
      "--- gene=CHAF1B, id=0 removed by soft count = 3127\n",
      "--- gene=CHAF1B, id=1 removed by soft count = 31\n",
      "--- gene=CHAF1B, id=22 removed by soft count = 34\n",
      "--- number of probes: 29 , kept by blast: 26 , if remove dups: 26\n",
      "-- number of probes kept for this region: 26\n",
      "-- checking probes in region: COL18A1\n",
      "--- number of probes: 109 , kept by blast: 109 , if remove dups: 109\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: COL6A1\n",
      "--- number of probes: 32 , kept by blast: 32 , if remove dups: 32\n",
      "-- number of probes kept for this region: 32\n",
      "-- checking probes in region: COL6A2\n",
      "--- gene=COL6A2, id=44 removed by soft count = 59\n",
      "--- gene=COL6A2, id=65 removed by hard count = 0\n",
      "--- gene=COL6A2, id=66 removed by hard count = 0\n",
      "--- number of probes: 105 , kept by blast: 102 , if remove dups: 102\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: CRYZL1\n",
      "--- gene=CRYZL1, id=16 removed by soft count = 1844\n",
      "--- gene=CRYZL1, id=18 removed by soft count = 35\n",
      "--- gene=CRYZL1, id=19 removed by soft count = 80\n",
      "--- number of probes: 29 , kept by blast: 26 , if remove dups: 26\n",
      "-- number of probes kept for this region: 26\n",
      "-- checking probes in region: CXADR\n",
      "--- gene=CXADR, id=0 removed by soft count = 82\n",
      "--- gene=CXADR, id=4 removed by hard count = 0\n",
      "--- gene=CXADR, id=5 removed by soft count = 67\n",
      "--- gene=CXADR, id=8 removed by soft count = 1902\n",
      "--- number of probes: 34 , kept by blast: 30 , if remove dups: 30\n",
      "-- number of probes kept for this region: 30\n",
      "-- checking probes in region: CYP4F29P\n",
      "--- gene=CYP4F29P, id=2 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=5 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=6 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=13 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=14 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=16 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=23 removed by hard count = 5\n",
      "--- gene=CYP4F29P, id=24 removed by hard count = 5\n",
      "--- gene=CYP4F29P, id=26 removed by soft count = 52\n",
      "--- gene=CYP4F29P, id=32 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=34 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=36 removed by soft count = 40\n",
      "--- gene=CYP4F29P, id=43 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=44 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=46 removed by soft count = 54\n",
      "--- gene=CYP4F29P, id=47 removed by soft count = 48\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- gene=CYP4F29P, id=48 removed by soft count = 47\n",
      "--- gene=CYP4F29P, id=49 removed by hard count = 4\n",
      "--- gene=CYP4F29P, id=50 removed by soft count = 32\n",
      "--- gene=CYP4F29P, id=51 removed by soft count = 63\n",
      "--- gene=CYP4F29P, id=52 removed by soft count = 102\n",
      "--- gene=CYP4F29P, id=53 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=54 removed by hard count = 3\n",
      "--- gene=CYP4F29P, id=55 removed by soft count = 45\n",
      "--- number of probes: 56 , kept by blast: 32 , if remove dups: 21\n",
      "-- number of probes kept for this region: 32\n",
      "-- checking probes in region: DIP2A\n",
      "--- gene=DIP2A, id=0 removed by soft count = 95\n",
      "--- gene=DIP2A, id=8 removed by soft count = 72\n",
      "--- gene=DIP2A, id=21 removed by soft count = 61\n",
      "--- gene=DIP2A, id=38 removed by soft count = 35\n",
      "--- gene=DIP2A, id=44 removed by soft count = 86\n",
      "--- number of probes: 45 , kept by blast: 40 , if remove dups: 40\n",
      "-- number of probes kept for this region: 40\n",
      "-- checking probes in region: DONSON\n",
      "--- gene=DONSON, id=3 removed by soft count = 37868\n",
      "--- gene=DONSON, id=28 removed by soft count = 56\n",
      "--- gene=DONSON, id=30 removed by soft count = 18976\n",
      "--- number of probes: 41 , kept by blast: 38 , if remove dups: 38\n",
      "-- number of probes kept for this region: 38\n",
      "-- checking probes in region: DSCR9\n",
      "--- gene=DSCR9, id=10 removed by soft count = 3823\n",
      "--- gene=DSCR9, id=11 removed by soft count = 142\n",
      "--- gene=DSCR9, id=17 removed by soft count = 138\n",
      "--- gene=DSCR9, id=55 removed by soft count = 64\n",
      "--- gene=DSCR9, id=70 removed by soft count = 392\n",
      "--- gene=DSCR9, id=71 removed by soft count = 187\n",
      "--- gene=DSCR9, id=72 removed by soft count = 370\n",
      "--- gene=DSCR9, id=73 removed by soft count = 77\n",
      "--- gene=DSCR9, id=74 removed by soft count = 91\n",
      "--- gene=DSCR9, id=75 removed by soft count = 86\n",
      "--- number of probes: 78 , kept by blast: 68 , if remove dups: 68\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 68\n",
      "-- checking probes in region: DYRK1A\n",
      "--- gene=DYRK1A, id=0 removed by soft count = 118\n",
      "--- gene=DYRK1A, id=1 removed by soft count = 76\n",
      "--- number of probes: 31 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: ETS2\n",
      "--- gene=ETS2, id=65 removed by soft count = 46\n",
      "--- gene=ETS2, id=66 removed by soft count = 53\n",
      "--- gene=ETS2, id=74 removed by hard count = 0\n",
      "--- gene=ETS2, id=89 removed by soft count = 87\n",
      "--- number of probes: 109 , kept by blast: 105 , if remove dups: 105\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: EVA1C\n",
      "--- gene=EVA1C, id=2 removed by soft count = 265\n",
      "--- gene=EVA1C, id=6 removed by soft count = 39\n",
      "--- number of probes: 20 , kept by blast: 18 , if remove dups: 18\n",
      "-- number of probes kept for this region: 18\n",
      "-- checking probes in region: FAM207A\n",
      "--- gene=FAM207A, id=1 removed by soft count = 265\n",
      "--- gene=FAM207A, id=2 removed by soft count = 157\n",
      "--- gene=FAM207A, id=4 removed by soft count = 595\n",
      "--- gene=FAM207A, id=5 removed by soft count = 1360\n",
      "--- gene=FAM207A, id=30 removed by soft count = 337\n",
      "--- gene=FAM207A, id=31 removed by soft count = 406\n",
      "--- gene=FAM207A, id=32 removed by soft count = 13055\n",
      "--- gene=FAM207A, id=58 removed by soft count = 34\n",
      "--- gene=FAM207A, id=59 removed by soft count = 81\n",
      "--- number of probes: 94 , kept by blast: 85 , if remove dups: 85\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: HLCS\n",
      "--- gene=HLCS, id=8 removed by soft count = 45\n",
      "--- gene=HLCS, id=10 removed by soft count = 44266\n",
      "--- gene=HLCS, id=19 removed by soft count = 270\n",
      "--- gene=HLCS, id=20 removed by soft count = 81\n",
      "--- gene=HLCS, id=21 removed by soft count = 145\n",
      "--- gene=HLCS, id=22 removed by soft count = 51\n",
      "--- gene=HLCS, id=23 removed by soft count = 315\n",
      "--- gene=HLCS, id=26 removed by soft count = 265\n",
      "--- gene=HLCS, id=28 removed by soft count = 139\n",
      "--- gene=HLCS, id=29 removed by soft count = 106\n",
      "--- gene=HLCS, id=30 removed by soft count = 120\n",
      "--- gene=HLCS, id=32 removed by soft count = 223\n",
      "--- gene=HLCS, id=33 removed by soft count = 369\n",
      "--- gene=HLCS, id=34 removed by soft count = 725\n",
      "--- number of probes: 39 , kept by blast: 25 , if remove dups: 25\n",
      "-- number of probes kept for this region: 25\n",
      "-- checking probes in region: HMGN1\n",
      "--- gene=HMGN1, id=54 removed by soft count = 1853\n",
      "--- gene=HMGN1, id=64 removed by soft count = 32\n",
      "--- gene=HMGN1, id=76 removed by soft count = 106\n",
      "--- gene=HMGN1, id=79 removed by soft count = 84\n",
      "--- gene=HMGN1, id=87 removed by soft count = 60\n",
      "--- gene=HMGN1, id=90 removed by soft count = 34\n",
      "--- number of probes: 96 , kept by blast: 90 , if remove dups: 87\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: HSPA13\n",
      "--- gene=HSPA13, id=21 removed by soft count = 56\n",
      "--- gene=HSPA13, id=23 removed by soft count = 38\n",
      "--- gene=HSPA13, id=24 removed by soft count = 84\n",
      "--- gene=HSPA13, id=25 removed by soft count = 42\n",
      "--- number of probes: 29 , kept by blast: 25 , if remove dups: 25\n",
      "-- number of probes kept for this region: 25\n",
      "-- checking probes in region: HUNK\n",
      "--- gene=HUNK, id=3 removed by soft count = 43\n",
      "--- number of probes: 34 , kept by blast: 33 , if remove dups: 33\n",
      "-- number of probes kept for this region: 33\n",
      "-- checking probes in region: IFNAR1\n",
      "--- gene=IFNAR1, id=1 removed by soft count = 96\n",
      "--- gene=IFNAR1, id=2 removed by hard count = 0\n",
      "--- gene=IFNAR1, id=3 removed by soft count = 293\n",
      "--- gene=IFNAR1, id=4 removed by soft count = 120\n",
      "--- gene=IFNAR1, id=5 removed by soft count = 132\n",
      "--- gene=IFNAR1, id=6 removed by soft count = 123\n",
      "--- gene=IFNAR1, id=7 removed by hard count = 3\n",
      "--- gene=IFNAR1, id=18 removed by soft count = 75\n",
      "--- gene=IFNAR1, id=20 removed by soft count = 351\n",
      "--- gene=IFNAR1, id=21 removed by soft count = 38\n",
      "--- gene=IFNAR1, id=22 removed by soft count = 45\n",
      "--- number of probes: 25 , kept by blast: 14 , if remove dups: 14\n",
      "-- number of probes kept for this region: 14\n",
      "-- checking probes in region: IFNAR2\n",
      "--- gene=IFNAR2, id=4 removed by soft count = 41\n",
      "--- number of probes: 21 , kept by blast: 20 , if remove dups: 20\n",
      "-- number of probes kept for this region: 20\n",
      "-- checking probes in region: IL10RB\n",
      "--- number of probes: 29 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: ITGB2\n",
      "--- gene=ITGB2, id=1 removed by soft count = 535\n",
      "--- gene=ITGB2, id=20 removed by hard count = 0\n",
      "--- number of probes: 35 , kept by blast: 33 , if remove dups: 33\n",
      "-- number of probes kept for this region: 33\n",
      "-- checking probes in region: JAM2\n",
      "--- gene=JAM2, id=2 removed by soft count = 135\n",
      "--- number of probes: 30 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: LCA5L\n",
      "--- gene=LCA5L, id=23 removed by soft count = 31\n",
      "--- gene=LCA5L, id=89 removed by soft count = 81\n",
      "--- gene=LCA5L, id=90 removed by soft count = 76\n",
      "--- gene=LCA5L, id=93 removed by soft count = 185\n",
      "--- number of probes: 94 , kept by blast: 90 , if remove dups: 90\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: LSS\n",
      "--- gene=LSS, id=11 removed by soft count = 39\n",
      "--- gene=LSS, id=15 removed by soft count = 36\n",
      "--- gene=LSS, id=33 removed by soft count = 3715\n",
      "--- number of probes: 35 , kept by blast: 32 , if remove dups: 32\n",
      "-- number of probes kept for this region: 32\n",
      "-- checking probes in region: LTN1\n",
      "--- gene=LTN1, id=0 removed by soft count = 46\n",
      "--- gene=LTN1, id=4 removed by soft count = 128\n",
      "--- gene=LTN1, id=5 removed by soft count = 834\n",
      "--- gene=LTN1, id=7 removed by soft count = 33\n",
      "--- number of probes: 37 , kept by blast: 33 , if remove dups: 33\n",
      "-- number of probes kept for this region: 33\n",
      "-- checking probes in region: MCM3AP\n",
      "--- number of probes: 20 , kept by blast: 20 , if remove dups: 20\n",
      "-- number of probes kept for this region: 20\n",
      "-- checking probes in region: MIS18A\n",
      "--- gene=MIS18A, id=2 removed by soft count = 143\n",
      "--- number of probes: 27 , kept by blast: 26 , if remove dups: 26\n",
      "-- number of probes kept for this region: 26\n",
      "-- checking probes in region: MORC3\n",
      "--- gene=MORC3, id=0 removed by soft count = 147\n",
      "--- gene=MORC3, id=12 removed by soft count = 162\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- number of probes: 15 , kept by blast: 13 , if remove dups: 13\n",
      "-- number of probes kept for this region: 13\n",
      "-- checking probes in region: MRPL39\n",
      "--- gene=MRPL39, id=77 removed by soft count = 117\n",
      "--- number of probes: 79 , kept by blast: 78 , if remove dups: 78\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 78\n",
      "-- checking probes in region: MRPS6\n",
      "--- gene=MRPS6, id=0 removed by soft count = 31\n",
      "--- gene=MRPS6, id=6 removed by soft count = 33\n",
      "--- gene=MRPS6, id=8 removed by soft count = 36\n",
      "--- gene=MRPS6, id=25 removed by soft count = 39\n",
      "--- number of probes: 34 , kept by blast: 30 , if remove dups: 30\n",
      "-- number of probes kept for this region: 30\n",
      "-- checking probes in region: MX1\n",
      "--- gene=MX1, id=0 removed by soft count = 87\n",
      "--- gene=MX1, id=1 removed by soft count = 34\n",
      "--- gene=MX1, id=14 removed by soft count = 40\n",
      "--- gene=MX1, id=17 removed by soft count = 94\n",
      "--- gene=MX1, id=24 removed by hard count = 5\n",
      "--- gene=MX1, id=25 removed by soft count = 1049\n",
      "--- gene=MX1, id=26 removed by soft count = 670\n",
      "--- gene=MX1, id=27 removed by soft count = 877\n",
      "--- gene=MX1, id=28 removed by soft count = 248\n",
      "--- number of probes: 30 , kept by blast: 21 , if remove dups: 21\n",
      "-- number of probes kept for this region: 21\n",
      "-- checking probes in region: MX2\n",
      "--- gene=MX2, id=47 removed by soft count = 94\n",
      "--- gene=MX2, id=66 removed by soft count = 53\n",
      "--- gene=MX2, id=69 removed by soft count = 9064\n",
      "--- gene=MX2, id=71 removed by soft count = 730\n",
      "--- number of probes: 79 , kept by blast: 75 , if remove dups: 75\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 75\n",
      "-- checking probes in region: N6AMT1\n",
      "--- gene=N6AMT1, id=6 removed by soft count = 103\n",
      "--- gene=N6AMT1, id=10 removed by soft count = 1012\n",
      "--- number of probes: 28 , kept by blast: 26 , if remove dups: 26\n",
      "-- number of probes kept for this region: 26\n",
      "-- checking probes in region: NCAM2\n",
      "--- gene=NCAM2, id=5 removed by soft count = 361\n",
      "--- gene=NCAM2, id=6 removed by soft count = 56\n",
      "--- gene=NCAM2, id=7 removed by soft count = 136\n",
      "--- gene=NCAM2, id=42 removed by soft count = 33\n",
      "--- gene=NCAM2, id=50 removed by soft count = 51\n",
      "--- number of probes: 103 , kept by blast: 98 , if remove dups: 98\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: NRIP1\n",
      "--- number of probes: 36 , kept by blast: 36 , if remove dups: 36\n",
      "-- number of probes kept for this region: 36\n",
      "-- checking probes in region: PAXBP1\n",
      "--- number of probes: 46 , kept by blast: 46 , if remove dups: 0\n",
      "-- number of probes kept for this region: 46\n",
      "-- checking probes in region: PCBP3\n",
      "--- gene=PCBP3, id=9 removed by soft count = 139\n",
      "--- gene=PCBP3, id=79 removed by soft count = 33\n",
      "--- gene=PCBP3, id=106 removed by soft count = 32\n",
      "--- number of probes: 107 , kept by blast: 104 , if remove dups: 104\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: PCNT\n",
      "--- gene=PCNT, id=3 removed by soft count = 9588\n",
      "--- number of probes: 30 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: PDE9A\n",
      "--- gene=PDE9A, id=86 removed by soft count = 58\n",
      "--- number of probes: 107 , kept by blast: 106 , if remove dups: 106\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: PDXK\n",
      "--- gene=PDXK, id=4 removed by soft count = 162\n",
      "--- gene=PDXK, id=6 removed by soft count = 64\n",
      "--- gene=PDXK, id=7 removed by soft count = 62\n",
      "--- number of probes: 24 , kept by blast: 21 , if remove dups: 21\n",
      "-- number of probes kept for this region: 21\n",
      "-- checking probes in region: PFKL\n",
      "--- gene=PFKL, id=10 removed by soft count = 134\n",
      "--- gene=PFKL, id=12 removed by soft count = 41\n",
      "--- gene=PFKL, id=13 removed by soft count = 346\n",
      "--- gene=PFKL, id=14 removed by soft count = 41\n",
      "--- gene=PFKL, id=17 removed by soft count = 105\n",
      "--- gene=PFKL, id=40 removed by soft count = 42\n",
      "--- number of probes: 91 , kept by blast: 85 , if remove dups: 84\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: PKNOX1\n",
      "--- gene=PKNOX1, id=1 removed by soft count = 484\n",
      "--- gene=PKNOX1, id=2 removed by soft count = 191\n",
      "--- gene=PKNOX1, id=3 removed by soft count = 599\n",
      "--- gene=PKNOX1, id=4 removed by soft count = 817\n",
      "--- gene=PKNOX1, id=5 removed by soft count = 1210\n",
      "--- gene=PKNOX1, id=15 removed by hard count = 0\n",
      "--- gene=PKNOX1, id=21 removed by soft count = 295\n",
      "--- gene=PKNOX1, id=22 removed by soft count = 235\n",
      "--- gene=PKNOX1, id=23 removed by soft count = 190\n",
      "--- gene=PKNOX1, id=24 removed by soft count = 872\n",
      "--- gene=PKNOX1, id=25 removed by soft count = 907\n",
      "--- gene=PKNOX1, id=26 removed by soft count = 419\n",
      "--- gene=PKNOX1, id=27 removed by soft count = 1150\n",
      "--- number of probes: 28 , kept by blast: 15 , if remove dups: 15\n",
      "-- number of probes kept for this region: 15\n",
      "-- checking probes in region: PRDM15\n",
      "--- gene=PRDM15, id=4 removed by soft count = 54\n",
      "--- gene=PRDM15, id=33 removed by soft count = 36\n",
      "--- number of probes: 35 , kept by blast: 33 , if remove dups: 33\n",
      "-- number of probes kept for this region: 33\n",
      "-- checking probes in region: PRMT2\n",
      "--- gene=PRMT2, id=0 removed by soft count = 156\n",
      "--- gene=PRMT2, id=1 removed by soft count = 70\n",
      "--- gene=PRMT2, id=2 removed by soft count = 138\n",
      "--- gene=PRMT2, id=3 removed by soft count = 93\n",
      "--- gene=PRMT2, id=4 removed by soft count = 61\n",
      "--- gene=PRMT2, id=5 removed by soft count = 36\n",
      "--- gene=PRMT2, id=6 removed by soft count = 116\n",
      "--- gene=PRMT2, id=7 removed by soft count = 97\n",
      "--- gene=PRMT2, id=9 removed by soft count = 197\n",
      "--- gene=PRMT2, id=10 removed by soft count = 80\n",
      "--- gene=PRMT2, id=11 removed by soft count = 48\n",
      "--- gene=PRMT2, id=12 removed by soft count = 278\n",
      "--- gene=PRMT2, id=13 removed by soft count = 152\n",
      "--- gene=PRMT2, id=14 removed by soft count = 543\n",
      "--- gene=PRMT2, id=15 removed by soft count = 917\n",
      "--- gene=PRMT2, id=16 removed by soft count = 38\n",
      "--- gene=PRMT2, id=17 removed by soft count = 116\n",
      "--- number of probes: 34 , kept by blast: 17 , if remove dups: 17\n",
      "-- number of probes kept for this region: 17\n",
      "-- checking probes in region: PSMG1\n",
      "--- gene=PSMG1, id=13 removed by soft count = 1232\n",
      "--- number of probes: 35 , kept by blast: 34 , if remove dups: 34\n",
      "-- number of probes kept for this region: 34\n",
      "-- checking probes in region: PTTG1IP\n",
      "--- gene=PTTG1IP, id=11 removed by hard count = 5\n",
      "--- number of probes: 22 , kept by blast: 21 , if remove dups: 20\n",
      "-- number of probes kept for this region: 21\n",
      "-- checking probes in region: PWP2\n",
      "--- number of probes: 73 , kept by blast: 73 , if remove dups: 3\n",
      "-- number of probes kept for this region: 73\n",
      "-- checking probes in region: RCAN1\n",
      "--- gene=RCAN1, id=23 removed by soft count = 32\n",
      "--- gene=RCAN1, id=24 removed by soft count = 158\n",
      "--- gene=RCAN1, id=25 removed by soft count = 277\n",
      "--- gene=RCAN1, id=26 removed by soft count = 1381\n",
      "--- gene=RCAN1, id=27 removed by soft count = 379\n",
      "--- gene=RCAN1, id=28 removed by soft count = 210\n",
      "--- gene=RCAN1, id=29 removed by soft count = 299\n",
      "--- gene=RCAN1, id=30 removed by soft count = 193\n",
      "--- gene=RCAN1, id=31 removed by soft count = 199\n",
      "--- number of probes: 32 , kept by blast: 23 , if remove dups: 23\n",
      "-- number of probes kept for this region: 23\n",
      "-- checking probes in region: RRP1\n",
      "--- gene=RRP1, id=3 removed by soft count = 438\n",
      "--- gene=RRP1, id=13 removed by soft count = 962\n",
      "--- gene=RRP1, id=15 removed by soft count = 269\n",
      "--- number of probes: 32 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: RRP1B\n",
      "--- gene=RRP1B, id=41 removed by soft count = 6612\n",
      "--- number of probes: 62 , kept by blast: 61 , if remove dups: 61\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 61\n",
      "-- checking probes in region: RUNX1\n",
      "--- gene=RUNX1, id=36 removed by soft count = 65\n",
      "--- gene=RUNX1, id=41 removed by hard count = 0\n",
      "--- gene=RUNX1, id=43 removed by soft count = 122\n",
      "--- gene=RUNX1, id=45 removed by soft count = 112\n",
      "--- gene=RUNX1, id=46 removed by soft count = 78\n",
      "--- gene=RUNX1, id=56 removed by soft count = 45\n",
      "--- number of probes: 72 , kept by blast: 66 , if remove dups: 66\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 66\n",
      "-- checking probes in region: SCAF4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- number of probes: 40 , kept by blast: 40 , if remove dups: 40\n",
      "-- number of probes kept for this region: 40\n",
      "-- checking probes in region: SETD4\n",
      "--- gene=SETD4, id=7 removed by soft count = 201\n",
      "--- gene=SETD4, id=9 removed by soft count = 353\n",
      "--- gene=SETD4, id=10 removed by soft count = 51\n",
      "--- gene=SETD4, id=15 removed by soft count = 3495\n",
      "--- gene=SETD4, id=16 removed by soft count = 200\n",
      "--- gene=SETD4, id=17 removed by soft count = 360\n",
      "--- gene=SETD4, id=18 removed by soft count = 1624\n",
      "--- number of probes: 25 , kept by blast: 18 , if remove dups: 18\n",
      "-- number of probes kept for this region: 18\n",
      "-- checking probes in region: SIK1\n",
      "--- gene=SIK1, id=20 removed by hard count = 0\n",
      "--- number of probes: 53 , kept by blast: 52 , if remove dups: 1\n",
      "-- number of probes kept for this region: 52\n",
      "-- checking probes in region: SIM2\n",
      "--- gene=SIM2, id=82 removed by soft count = 38\n",
      "--- number of probes: 107 , kept by blast: 106 , if remove dups: 106\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: SLC19A1\n",
      "--- gene=SLC19A1, id=26 removed by soft count = 192\n",
      "--- gene=SLC19A1, id=27 removed by soft count = 32\n",
      "--- gene=SLC19A1, id=29 removed by soft count = 80\n",
      "--- gene=SLC19A1, id=33 removed by soft count = 48\n",
      "--- gene=SLC19A1, id=35 removed by soft count = 1380\n",
      "--- gene=SLC19A1, id=40 removed by soft count = 296\n",
      "--- number of probes: 41 , kept by blast: 35 , if remove dups: 35\n",
      "-- number of probes kept for this region: 35\n",
      "-- checking probes in region: SLC37A1\n",
      "--- gene=SLC37A1, id=84 removed by soft count = 94\n",
      "--- gene=SLC37A1, id=85 removed by soft count = 121\n",
      "--- gene=SLC37A1, id=86 removed by soft count = 113\n",
      "--- gene=SLC37A1, id=87 removed by soft count = 169\n",
      "--- gene=SLC37A1, id=91 removed by soft count = 48\n",
      "--- gene=SLC37A1, id=92 removed by soft count = 48\n",
      "--- gene=SLC37A1, id=93 removed by soft count = 82\n",
      "--- gene=SLC37A1, id=94 removed by soft count = 36\n",
      "--- number of probes: 95 , kept by blast: 87 , if remove dups: 87\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: SOD1\n",
      "--- gene=SOD1, id=5 removed by soft count = 43\n",
      "--- gene=SOD1, id=7 removed by soft count = 360\n",
      "--- gene=SOD1, id=8 removed by soft count = 46\n",
      "--- gene=SOD1, id=80 removed by soft count = 31\n",
      "--- gene=SOD1, id=84 removed by soft count = 37\n",
      "--- number of probes: 91 , kept by blast: 86 , if remove dups: 86\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: SON\n",
      "--- number of probes: 56 , kept by blast: 56 , if remove dups: 56\n",
      "-- number of probes kept for this region: 56\n",
      "-- checking probes in region: SUMO3\n",
      "--- number of probes: 36 , kept by blast: 36 , if remove dups: 36\n",
      "-- number of probes kept for this region: 36\n",
      "-- checking probes in region: SYNJ1\n",
      "--- gene=SYNJ1, id=1 removed by soft count = 43\n",
      "--- gene=SYNJ1, id=31 removed by soft count = 120\n",
      "--- number of probes: 32 , kept by blast: 30 , if remove dups: 30\n",
      "-- number of probes kept for this region: 30\n",
      "-- checking probes in region: TIAM1\n",
      "--- gene=TIAM1, id=12 removed by soft count = 79\n",
      "--- gene=TIAM1, id=23 removed by soft count = 78\n",
      "--- gene=TIAM1, id=26 removed by soft count = 36\n",
      "--- gene=TIAM1, id=27 removed by soft count = 45\n",
      "--- number of probes: 30 , kept by blast: 26 , if remove dups: 26\n",
      "-- number of probes kept for this region: 26\n",
      "-- checking probes in region: TRAPPC10\n",
      "--- gene=TRAPPC10, id=0 removed by soft count = 243\n",
      "--- gene=TRAPPC10, id=10 removed by soft count = 44\n",
      "--- gene=TRAPPC10, id=13 removed by soft count = 289\n",
      "--- gene=TRAPPC10, id=14 removed by soft count = 798\n",
      "--- gene=TRAPPC10, id=15 removed by soft count = 279\n",
      "--- gene=TRAPPC10, id=18 removed by soft count = 1787\n",
      "--- gene=TRAPPC10, id=19 removed by soft count = 2143\n",
      "--- gene=TRAPPC10, id=36 removed by soft count = 11773\n",
      "--- number of probes: 37 , kept by blast: 29 , if remove dups: 29\n",
      "-- number of probes kept for this region: 29\n",
      "-- checking probes in region: TTC3\n",
      "--- gene=TTC3, id=1 removed by soft count = 93\n",
      "--- gene=TTC3, id=38 removed by soft count = 94\n",
      "--- gene=TTC3, id=39 removed by soft count = 142\n",
      "--- number of probes: 41 , kept by blast: 38 , if remove dups: 38\n",
      "-- number of probes kept for this region: 38\n",
      "-- checking probes in region: UBE2G2\n",
      "--- gene=UBE2G2, id=0 removed by soft count = 117\n",
      "--- gene=UBE2G2, id=1 removed by soft count = 293\n",
      "--- gene=UBE2G2, id=2 removed by soft count = 221\n",
      "--- gene=UBE2G2, id=3 removed by soft count = 148\n",
      "--- gene=UBE2G2, id=6 removed by soft count = 127\n",
      "--- gene=UBE2G2, id=7 removed by soft count = 69\n",
      "--- gene=UBE2G2, id=12 removed by soft count = 1372\n",
      "--- gene=UBE2G2, id=13 removed by soft count = 32\n",
      "--- gene=UBE2G2, id=14 removed by soft count = 652\n",
      "--- gene=UBE2G2, id=15 removed by soft count = 341\n",
      "--- number of probes: 100 , kept by blast: 90 , if remove dups: 90\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: URB1\n",
      "--- gene=URB1, id=8 removed by soft count = 5913\n",
      "--- gene=URB1, id=80 removed by soft count = 620\n",
      "--- number of probes: 87 , kept by blast: 85 , if remove dups: 85\n",
      "--- remove duplicated probes\n",
      "-- number of probes kept for this region: 80\n",
      "-- checking probes in region: USP16\n",
      "--- gene=USP16, id=0 removed by soft count = 9542\n",
      "--- number of probes: 19 , kept by blast: 18 , if remove dups: 18\n",
      "-- number of probes kept for this region: 18\n",
      "-- checking probes in region: USP25\n",
      "--- gene=USP25, id=0 removed by soft count = 224\n",
      "--- gene=USP25, id=1 removed by soft count = 784\n",
      "--- gene=USP25, id=2 removed by soft count = 39\n",
      "--- gene=USP25, id=3 removed by soft count = 35\n",
      "--- gene=USP25, id=4 removed by soft count = 608\n",
      "--- gene=USP25, id=5 removed by soft count = 333\n",
      "--- gene=USP25, id=18 removed by hard count = 0\n",
      "--- number of probes: 31 , kept by blast: 24 , if remove dups: 24\n",
      "-- number of probes kept for this region: 24\n",
      "-- checking probes in region: WDR4\n",
      "--- number of probes: 18 , kept by blast: 18 , if remove dups: 18\n",
      "-- number of probes kept for this region: 18\n",
      "-- checking probes in region: WRB\n",
      "--- gene=WRB, id=1 removed by soft count = 101\n",
      "--- gene=WRB, id=2 removed by soft count = 53\n",
      "--- gene=WRB, id=3 removed by soft count = 106\n",
      "--- gene=WRB, id=10 removed by soft count = 344\n",
      "--- gene=WRB, id=21 removed by soft count = 86\n",
      "--- number of probes: 22 , kept by blast: 17 , if remove dups: 17\n",
      "-- number of probes kept for this region: 17\n",
      "- summarize\n",
      "-- region: ADAMTS1 , number of probes: 80\n",
      "-- region: ADAMTS5 , number of probes: 31\n",
      "-- region: ADARB1 , number of probes: 34\n",
      "-- region: AGPAT3 , number of probes: 20\n",
      "-- region: ANKRD20A11P , number of probes: 43\n",
      "-- region: APP , number of probes: 24\n",
      "-- region: ATP5PF , number of probes: 29\n",
      "-- region: BACE2 , number of probes: 35\n",
      "-- region: BACH1 , number of probes: 67\n",
      "-- region: BRWD1 , number of probes: 25\n",
      "-- region: BTG3 , number of probes: 50\n",
      "-- region: C21orf91 , number of probes: 28\n",
      "-- region: C2CD2 , number of probes: 80\n",
      "-- region: CBR3 , number of probes: 16\n",
      "-- region: CCT8 , number of probes: 22\n",
      "-- region: CHAF1B , number of probes: 26\n",
      "-- region: COL18A1 , number of probes: 80\n",
      "-- region: COL6A1 , number of probes: 32\n",
      "-- region: COL6A2 , number of probes: 80\n",
      "-- region: CRYZL1 , number of probes: 26\n",
      "-- region: CXADR , number of probes: 30\n",
      "-- region: CYP4F29P , number of probes: 32\n",
      "-- region: DIP2A , number of probes: 40\n",
      "-- region: DONSON , number of probes: 38\n",
      "-- region: DSCR9 , number of probes: 68\n",
      "-- region: DYRK1A , number of probes: 29\n",
      "-- region: ETS2 , number of probes: 80\n",
      "-- region: EVA1C , number of probes: 18\n",
      "-- region: FAM207A , number of probes: 80\n",
      "-- region: HLCS , number of probes: 25\n",
      "-- region: HMGN1 , number of probes: 80\n",
      "-- region: HSPA13 , number of probes: 25\n",
      "-- region: HUNK , number of probes: 33\n",
      "-- region: IFNAR1 , number of probes: 14\n",
      "-- region: IFNAR2 , number of probes: 20\n",
      "-- region: IL10RB , number of probes: 29\n",
      "-- region: ITGB2 , number of probes: 33\n",
      "-- region: JAM2 , number of probes: 29\n",
      "-- region: LCA5L , number of probes: 80\n",
      "-- region: LSS , number of probes: 32\n",
      "-- region: LTN1 , number of probes: 33\n",
      "-- region: MCM3AP , number of probes: 20\n",
      "-- region: MIS18A , number of probes: 26\n",
      "-- region: MORC3 , number of probes: 13\n",
      "-- region: MRPL39 , number of probes: 78\n",
      "-- region: MRPS6 , number of probes: 30\n",
      "-- region: MX1 , number of probes: 21\n",
      "-- region: MX2 , number of probes: 75\n",
      "-- region: N6AMT1 , number of probes: 26\n",
      "-- region: NCAM2 , number of probes: 80\n",
      "-- region: NRIP1 , number of probes: 36\n",
      "-- region: PAXBP1 , number of probes: 46\n",
      "-- region: PCBP3 , number of probes: 80\n",
      "-- region: PCNT , number of probes: 29\n",
      "-- region: PDE9A , number of probes: 80\n",
      "-- region: PDXK , number of probes: 21\n",
      "-- region: PFKL , number of probes: 80\n",
      "-- region: PKNOX1 , number of probes: 15\n",
      "-- region: PRDM15 , number of probes: 33\n",
      "-- region: PRMT2 , number of probes: 17\n",
      "-- region: PSMG1 , number of probes: 34\n",
      "-- region: PTTG1IP , number of probes: 21\n",
      "-- region: PWP2 , number of probes: 73\n",
      "-- region: RCAN1 , number of probes: 23\n",
      "-- region: RRP1 , number of probes: 29\n",
      "-- region: RRP1B , number of probes: 61\n",
      "-- region: RUNX1 , number of probes: 66\n",
      "-- region: SCAF4 , number of probes: 40\n",
      "-- region: SETD4 , number of probes: 18\n",
      "-- region: SIK1 , number of probes: 52\n",
      "-- region: SIM2 , number of probes: 80\n",
      "-- region: SLC19A1 , number of probes: 35\n",
      "-- region: SLC37A1 , number of probes: 80\n",
      "-- region: SOD1 , number of probes: 80\n",
      "-- region: SON , number of probes: 56\n",
      "-- region: SUMO3 , number of probes: 36\n",
      "-- region: SYNJ1 , number of probes: 30\n",
      "-- region: TIAM1 , number of probes: 26\n",
      "-- region: TRAPPC10 , number of probes: 29\n",
      "-- region: TTC3 , number of probes: 38\n",
      "-- region: UBE2G2 , number of probes: 80\n",
      "-- region: URB1 , number of probes: 80\n",
      "-- region: USP16 , number of probes: 18\n",
      "-- region: USP25 , number of probes: 24\n",
      "-- region: WDR4 , number of probes: 18\n",
      "-- region: WRB , number of probes: 17\n",
      "- Number of probes kept: 3626\n",
      "- Saving to file: X:\\Libraries\\CTP-06\\chr21_promoters\\.\\blast_full_probes.fasta\n"
     ]
    }
   ],
   "source": [
    "kept_pbs, blast_keep_dic, hard_count_list, soft_count_list = quality_check.Screening_Probes_by_Blast(library_folder, 80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## extra selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-14T02:32:10.293226Z",
     "start_time": "2019-06-14T02:31:57.981059Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Checking primer usage, total probes: 3626\n",
      "True\n",
      "gene: ADAMTS1 -> 80\n",
      "gene: ADAMTS5 -> 31\n",
      "gene: ADARB1 -> 34\n",
      "gene: AGPAT3 -> 20\n",
      "gene: ANKRD20A11P -> 43\n",
      "gene: APP -> 24\n",
      "gene: ATP5PF -> 29\n",
      "gene: BACE2 -> 35\n",
      "gene: BACH1 -> 67\n",
      "gene: BRWD1 -> 25\n",
      "gene: BTG3 -> 50\n",
      "gene: C21orf91 -> 28\n",
      "gene: C2CD2 -> 80\n",
      "gene: CBR3 -> 16\n",
      "gene: CCT8 -> 22\n",
      "gene: CHAF1B -> 26\n",
      "gene: COL18A1 -> 80\n",
      "gene: COL6A1 -> 32\n",
      "gene: COL6A2 -> 80\n",
      "gene: CRYZL1 -> 26\n",
      "gene: CXADR -> 30\n",
      "gene: CYP4F29P -> 32\n",
      "gene: DIP2A -> 40\n",
      "gene: DONSON -> 38\n",
      "gene: DSCR9 -> 68\n",
      "gene: DYRK1A -> 29\n",
      "gene: ETS2 -> 80\n",
      "gene: EVA1C -> 18\n",
      "gene: FAM207A -> 80\n",
      "gene: HLCS -> 25\n",
      "gene: HMGN1 -> 80\n",
      "gene: HSPA13 -> 25\n",
      "gene: HUNK -> 33\n",
      "gene: IFNAR1 -> 14\n",
      "gene: IFNAR2 -> 20\n",
      "gene: IL10RB -> 29\n",
      "gene: ITGB2 -> 33\n",
      "gene: JAM2 -> 29\n",
      "gene: LCA5L -> 80\n",
      "gene: LSS -> 32\n",
      "gene: LTN1 -> 33\n",
      "gene: MCM3AP -> 20\n",
      "gene: MIS18A -> 26\n",
      "gene: MORC3 -> 13\n",
      "gene: MRPL39 -> 78\n",
      "gene: MRPS6 -> 30\n",
      "gene: MX1 -> 21\n",
      "gene: MX2 -> 75\n",
      "gene: N6AMT1 -> 26\n",
      "gene: NCAM2 -> 80\n",
      "gene: NRIP1 -> 36\n",
      "gene: PAXBP1 -> 46\n",
      "gene: PCBP3 -> 80\n",
      "gene: PCNT -> 29\n",
      "gene: PDE9A -> 80\n",
      "gene: PDXK -> 21\n",
      "gene: PFKL -> 80\n",
      "gene: PKNOX1 -> 15\n",
      "gene: PRDM15 -> 33\n",
      "gene: PRMT2 -> 17\n",
      "gene: PSMG1 -> 34\n",
      "gene: PTTG1IP -> 21\n",
      "gene: PWP2 -> 73\n",
      "gene: RCAN1 -> 23\n",
      "gene: RRP1 -> 29\n",
      "gene: RRP1B -> 61\n",
      "gene: RUNX1 -> 66\n",
      "gene: SCAF4 -> 40\n",
      "gene: SETD4 -> 18\n",
      "gene: SIK1 -> 52\n",
      "gene: SIM2 -> 80\n",
      "gene: SLC19A1 -> 35\n",
      "gene: SLC37A1 -> 80\n",
      "gene: SOD1 -> 80\n",
      "gene: SON -> 56\n",
      "gene: SUMO3 -> 36\n",
      "gene: SYNJ1 -> 30\n",
      "gene: TIAM1 -> 26\n",
      "gene: TRAPPC10 -> 29\n",
      "gene: TTC3 -> 38\n",
      "gene: UBE2G2 -> 80\n",
      "gene: URB1 -> 80\n",
      "gene: USP16 -> 18\n",
      "gene: USP25 -> 24\n",
      "gene: WDR4 -> 18\n",
      "gene: WRB -> 17\n",
      "False\n",
      "ADAMTS1 -> ['NDB_1147', 'NDB_1147', 'NDB_1147']\n",
      "ADAMTS5 -> ['NDB_1130', 'NDB_1130', 'NDB_1130']\n",
      "ADARB1 -> ['NDB_1157', 'NDB_1157', 'NDB_1157']\n",
      "AGPAT3 -> ['NDB_1176', 'NDB_1176', 'NDB_1176']\n",
      "ANKRD20A11P -> ['NDB_1209', 'NDB_1209', 'NDB_1209']\n",
      "APP -> ['NDB_1146', 'NDB_1146', 'NDB_1146']\n",
      "ATP5PF -> ['NDB_1136', 'NDB_1136', 'NDB_1136']\n",
      "BACE2 -> ['NDB_1137', 'NDB_1137', 'NDB_1137']\n",
      "BACH1 -> ['NDB_1175', 'NDB_1175', 'NDB_1175']\n",
      "BRWD1 -> ['NDB_1170', 'NDB_1170', 'NDB_1170']\n",
      "BTG3 -> ['NDB_1163', 'NDB_1163', 'NDB_1163']\n",
      "C21orf91 -> ['NDB_1132', 'NDB_1132', 'NDB_1132']\n",
      "C2CD2 -> ['NDB_1131', 'NDB_1131', 'NDB_1131']\n",
      "CBR3 -> ['NDB_1126', 'NDB_1126', 'NDB_1126']\n",
      "CCT8 -> ['NDB_1127', 'NDB_1127', 'NDB_1127']\n",
      "CHAF1B -> ['NDB_1174', 'NDB_1174', 'NDB_1174']\n",
      "COL18A1 -> ['NDB_1166', 'NDB_1166', 'NDB_1166']\n",
      "COL6A1 -> ['NDB_1208', 'NDB_1208', 'NDB_1208']\n",
      "COL6A2 -> ['NDB_1140', 'NDB_1140', 'NDB_1140']\n",
      "CRYZL1 -> ['NDB_1134', 'NDB_1134', 'NDB_1134']\n",
      "CXADR -> ['NDB_1203', 'NDB_1203', 'NDB_1203']\n",
      "CYP4F29P -> ['NDB_1199', 'NDB_1199', 'NDB_1199']\n",
      "DIP2A -> ['NDB_1193', 'NDB_1193', 'NDB_1193']\n",
      "DONSON -> ['NDB_1158', 'NDB_1158', 'NDB_1158']\n",
      "DSCR9 -> ['NDB_1168', 'NDB_1168', 'NDB_1168']\n",
      "DYRK1A -> ['NDB_1207', 'NDB_1207', 'NDB_1207']\n",
      "ETS2 -> ['NDB_1153', 'NDB_1153', 'NDB_1153']\n",
      "EVA1C -> ['NDB_1202', 'NDB_1202', 'NDB_1202']\n",
      "FAM207A -> ['NDB_1177', 'NDB_1177', 'NDB_1177']\n",
      "HLCS -> ['NDB_1160', 'NDB_1160', 'NDB_1160']\n",
      "HMGN1 -> ['NDB_1161', 'NDB_1161', 'NDB_1161']\n",
      "HSPA13 -> ['NDB_1159', 'NDB_1159', 'NDB_1159']\n",
      "HUNK -> ['NDB_1200', 'NDB_1200', 'NDB_1200']\n",
      "IFNAR1 -> ['NDB_1149', 'NDB_1149', 'NDB_1149']\n",
      "IFNAR2 -> ['NDB_1145', 'NDB_1145', 'NDB_1145']\n",
      "IL10RB -> ['NDB_1171', 'NDB_1171', 'NDB_1171']\n",
      "ITGB2 -> ['NDB_1189', 'NDB_1189', 'NDB_1189']\n",
      "JAM2 -> ['NDB_1192', 'NDB_1192', 'NDB_1192']\n",
      "LCA5L -> ['NDB_1194', 'NDB_1194', 'NDB_1194']\n",
      "LSS -> ['NDB_1139', 'NDB_1139', 'NDB_1139']\n",
      "LTN1 -> ['NDB_1169', 'NDB_1169', 'NDB_1169']\n",
      "MCM3AP -> ['NDB_1141', 'NDB_1141', 'NDB_1141']\n",
      "MIS18A -> ['NDB_1196', 'NDB_1196', 'NDB_1196']\n",
      "MORC3 -> ['NDB_1155', 'NDB_1155', 'NDB_1155']\n",
      "MRPL39 -> ['NDB_1205', 'NDB_1205', 'NDB_1205']\n",
      "MRPS6 -> ['NDB_1150', 'NDB_1150', 'NDB_1150']\n",
      "MX1 -> ['NDB_1151', 'NDB_1151', 'NDB_1151']\n",
      "MX2 -> ['NDB_1152', 'NDB_1152', 'NDB_1152']\n",
      "N6AMT1 -> ['NDB_1186', 'NDB_1186', 'NDB_1186']\n",
      "NCAM2 -> ['NDB_1135', 'NDB_1135', 'NDB_1135']\n",
      "NRIP1 -> ['NDB_1156', 'NDB_1156', 'NDB_1156']\n",
      "PAXBP1 -> ['NDB_1211', 'NDB_1211', 'NDB_1211']\n",
      "PCBP3 -> ['NDB_1164', 'NDB_1164', 'NDB_1164']\n",
      "PCNT -> ['NDB_1133', 'NDB_1133', 'NDB_1133']\n",
      "PDE9A -> ['NDB_1173', 'NDB_1173', 'NDB_1173']\n",
      "PDXK -> ['NDB_1172', 'NDB_1172', 'NDB_1172']\n",
      "PFKL -> ['NDB_1184', 'NDB_1184', 'NDB_1184']\n",
      "PKNOX1 -> ['NDB_1138', 'NDB_1138', 'NDB_1138']\n",
      "PRDM15 -> ['NDB_1198', 'NDB_1198', 'NDB_1198']\n",
      "PRMT2 -> ['NDB_1180', 'NDB_1180', 'NDB_1180']\n",
      "PSMG1 -> ['NDB_1210', 'NDB_1210', 'NDB_1210']\n",
      "PTTG1IP -> ['NDB_1183', 'NDB_1183', 'NDB_1183']\n",
      "PWP2 -> ['NDB_1195', 'NDB_1195', 'NDB_1195']\n",
      "RCAN1 -> ['NDB_1191', 'NDB_1191', 'NDB_1191']\n",
      "RRP1 -> ['NDB_1181', 'NDB_1181', 'NDB_1181']\n",
      "RRP1B -> ['NDB_1190', 'NDB_1190', 'NDB_1190']\n",
      "RUNX1 -> ['NDB_1129', 'NDB_1129', 'NDB_1129']\n",
      "SCAF4 -> ['NDB_1204', 'NDB_1204', 'NDB_1204']\n",
      "SETD4 -> ['NDB_1188', 'NDB_1188', 'NDB_1188']\n",
      "SIK1 -> ['NDB_1182', 'NDB_1182', 'NDB_1182']\n",
      "SIM2 -> ['NDB_1206', 'NDB_1206', 'NDB_1206']\n",
      "SLC19A1 -> ['NDB_1201', 'NDB_1201', 'NDB_1201']\n",
      "SLC37A1 -> ['NDB_1187', 'NDB_1187', 'NDB_1187']\n",
      "SOD1 -> ['NDB_1162', 'NDB_1162', 'NDB_1162']\n",
      "SON -> ['NDB_1154', 'NDB_1154', 'NDB_1154']\n",
      "SUMO3 -> ['NDB_1128', 'NDB_1128', 'NDB_1128']\n",
      "SYNJ1 -> ['NDB_1142', 'NDB_1142', 'NDB_1142']\n",
      "TIAM1 -> ['NDB_1197', 'NDB_1197', 'NDB_1197']\n",
      "TRAPPC10 -> ['NDB_1178', 'NDB_1178', 'NDB_1178']\n",
      "TTC3 -> ['NDB_1143', 'NDB_1143', 'NDB_1143']\n",
      "UBE2G2 -> ['NDB_1144', 'NDB_1144', 'NDB_1144']\n",
      "URB1 -> ['NDB_1165', 'NDB_1165', 'NDB_1165']\n",
      "USP16 -> ['NDB_1167', 'NDB_1167', 'NDB_1167']\n",
      "USP25 -> ['NDB_1179', 'NDB_1179', 'NDB_1179']\n",
      "WDR4 -> ['NDB_1148', 'NDB_1148', 'NDB_1148']\n",
      "WRB -> ['NDB_1185', 'NDB_1185', 'NDB_1185']\n",
      "True\n",
      "NDB_1126 -> ['CBR3', 'CBR3', 'CBR3']\n",
      "NDB_1127 -> ['CCT8', 'CCT8', 'CCT8']\n",
      "NDB_1128 -> ['SUMO3', 'SUMO3', 'SUMO3']\n",
      "NDB_1129 -> ['RUNX1', 'RUNX1', 'RUNX1']\n",
      "NDB_1130 -> ['ADAMTS5', 'ADAMTS5', 'ADAMTS5']\n",
      "NDB_1131 -> ['C2CD2', 'C2CD2', 'C2CD2']\n",
      "NDB_1132 -> ['C21orf91', 'C21orf91', 'C21orf91']\n",
      "NDB_1133 -> ['PCNT', 'PCNT', 'PCNT']\n",
      "NDB_1134 -> ['CRYZL1', 'CRYZL1', 'CRYZL1']\n",
      "NDB_1135 -> ['NCAM2', 'NCAM2', 'NCAM2']\n",
      "NDB_1136 -> ['ATP5PF', 'ATP5PF', 'ATP5PF']\n",
      "NDB_1137 -> ['BACE2', 'BACE2', 'BACE2']\n",
      "NDB_1138 -> ['PKNOX1', 'PKNOX1', 'PKNOX1']\n",
      "NDB_1139 -> ['LSS', 'LSS', 'LSS']\n",
      "NDB_1140 -> ['COL6A2', 'COL6A2', 'COL6A2']\n",
      "NDB_1141 -> ['MCM3AP', 'MCM3AP', 'MCM3AP']\n",
      "NDB_1142 -> ['SYNJ1', 'SYNJ1', 'SYNJ1']\n",
      "NDB_1143 -> ['TTC3', 'TTC3', 'TTC3']\n",
      "NDB_1144 -> ['UBE2G2', 'UBE2G2', 'UBE2G2']\n",
      "NDB_1145 -> ['IFNAR2', 'IFNAR2', 'IFNAR2']\n",
      "NDB_1146 -> ['APP', 'APP', 'APP']\n",
      "NDB_1147 -> ['ADAMTS1', 'ADAMTS1', 'ADAMTS1']\n",
      "NDB_1148 -> ['WDR4', 'WDR4', 'WDR4']\n",
      "NDB_1149 -> ['IFNAR1', 'IFNAR1', 'IFNAR1']\n",
      "NDB_1150 -> ['MRPS6', 'MRPS6', 'MRPS6']\n",
      "NDB_1151 -> ['MX1', 'MX1', 'MX1']\n",
      "NDB_1152 -> ['MX2', 'MX2', 'MX2']\n",
      "NDB_1153 -> ['ETS2', 'ETS2', 'ETS2']\n",
      "NDB_1154 -> ['SON', 'SON', 'SON']\n",
      "NDB_1155 -> ['MORC3', 'MORC3', 'MORC3']\n",
      "NDB_1156 -> ['NRIP1', 'NRIP1', 'NRIP1']\n",
      "NDB_1157 -> ['ADARB1', 'ADARB1', 'ADARB1']\n",
      "NDB_1158 -> ['DONSON', 'DONSON', 'DONSON']\n",
      "NDB_1159 -> ['HSPA13', 'HSPA13', 'HSPA13']\n",
      "NDB_1160 -> ['HLCS', 'HLCS', 'HLCS']\n",
      "NDB_1161 -> ['HMGN1', 'HMGN1', 'HMGN1']\n",
      "NDB_1162 -> ['SOD1', 'SOD1', 'SOD1']\n",
      "NDB_1163 -> ['BTG3', 'BTG3', 'BTG3']\n",
      "NDB_1164 -> ['PCBP3', 'PCBP3', 'PCBP3']\n",
      "NDB_1165 -> ['URB1', 'URB1', 'URB1']\n",
      "NDB_1166 -> ['COL18A1', 'COL18A1', 'COL18A1']\n",
      "NDB_1167 -> ['USP16', 'USP16', 'USP16']\n",
      "NDB_1168 -> ['DSCR9', 'DSCR9', 'DSCR9']\n",
      "NDB_1169 -> ['LTN1', 'LTN1', 'LTN1']\n",
      "NDB_1170 -> ['BRWD1', 'BRWD1', 'BRWD1']\n",
      "NDB_1171 -> ['IL10RB', 'IL10RB', 'IL10RB']\n",
      "NDB_1172 -> ['PDXK', 'PDXK', 'PDXK']\n",
      "NDB_1173 -> ['PDE9A', 'PDE9A', 'PDE9A']\n",
      "NDB_1174 -> ['CHAF1B', 'CHAF1B', 'CHAF1B']\n",
      "NDB_1175 -> ['BACH1', 'BACH1', 'BACH1']\n",
      "NDB_1176 -> ['AGPAT3', 'AGPAT3', 'AGPAT3']\n",
      "NDB_1177 -> ['FAM207A', 'FAM207A', 'FAM207A']\n",
      "NDB_1178 -> ['TRAPPC10', 'TRAPPC10', 'TRAPPC10']\n",
      "NDB_1179 -> ['USP25', 'USP25', 'USP25']\n",
      "NDB_1180 -> ['PRMT2', 'PRMT2', 'PRMT2']\n",
      "NDB_1181 -> ['RRP1', 'RRP1', 'RRP1']\n",
      "NDB_1182 -> ['SIK1', 'SIK1', 'SIK1']\n",
      "NDB_1183 -> ['PTTG1IP', 'PTTG1IP', 'PTTG1IP']\n",
      "NDB_1184 -> ['PFKL', 'PFKL', 'PFKL']\n",
      "NDB_1185 -> ['WRB', 'WRB', 'WRB']\n",
      "NDB_1186 -> ['N6AMT1', 'N6AMT1', 'N6AMT1']\n",
      "NDB_1187 -> ['SLC37A1', 'SLC37A1', 'SLC37A1']\n",
      "NDB_1188 -> ['SETD4', 'SETD4', 'SETD4']\n",
      "NDB_1189 -> ['ITGB2', 'ITGB2', 'ITGB2']\n",
      "NDB_1190 -> ['RRP1B', 'RRP1B', 'RRP1B']\n",
      "NDB_1191 -> ['RCAN1', 'RCAN1', 'RCAN1']\n",
      "NDB_1192 -> ['JAM2', 'JAM2', 'JAM2']\n",
      "NDB_1193 -> ['DIP2A', 'DIP2A', 'DIP2A']\n",
      "NDB_1194 -> ['LCA5L', 'LCA5L', 'LCA5L']\n",
      "NDB_1195 -> ['PWP2', 'PWP2', 'PWP2']\n",
      "NDB_1196 -> ['MIS18A', 'MIS18A', 'MIS18A']\n",
      "NDB_1197 -> ['TIAM1', 'TIAM1', 'TIAM1']\n",
      "NDB_1198 -> ['PRDM15', 'PRDM15', 'PRDM15']\n",
      "NDB_1199 -> ['CYP4F29P', 'CYP4F29P', 'CYP4F29P']\n",
      "NDB_1200 -> ['HUNK', 'HUNK', 'HUNK']\n",
      "NDB_1201 -> ['SLC19A1', 'SLC19A1', 'SLC19A1']\n",
      "NDB_1202 -> ['EVA1C', 'EVA1C', 'EVA1C']\n",
      "NDB_1203 -> ['CXADR', 'CXADR', 'CXADR']\n",
      "NDB_1204 -> ['SCAF4', 'SCAF4', 'SCAF4']\n",
      "NDB_1205 -> ['MRPL39', 'MRPL39', 'MRPL39']\n",
      "NDB_1206 -> ['SIM2', 'SIM2', 'SIM2']\n",
      "NDB_1207 -> ['DYRK1A', 'DYRK1A', 'DYRK1A']\n",
      "NDB_1208 -> ['COL6A1', 'COL6A1', 'COL6A1']\n",
      "NDB_1209 -> ['ANKRD20A11P', 'ANKRD20A11P', 'ANKRD20A11P']\n",
      "NDB_1210 -> ['PSMG1', 'PSMG1', 'PSMG1']\n",
      "NDB_1211 -> ['PAXBP1', 'PAXBP1', 'PAXBP1']\n",
      "True\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- saving internal 17-mer map to file:X:\\Libraries\\CTP-06\\chr21_promoters\\probe_table_17.npz\n",
      "True\n",
      "-- total probes removed by internal screening: 0\n"
     ]
    }
   ],
   "source": [
    "primer_check = check._check_primer_usage(kept_pb_list, primers[0], primers[1])\n",
    "print(primer_check)\n",
    "reg_size_dic, len_check = check._check_region_size(kept_pb_list,min_size=15)\n",
    "print(len_check)\n",
    "reg_readout_dic, reg2readout_check = check._check_region_to_readouts(kept_pb_list, readout_dict)\n",
    "print(reg2readout_check)\n",
    "\n",
    "readout_reg_dic, readout2reg_check = check._check_readout_to_region(reg_readout_dic, kept_pb_list, readout_dict)\n",
    "print(readout2reg_check)\n",
    "\n",
    "int_map = check._construct_internal_map(kept_pb_list, library_folder)\n",
    "\n",
    "readout_count_dic, readout_count_check = check._check_readout_in_probes(readout_reg_dic, reg_size_dic, int_map, readout_dict)\n",
    "print(readout_count_check)\n",
    "\n",
    "kept_records, removed_count = check._check_between_probes(kept_pb_list, int_map)\n",
    "# save kept records\n",
    "with open(os.path.join(library_folder, 'final_probes', 'extra_filtered_full_probes.fasta'), 'w') as output_handle:\n",
    "    SeqIO.write(kept_records, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
