{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This a jupyter notebook guide on compartment analysis for chr21\n",
    "\n",
    "by Pu Zheng and Jun-Han Su\n",
    "\n",
    "2020.06.06\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import sys, os, glob, time, copy\n",
    "import numpy as np\n",
    "import scipy\n",
    "import pickle\n",
    "\n",
    "sys.path.append(os.path.abspath(r\"..\\.\"))\n",
    "\n",
    "import source as ia\n",
    "\n",
    "from scipy.signal import find_peaks\n",
    "from scipy.spatial.distance import cdist,pdist,squareform\n",
    "\n",
    "print(os.getpid()) # print this so u can terminate through cmd / task-manager"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required plotting setting\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rc('font', family='serif')\n",
    "plt.rc('font', serif='Arial')\n",
    "_font_size = 7.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required plotting parameters\n",
    "from source.figure_tools import _dpi,_single_col_width,_double_col_width,_single_row_height,_ref_bar_length, _ticklabel_size,_ticklabel_width,_font_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure folder\n",
    "parent_figure_folder = r'\\\\10.245.74.158\\Chromatin_NAS_4\\Chromatin_Share\\final_figures'\n",
    "figure_folder = os.path.join(parent_figure_folder, 'Chr21_compartment_figures')\n",
    "print(figure_folder)\n",
    "if not os.path.exists(figure_folder):\n",
    "    os.makedirs(figure_folder)\n",
    "    print(\"generating this folder\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = r'E:\\Users\\puzheng\\Dropbox\\2020 Chromatin Imaging Manuscript\\Revision\\DataForReviewers'\n",
    "rep1_filename = os.path.join(data_folder, 'chromosome21.tsv')\n",
    "rep2_filename = os.path.join(data_folder, 'chromosome21-cell_cycle.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.1 load replicate 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load from file and extract info\n",
    "import csv\n",
    "rep1_info_dict = {}\n",
    "with open(rep1_filename, 'r') as _handle:\n",
    "    _reader = csv.reader(_handle, delimiter='\\t', quotechar='|')\n",
    "    _headers = next(_reader)\n",
    "    print(_headers)\n",
    "    # create keys for each header\n",
    "    for _h in _headers:\n",
    "        rep1_info_dict[_h] = []\n",
    "    # loop through content\n",
    "    for _contents in _reader:\n",
    "        for _h, _info in zip(_headers,_contents):\n",
    "            rep1_info_dict[_h].append(_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook as tqdm\n",
    "\n",
    "# clean up infoa\n",
    "data_rep1 = {'params':{}}\n",
    "\n",
    "# clean up genomic coordiantes\n",
    "region_names = np.array([_n for _n in sorted(np.unique(rep1_info_dict['Genomic coordinate']), \n",
    "                                             key=lambda s:int(s.split(':')[1].split('-')[0]))])\n",
    "region_starts = np.array([int(_n.split(':')[1].split('-')[0]) for _n in region_names])\n",
    "region_ends = np.array([int(_n.split(':')[1].split('-')[1]) for _n in region_names])[np.argsort(region_starts)]\n",
    "region_starts = np.sort(region_starts)\n",
    "\n",
    "mid_positions = ((region_starts + region_ends)/2).astype(np.int)\n",
    "mid_positions_Mb = np.round(mid_positions / 1e6, 2) \n",
    "\n",
    "# clean up chrom copy number\n",
    "chr_nums = np.array([int(_info) for _info in rep1_info_dict['Chromosome copy number']])\n",
    "chr_ids, region_cts = np.unique(chr_nums, return_counts=True)\n",
    "dna_zxys_list = [[[] for _start in region_starts] for _id in chr_ids]\n",
    "\n",
    "# clean up zxy\n",
    "for _z,_x,_y,_reg_info, _cid in tqdm(zip(rep1_info_dict['Z(nm)'],rep1_info_dict['X(nm)'],\\\n",
    "                                         rep1_info_dict['Y(nm)'],rep1_info_dict['Genomic coordinate'],\\\n",
    "                                         rep1_info_dict['Chromosome copy number'])):\n",
    "    # get chromosome inds\n",
    "    _cid = int(_cid)\n",
    "    _cind = np.where(chr_ids == _cid)[0][0]\n",
    "    \n",
    "    # get region indices\n",
    "    _start = int(_reg_info.split(':')[1].split('-')[0])\n",
    "    _rind = np.where(region_starts==_start)[0][0]\n",
    "    \n",
    "    dna_zxys_list[_cind][_rind] = np.array([float(_z),float(_x), float(_y)])\n",
    "\n",
    "# merge together\n",
    "dna_zxys_list = np.array(dna_zxys_list)\n",
    "data_rep1['chrom_ids'] = chr_ids\n",
    "data_rep1['region_names'] = region_names\n",
    "data_rep1['mid_position_Mb'] = mid_positions_Mb\n",
    "data_rep1['dna_zxys'] = dna_zxys_list\n",
    "\n",
    "# clean up tss and transcription\n",
    "if 'Gene names' in rep1_info_dict:\n",
    "    import re\n",
    "    # first extract number of genes\n",
    "    gene_names = []\n",
    "    for _gene_info, _trans_info, _tss_coord in zip(rep1_info_dict['Gene names'],\n",
    "                                                   rep1_info_dict['Transcription'],\n",
    "                                                   rep1_info_dict['TSS ZXY(nm)']):\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            for _gene in _genes:\n",
    "                if _gene not in gene_names:\n",
    "                    gene_names.append(_gene)\n",
    "    print(f\"{len(gene_names)} genes exist in this dataset.\")\n",
    "    \n",
    "    # initialize gene and transcription\n",
    "    tss_zxys_list = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    transcription_profiles = [[[] for _gene in gene_names] for _id in chr_ids]\n",
    "    # loop through to get info\n",
    "    for _cid, _gene_info, _trans_info, _tss_locations in tqdm(zip(rep1_info_dict['Chromosome copy number'],\n",
    "                                                                  rep1_info_dict['Gene names'],\n",
    "                                                                  rep1_info_dict['Transcription'],\n",
    "                                                                  rep1_info_dict['TSS ZXY(nm)'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        # process if there are genes in this region:\n",
    "        if _gene_info != '':\n",
    "            # split by semicolon\n",
    "            _genes = _gene_info.split(';')[:-1]\n",
    "            _transcribes = _trans_info.split(';')[:-1]\n",
    "            _tss_zxys = _tss_locations.split(';')[:-1]\n",
    "            for _gene, _transcribe, _tss_zxy in zip(_genes, _transcribes, _tss_zxys):\n",
    "                # get gene index\n",
    "                _gind = gene_names.index(_gene)\n",
    "                # get transcription profile\n",
    "                if _transcribe == 'on':\n",
    "                    transcription_profiles[_cind][_gind] = True\n",
    "                else:\n",
    "                    transcription_profiles[_cind][_gind] = False\n",
    "                # get coordinates\n",
    "                _tss_zxy = np.array([np.float(_c) for _c in re.split(r'\\s+', _tss_zxy.split('[')[1].split(']')[0]) if _c != ''])\n",
    "                tss_zxys_list[_cind][_gind] = _tss_zxy\n",
    "                \n",
    "    tss_zxys_list = np.array(tss_zxys_list)\n",
    "    transcription_profiles = np.array(transcription_profiles)\n",
    "    data_rep1['gene_names'] = gene_names\n",
    "    data_rep1['tss_zxys'] = tss_zxys_list\n",
    "    data_rep1['trans_pfs'] = transcription_profiles\n",
    "\n",
    "# clean up cell_cycle states\n",
    "if 'Cell cycle state' in rep1_info_dict:\n",
    "    cell_cycle_types = np.unique(rep1_info_dict['Cell cycle state'])\n",
    "    cell_cycle_flag_dict = {_k:[[] for _id in chr_ids] for _k in cell_cycle_types if _k != 'ND'}\n",
    "    for _cid, _state in tqdm(zip(rep1_info_dict['Chromosome copy number'],rep1_info_dict['Cell cycle state'])):\n",
    "        # get chromosome inds\n",
    "        _cid = int(_cid)\n",
    "        _cind = np.where(chr_ids == _cid)[0][0]\n",
    "        if np.array([_v[_cind]==[] for _k,_v in cell_cycle_flag_dict.items()]).any():\n",
    "            for _k,_v in cell_cycle_flag_dict.items():\n",
    "                if _k == _state:\n",
    "                    _v[_cind] = True\n",
    "                else:\n",
    "                    _v[_cind] = False\n",
    "    # append to data\n",
    "    for _k, _v in cell_cycle_flag_dict.items():\n",
    "        data_rep1[f'{_k}_flags'] = np.array(_v)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get population averaged maps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## imaging-based proximity freq. matrix and median distance matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zxys_rep1_list = np.array(data_rep1['dna_zxys'])\n",
    "distmap_rep1_list = np.array([squareform(pdist(_zxy)) for _zxy in zxys_rep1_list])\n",
    "\n",
    "# generate median distance map\n",
    "median_distance_map_rep1 = np.nanmedian(distmap_rep1_list, axis = 0)\n",
    "# generate contact map\n",
    "contact_th = 500\n",
    "contact_rep1_map = np.sum(distmap_rep1_list<contact_th, axis=0) / np.sum(np.isnan(distmap_rep1_list)==False, axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## corresponding Hi-C data from Rao et al."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hic_filename = os.path.join(data_folder, 'Hi-C matrices', 'Hi-C_contacts_chromosome21.tsv')\n",
    "\n",
    "hic_map = []\n",
    "\n",
    "with open(hic_filename, 'r') as _handle:\n",
    "    _reader = csv.reader(_handle, delimiter='\\t', quotechar='|')\n",
    "    col_regions = next(_reader)[1:]\n",
    "    row_regions = []\n",
    "    # loop through content\n",
    "    for _contents in _reader:\n",
    "        row_regions.append(_contents[0])\n",
    "        hic_map.append([int(_c) for _c in _contents[1:]])\n",
    "hic_map = np.array(hic_map)\n",
    "# sort row and col to match tsv dataset\n",
    "row_order = np.concatenate([np.where(data_rep1['region_names']==_rn)[0] for _rn in row_regions])\n",
    "col_order = np.concatenate([np.where(data_rep1['region_names']==_cn)[0] for _cn in col_regions])\n",
    "hic_map = hic_map[row_order][:, col_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.2 call compartments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compartment calling by PCA\n",
    "\n",
    "# Generate correlation map\n",
    "gaussian_sigma = 2.75\n",
    "\n",
    "# normalize genomic distance effects\n",
    "genomic_distance_map = squareform(pdist(data_rep1['mid_position_Mb'][:,np.newaxis]))\n",
    "genomic_distance_entries = genomic_distance_map[np.triu_indices(len(genomic_distance_map),1)]\n",
    "\n",
    "\n",
    "median_entries = median_distance_map_rep1[np.triu_indices(len(median_distance_map_rep1),1)]\n",
    "kept = (genomic_distance_entries > 0) * (median_entries > 0)\n",
    "median_lr = scipy.stats.linregress(np.log(genomic_distance_entries[kept]), np.log(median_entries[kept]))\n",
    "print(median_lr)\n",
    "median_norm_map = np.exp(np.log(genomic_distance_map) * median_lr.slope + median_lr.intercept)\n",
    "for _i in range(len(median_norm_map)):\n",
    "    median_norm_map[_i,_i] = 1\n",
    "median_normed_map = median_distance_map_rep1 / median_norm_map\n",
    "# apply gaussian\n",
    "from scipy.ndimage import gaussian_filter\n",
    "\n",
    "median_corr_map_rep1 = np.corrcoef(gaussian_filter(median_normed_map, gaussian_sigma))\n",
    "\n",
    "# normalize genomic distance effects\n",
    "contact_entries = contact_rep1_map[np.triu_indices(len(contact_rep1_map),1)]\n",
    "kept = (genomic_distance_entries > 0) * (contact_entries > 0)\n",
    "contact_lr = scipy.stats.linregress(np.log(genomic_distance_entries[kept]), np.log(contact_entries[kept]))\n",
    "print(contact_lr)\n",
    "contact_norm_map = np.exp(np.log(genomic_distance_map) * contact_lr.slope + contact_lr.intercept)\n",
    "#for _i in range(len(normalization_mat)):\n",
    "#    contact_norm_map[_i,_i] = 1\n",
    "contact_normed_map = contact_rep1_map / contact_norm_map\n",
    "# apply gaussian\n",
    "from scipy.ndimage import gaussian_filter\n",
    "\n",
    "contact_corr_map_rep1 = np.corrcoef(gaussian_filter(contact_normed_map, gaussian_sigma))\n",
    "\n",
    "# normalize genomic distance effects\n",
    "hic_entries = hic_map[np.triu_indices(len(hic_map),1)]\n",
    "kept = (genomic_distance_entries > 0) * (hic_entries > 0)\n",
    "hic_lr = scipy.stats.linregress(np.log(genomic_distance_entries[kept]), np.log(hic_entries[kept]))\n",
    "print(hic_lr)\n",
    "hic_norm_map = np.exp(np.log(genomic_distance_map) * hic_lr.slope + hic_lr.intercept)\n",
    "#for _i in range(len(normalization_mat)):\n",
    "#    hic_norm_map[_i,_i] = 1\n",
    "hic_normed_map = hic_map / hic_norm_map\n",
    "# apply gaussian\n",
    "from scipy.ndimage import gaussian_filter\n",
    "\n",
    "hic_corr_map = np.corrcoef(gaussian_filter(hic_normed_map, gaussian_sigma))\n",
    "\n",
    "# Do PCA\n",
    "from sklearn.decomposition import PCA\n",
    "median_model = PCA(1)\n",
    "median_model.fit(median_corr_map_rep1)\n",
    "median_pc1_rep1 = np.reshape(median_model.fit_transform(median_corr_map_rep1), -1)\n",
    "\n",
    "contact_model = PCA(1)\n",
    "contact_model.fit(contact_corr_map_rep1)\n",
    "contact_pc1_rep1 = np.reshape(contact_model.fit_transform(contact_corr_map_rep1), -1)\n",
    "\n",
    "hic_model = PCA(1)\n",
    "hic_model.fit(hic_corr_map)\n",
    "hic_pc1 = np.reshape(hic_model.fit_transform(hic_corr_map), -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define AB compartment by merging small sub-compartments\n",
    "\n",
    "temp_AB_dict = {'A':np.where(contact_pc1_rep1 >= 0)[0],\n",
    "                'B':np.where(contact_pc1_rep1 < 0)[0],}\n",
    "\n",
    "temp_AB_vector = np.ones(len(zxys_rep1_list[0])).astype(np.int) * -1\n",
    "temp_AB_vector[temp_AB_dict['A']] = 1\n",
    "temp_AB_vector[temp_AB_dict['B']] = 0\n",
    "\n",
    "num_small_compartment = np.inf\n",
    "prev_v = temp_AB_vector[0]\n",
    "while num_small_compartment > 0:\n",
    "    # find indices for all sub-comaprtments\n",
    "    all_comp_inds = []\n",
    "    _comp_inds = []\n",
    "    prev_v = temp_AB_vector[0] # initialize previous compartment\n",
    "    for _i, _v in enumerate(temp_AB_vector):\n",
    "        if prev_v != _v:\n",
    "            all_comp_inds.append(_comp_inds)\n",
    "            _comp_inds = [_i]\n",
    "        else:\n",
    "            _comp_inds.append(_i)            \n",
    "        prev_v = _v\n",
    "    if _comp_inds != []:\n",
    "        all_comp_inds.append(_comp_inds)\n",
    "    # calculate length of each compartment\n",
    "    all_comp_lens = np.array([len(_c) for _c in all_comp_inds])\n",
    "    # update number of small comparment\n",
    "    num_small_compartment = np.sum(all_comp_lens < 4)\n",
    "    print(all_comp_lens, num_small_compartment)\n",
    "    # choose the smallest compartment to flip its AB\n",
    "    flip_ind = np.argmin(all_comp_lens)\n",
    "    temp_AB_vector[np.array(all_comp_inds[flip_ind])] = 1 - temp_AB_vector[np.array(all_comp_inds[flip_ind])]\n",
    "    \n",
    "# based on this cleaned AB_vector, recreate AB_dict\n",
    "data_rep1['AB_dict'] = {\n",
    "    'A': np.where(temp_AB_vector==1)[0],\n",
    "    'B': np.where(temp_AB_vector==0)[0],\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots related to compartment calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fig2A: correlation map with AB calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lims = [0,len(contact_corr_map_rep1)]\n",
    "\n",
    "xlims = np.array([min(lims), max(lims)])\n",
    "ylims = np.array([min(lims), max(lims)])\n",
    "\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "domain_line_color = [1,1,0,1]\n",
    "domain_line_width = 1.5\n",
    "bad_color=[0,0,0,1]\n",
    "\n",
    "fig, ax1 = plt.subplots(figsize=(_single_col_width, _single_col_width), dpi=600)\n",
    "\n",
    "# create a color map\n",
    "current_cmap = matplotlib.cm.get_cmap('seismic')\n",
    "current_cmap.set_bad(color=[0.5,0.5,0.5,1])\n",
    "\n",
    "_pf = ax1.imshow(contact_corr_map_rep1, cmap=current_cmap, vmin=-1, vmax=1)\n",
    "\n",
    "ax1.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=0,\n",
    "                pad=1, labelleft=False, labelbottom=False) # remove bottom ticklabels for ax1\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax1.spines.items()]\n",
    "\n",
    "\n",
    "# locate ax1\n",
    "divider = make_axes_locatable(ax1)\n",
    "\n",
    "# colorbar ax\n",
    "cax = divider.append_axes('right', size='6%', pad=\"4%\")\n",
    "cbar = plt.colorbar(_pf,cax=cax, ax=ax1, ticks=[-1,1])\n",
    "cbar.ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "                pad=1, labelleft=False) # remove bottom ticklabels for ax1\n",
    "cbar.outline.set_linewidth(_ticklabel_width)\n",
    "cbar.set_label('Pearson correlation', \n",
    "               fontsize=_font_size, labelpad=0, rotation=270)\n",
    "\n",
    "# create bottom ax\n",
    "bot_ax = divider.append_axes('bottom', size='10%', pad=\"0%\", \n",
    "                             sharex=ax1, xticks=[])\n",
    "bot_ax.bar(data_rep1['AB_dict']['A'], height=1, color='r', width=1, label='A')\n",
    "bot_ax.bar(data_rep1['AB_dict']['B'], height=-1, color='b', width=1, label='B')\n",
    "bot_ax.set_yticks([])\n",
    "bot_ax.set_yticklabels([])\n",
    "bot_ax.set_ylim([-1,1])\n",
    "\n",
    "_xticks = [0, len(contact_corr_map_rep1)-1]\n",
    "bot_ax.set_xticks(_xticks)\n",
    "bot_ax.set_xticklabels(np.round(mid_positions_Mb[_xticks],1))\n",
    "# tick params\n",
    "bot_ax.tick_params('both', labelsize=_font_size-0.5, \n",
    "                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "                pad=1, labelleft=False, labelbottom=True) # remove bottom ticklabels for ax1\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in bot_ax.spines.items()]\n",
    "# set labels\n",
    "bot_ax.set_xlabel(f'Genomic Positions (Mb)', fontsize=_font_size, labelpad=1)\n",
    "\n",
    "# create left ax\n",
    "#left_ax = divider.append_axes('left', size='10%', pad=\"0%\", sharey=ax1, xticks=[])\n",
    "\n",
    "#left_ax.barh(data_rep1['AB_dict']['A'], height=1, color='r', width=1, label='A')\n",
    "#left_ax.barh(data_rep1['AB_dict']['B'], height=1, left=-1, color='b', width=1, label='B')\n",
    "\n",
    "#left_ax.set_xticks([-0.5, 0.5])\n",
    "#left_ax.set_xticklabels(['B', 'A'])\n",
    "\n",
    "#_yticks = _xticks\n",
    "#left_ax.set_yticks(_yticks)\n",
    "#left_ax.set_yticklabels(mid_positions_Mb[_yticks])\n",
    "\n",
    "#left_ax.tick_params('both', labelsize=_font_size-1, \n",
    "#                width=_ticklabel_width, length=_ticklabel_size-1,\n",
    "#                pad=1, labelleft=True) # remove bottom ticklabels for ax1\n",
    "#[i[1].set_linewidth(_ticklabel_width) for i in left_ax.spines.items()]\n",
    "#\n",
    "#left_ax.set_ylabel(f'Genomic Positions (Mb)', fontsize=_font_size, labelpad=1)\n",
    "\n",
    "# set limits\n",
    "bot_ax.set_xlim(xlims-0.5)\n",
    "#left_ax.set_ylim([max(ylims)-0.5, min(ylims)-0.5])\n",
    "\n",
    "ax1.set_title(f\"Chr21 (~3,500 cells)\", fontsize=_font_size+0.5)\n",
    "\n",
    "# save\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.16, right=0.88)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'Fig2A_chr21_contact_corr_map_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FigS2A: plot PC1 for chr21 calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## pc1 barplot\n",
    "fig, ax = plt.subplots(figsize=(_double_col_width, _single_col_width), dpi=600)\n",
    "grid = plt.GridSpec(2, 1, height_ratios=[1,1], hspace=0., wspace=0.)\n",
    "contact_ax = plt.subplot(grid[0])\n",
    "contact_ax.bar(np.where(contact_pc1_rep1>=0)[0],\n",
    "               contact_pc1_rep1[contact_pc1_rep1>=0],\n",
    "               width=1, color='r', label='A')\n",
    "contact_ax.bar(np.where(contact_pc1_rep1<0)[0],\n",
    "               contact_pc1_rep1[contact_pc1_rep1<0],\n",
    "               width=1, color='b', label='B')\n",
    "contact_ax.tick_params('both', labelsize=_font_size, \n",
    "            width=_ticklabel_width, length=_ticklabel_size,\n",
    "            pad=1,labelbottom=False) # remove bottom ticklabels for ax1\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in contact_ax.spines.items()]\n",
    "contact_ax.set_ylim([-15,25])\n",
    "contact_ax.set_yticks([-10,0,10,20])\n",
    "contact_ax.set_ylabel(\"Contact PC1\", fontsize=_font_size, labelpad=0)\n",
    "\n",
    "\n",
    "# hic-ax\n",
    "hic_ax = plt.subplot(grid[1], sharex=contact_ax)\n",
    "\n",
    "hic_ax.bar(np.where(hic_pc1>=0)[0],\n",
    "               hic_pc1[hic_pc1>=0],\n",
    "               width=1, color='r', label='A')\n",
    "hic_ax.bar(np.where(hic_pc1<0)[0],\n",
    "               hic_pc1[hic_pc1<0],\n",
    "               width=1, color='b', label='B')\n",
    "hic_ax.tick_params('both', labelsize=_font_size, \n",
    "            width=_ticklabel_width, length=_ticklabel_size,\n",
    "            pad=1,) # remove bottom ticklabels for ax1\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in hic_ax.spines.items()]\n",
    "hic_ax.set_ylim([-15,25])\n",
    "hic_ax.set_yticks([-10,0,10,20])\n",
    "hic_ax.set_ylabel(\"Hi-C PC1\", fontsize=_font_size, labelpad=0)\n",
    "\n",
    "# set x\n",
    "hic_ax.set_xlim([0, len(contact_pc1_rep1)])\n",
    "_xticks = [0, len(contact_pc1_rep1)-1]\n",
    "hic_ax.set_xticks(_xticks)\n",
    "hic_ax.set_xticklabels(mid_positions_Mb[_xticks])\n",
    "hic_ax.set_xlabel(f'Genomic Positions (Mb)', fontsize=_font_size, labelpad=1)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.1)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f'FigS2A_chr21_PC1_barplot_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1.3 density scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import multiprocessing as mp\n",
    "num_threads=32\n",
    "density_var = 108 # nm\n",
    "\n",
    "# density score for 50kb genomic regions\n",
    "_dna_density_args = [(_zxys,_zxys, data_rep1['AB_dict'], [density_var,density_var,density_var], True) \n",
    "                     for _zxys in data_rep1['dna_zxys']]\n",
    "_dna_density_time = time.time()\n",
    "\n",
    "print(f\"Multiprocessing calculate dna_density_scores\", end=' ')\n",
    "if 'dna_density_scores' not in data_rep1:\n",
    "    with mp.Pool(num_threads) as dna_density_pool:\n",
    "        dna_density_dicts = dna_density_pool.starmap(ia.compartment_tools.scoring.spot_density_scores, _dna_density_args)\n",
    "        dna_density_pool.close()\n",
    "        dna_density_pool.join()\n",
    "        dna_density_pool.terminate()\n",
    "    # save\n",
    "    data_rep1['dna_density_scores'] = dna_density_dicts\n",
    "print(f\"in {time.time()-_dna_density_time:.3f}s.\")\n",
    "\n",
    "# density score for tss\n",
    "_gene_density_args = [(_gzxys,_zxys, data_rep1['AB_dict'], [density_var,density_var,density_var], True) \n",
    "                     for _gzxys, _zxys in zip(data_rep1['tss_zxys'], data_rep1['dna_zxys']) ]\n",
    "_gene_density_time = time.time()\n",
    "\n",
    "print(f\"Multiprocessing calculate gene_density_scores\", end=' ')\n",
    "if 'gene_density_scores' not in data_rep1:\n",
    "    with mp.Pool(num_threads) as gene_density_pool:\n",
    "        gene_density_dicts = gene_density_pool.starmap(ia.compartment_tools.scoring.spot_density_scores, _gene_density_args)\n",
    "        gene_density_pool.close()\n",
    "        gene_density_pool.join()\n",
    "        gene_density_pool.terminate()\n",
    "    # save\n",
    "    data_rep1['gene_density_scores'] = gene_density_dicts\n",
    "print(f\"in {time.time()-_gene_density_time:.3f}s.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FigS2C: mean density for chr21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate mean A, B density\n",
    "mean_A_scores = np.nanmedian([_s['A'] for _s in data_rep1['dna_density_scores']], axis=0)\n",
    "mean_B_scores = np.nanmedian([_s['B'] for _s in data_rep1['dna_density_scores']], axis=0)\n",
    "\n",
    "# Plot\n",
    "fig = plt.figure(figsize=(_double_col_width, _single_col_width),dpi=600)\n",
    "grid = plt.GridSpec(2, 1, height_ratios=[7,1], hspace=0., wspace=0.2)\n",
    "\n",
    "main_ax = plt.subplot(grid[0], xticklabels=[])\n",
    "\n",
    "main_ax.plot(mean_A_scores, 'r.--', label='A density', markersize=2, linewidth=1)\n",
    "main_ax.plot(mean_B_scores, 'b.--', label='B density', markersize=2, linewidth=1)\n",
    "\n",
    "# ticks\n",
    "main_ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size,\n",
    "                pad=1, labelbottom=False) # remove bottom ticklabels for ax1\n",
    "main_ax.tick_params('x', length=0)\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in main_ax.spines.items()]\n",
    "main_ax.set_ylabel(f\"Mean density scores\", fontsize=_font_size, labelpad=1)\n",
    "\n",
    "handles, labels = main_ax.get_legend_handles_labels()\n",
    "main_ax.legend(handles[::-1], labels[::-1], fontsize=_font_size, loc='upper right')\n",
    "main_ax.set_xlim(0,len(mean_A_scores))\n",
    "\n",
    "comp_ax = plt.subplot(grid[1], xticklabels=[], sharex=main_ax)\n",
    "comp_ax.eventplot([data_rep1['AB_dict']['A'], data_rep1['AB_dict']['B']], lineoffsets=[0.5,-0.5],linelengths=1, linewidths=1, \n",
    "                   colors=np.array([[1, 0, 0],[0, 0, 1]]))\n",
    "#comp_ax.imshow(comp_vector[np.newaxis,:], cmap='seismic', vmin=-1, vmax=1)\n",
    "comp_ax.tick_params('both', labelsize=_font_size, \n",
    "                    width=_ticklabel_width, length=_ticklabel_size,\n",
    "                    pad=1, labelbottom=True) # remove bottom ticklabels for ax1\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in comp_ax.spines.items()]\n",
    "\n",
    "comp_ax.set_yticks([0.5,-0.5])\n",
    "comp_ax.set_yticklabels([\"A\",\"B\"])\n",
    "comp_ax.set_ylim([-1,1])\n",
    "_xticks = [0, len(mean_A_scores)-1]\n",
    "comp_ax.set_xticks(_xticks)\n",
    "comp_ax.set_xticklabels(mid_positions_Mb[_xticks])\n",
    "comp_ax.set_xlabel(f'Genomic Positions (Mb)', fontsize=_font_size, labelpad=1)\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.1, left=0.05)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, 'FigS2C_chr21_mean_AB_density_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0.1.4 segregation scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def randomize_AB_dict(AB_dict):\n",
    "    all_regs = np.sort(np.concatenate(list(AB_dict.values())))\n",
    "    AB_identities = np.zeros(len(all_regs))\n",
    "    AB_identities[np.array([_i for _i,_r in enumerate(all_regs) \n",
    "                            if _r in AB_dict['A']])] = 1\n",
    "    # randomize new start\n",
    "    new_start = np.random.randint(0, len(all_regs))\n",
    "    new_AB_identities = np.concatenate([AB_identities[new_start:], AB_identities[:new_start]])\n",
    "    # recreate AB_dict\n",
    "    new_AB_dict = {'A': np.sort(all_regs[np.where(new_AB_identities==1)[0]]),\n",
    "                   'B': np.sort(all_regs[np.where(new_AB_identities==0)[0]]),}\n",
    "    return new_AB_dict\n",
    "\n",
    "# calculate dynamic fraction scores\n",
    "from scipy.stats import scoreatpercentile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AB_identities_rep1 = np.ones(len(data_rep1['dna_zxys'][0])) * np.nan\n",
    "AB_identities_rep1[data_rep1['AB_dict']['A']] = 1\n",
    "AB_identities_rep1[data_rep1['AB_dict']['B']] = 0\n",
    "\n",
    "from tqdm import tqdm\n",
    "# calculate re-thresholded fraction scores\n",
    "A_fracs, B_fracs = [], []\n",
    "A_ths, B_ths = [], []\n",
    "cloud_th_per=67\n",
    "for _sd in tqdm(data_rep1['dna_density_scores']):\n",
    "    # define A,B threshold based on their own densities\n",
    "    _A_th = scoreatpercentile(_sd['A'][data_rep1['AB_dict']['A']], 100-cloud_th_per)\n",
    "    _B_th = scoreatpercentile(_sd['B'][data_rep1['AB_dict']['B']], 100-cloud_th_per)\n",
    "    # calculate purity within A,B clouds\n",
    "    A_fracs.append(np.nanmean(AB_identities_rep1[np.where(_sd['A'] >= _A_th)[0]]))\n",
    "    B_fracs.append(1-np.nanmean(AB_identities_rep1[np.where(_sd['B'] >= _B_th)[0]]))\n",
    "    # store AB thresholds for references\n",
    "    A_ths.append(_A_th)\n",
    "    B_ths.append(_B_th)\n",
    "# calculate re-thresholded fraction scores\n",
    "rand_A_fracs, rand_B_fracs = [], []\n",
    "\n",
    "for _sd in tqdm(data_rep1['dna_density_scores']):\n",
    "    # randomize AB dict\n",
    "    _rand_AB_dict = randomize_AB_dict(data_rep1['AB_dict'])\n",
    "    _rand_A_inds, _rand_B_inds = np.array(_rand_AB_dict['A']), np.array(_rand_AB_dict['B'])\n",
    "    # generate randomized AB_identities_rep1 vector for purity calculation\n",
    "    _rand_AB_identities_rep1 = np.ones(len(data_rep1['dna_zxys'][0])) * np.nan\n",
    "    _rand_AB_identities_rep1[_rand_AB_dict['A']] = 1\n",
    "    _rand_AB_identities_rep1[_rand_AB_dict['B']] = 0\n",
    "    # define A,B threshold based on their own densities\n",
    "    _A_th = scoreatpercentile(_sd['A'][_rand_A_inds], 100-cloud_th_per)\n",
    "    _B_th = scoreatpercentile(_sd['B'][_rand_B_inds], 100-cloud_th_per)\n",
    "    # calculate purity within A,B clouds\n",
    "    rand_A_fracs.append(np.nanmean(_rand_AB_identities_rep1[np.where(_sd['A'] >= _A_th)[0]]))\n",
    "    rand_B_fracs.append(1-np.nanmean(_rand_AB_identities_rep1[np.where(_sd['B'] >= _B_th)[0]]))\n",
    "\n",
    "# Save\n",
    "data_rep1['segregation_scores'] = (np.array(A_fracs) + np.array(B_fracs)) / 2\n",
    "data_rep1['randomized_segregation_scores'] = (np.array(rand_A_fracs) + np.array(rand_B_fracs)) / 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fig 2C: segregation score histogram for Chr21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width),dpi=600)\n",
    "ax.hist(data_rep1['segregation_scores'], 100, range=(0.,1), \n",
    "        density=True, alpha=0.5, \n",
    "        color=[1,0.5,0], label='Chr21')\n",
    "ax.hist(data_rep1['randomized_segregation_scores'], 100, range=(0.,1), \n",
    "        density=True, alpha=0.5, \n",
    "        color=[0.3,0.4,0.4], label='randomized control')\n",
    "\n",
    "ax.legend(fontsize=_font_size-1, loc='upper right')\n",
    "\n",
    "ax.set_xlabel(\"Segregation score\", fontsize=_font_size, labelpad=1)\n",
    "ax.set_ylabel(\"Probability density\", fontsize=_font_size, labelpad=1)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size,\n",
    "                pad=1, labelleft=True) # remove bottom ticklabels for a_ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.set_xlim([0.4,1])\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.15, left=0.15)\n",
    "plt.savefig(os.path.join(figure_folder, 'Fig2C_chr21_segregation_hist_rep1.pdf'), transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2H: A/B density ratio difference w/wo transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_density_dicts = data_rep1['gene_density_scores']\n",
    "gene_A_scores = np.array([_gsd['A'] for _gsd in gene_density_dicts])\n",
    "gene_B_scores = np.array([_gsd['B'] for _gsd in gene_density_dicts])\n",
    "\n",
    "trans_pfs = np.array(data_rep1['trans_pfs'])\n",
    "\n",
    "on_gene_AB_ratio, off_gene_AB_ratio = [], []\n",
    "on_gene_total_density = []\n",
    "off_gene_total_density = []\n",
    "\n",
    "for _gind in range(gene_A_scores.shape[1]):\n",
    "    # extract AB and transcription\n",
    "    _a_scores = gene_A_scores[:, _gind]\n",
    "    _b_scores = gene_B_scores[:, _gind]\n",
    "    _transcriptions = trans_pfs[:,_gind]\n",
    "    # get AB ratio\n",
    "    _log_ab_ratios = np.log(_a_scores)/np.log(2) - np.log(_b_scores)/np.log(2)\n",
    "    _total_density =  _a_scores+_b_scores\n",
    "    # append\n",
    "    on_gene_AB_ratio.append(np.nanmedian(_log_ab_ratios[_transcriptions & (np.isinf(_log_ab_ratios)==False)]))\n",
    "    off_gene_AB_ratio.append(np.nanmedian(_log_ab_ratios[~_transcriptions & (np.isinf(_log_ab_ratios)==False)]))\n",
    "    on_gene_total_density.append(np.nanmedian(_total_density[_transcriptions & (np.isinf(_total_density)==False)]))\n",
    "    off_gene_total_density.append(np.nanmedian(_total_density[~_transcriptions & (np.isinf(_total_density)==False)]))\n",
    "\n",
    "\n",
    "# convert into arrays\n",
    "on_gene_AB_ratio = np.array(on_gene_AB_ratio)\n",
    "off_gene_AB_ratio = np.array(off_gene_AB_ratio)\n",
    "on_gene_total_density = np.array(on_gene_total_density)\n",
    "off_gene_total_density = np.array(off_gene_total_density)\n",
    "\n",
    "gene_ratio_diff = on_gene_AB_ratio - off_gene_AB_ratio\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width),dpi=600)\n",
    "\n",
    "ax.plot(np.sort(gene_ratio_diff), '.', color=[1,0.5,0], label='All genes', markersize=3)\n",
    "ax.plot(np.zeros(len(gene_ratio_diff)), 'black', label='ref', linewidth=0.75)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, width=_ticklabel_width, length=_ticklabel_size, pad=1) \n",
    "        \n",
    "        # remove bottom ticklabels for ax\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "ax.set_xticks(np.arange(len(gene_ratio_diff)))\n",
    "ax.set_yticks(np.arange(-6,6,2))\n",
    "ax.set_xticklabels([])\n",
    "ax.set_xlabel('Genes', fontsize=_font_size, labelpad=2)\n",
    "ax.set_ylabel('Log2 difference of \\nA/B density ratio', fontsize=_font_size, labelpad=2)\n",
    "ax.set_xlim([-1, len(gene_ratio_diff)])\n",
    "ax.set_ylim([-2,4])\n",
    "\n",
    "#ax.legend(fontsize=_font_size, framealpha=1, loc='upper left')\n",
    "ax.text(len(gene_ratio_diff), max(ax.get_ylim())/30, f'{np.round(np.mean(gene_ratio_diff>0),2)*100:.0f}%', \n",
    "        fontsize=_font_size-1, verticalalignment='bottom', horizontalalignment='right',)\n",
    "ax.text(len(gene_ratio_diff), -max(ax.get_ylim())/30, f'{np.round(np.mean(gene_ratio_diff<0),2)*100:.0f}%', \n",
    "        fontsize=_font_size-1, verticalalignment='top', horizontalalignment='right',)\n",
    "\n",
    "ax.set_title(\"Transcribing/Silent\", fontsize=_font_size)\n",
    "plt.gcf().subplots_adjust(left=0.2, bottom=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, 'Fig2H_diff_abratio_w-wo_transcribe_rep1.pdf'), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2I: firing rate difference with high/low A/B density ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_density_dicts = data_rep1['gene_density_scores']\n",
    "gene_A_scores = np.array([_gsd['A'] for _gsd in gene_density_dicts])\n",
    "gene_B_scores = np.array([_gsd['B'] for _gsd in gene_density_dicts])\n",
    "\n",
    "trans_pfs = np.array(data_rep1['trans_pfs'])\n",
    "\n",
    "abratio_ratio = []\n",
    "for _gid in np.arange(trans_pfs.shape[1]):\n",
    "    # extract info\n",
    "    _ga = gene_A_scores[:,_gid]\n",
    "    _gb = gene_B_scores[:,_gid]\n",
    "    _gratio = np.log(_ga / _gb) \n",
    "    \n",
    "    # sel_inds\n",
    "    _inds = np.argsort(_gratio)\n",
    "    _sel_v = np.where((np.isnan(_gratio)==False) * (np.isinf(_gratio)==False))[0]\n",
    "    _sel_inds = np.array([_i for _i in _inds if _i in _sel_v], dtype=np.int)\n",
    "    \n",
    "    _li = _sel_inds[:int(len(_sel_inds)/4)]\n",
    "    _hi = _sel_inds[-int(len(_sel_inds)/4):]\n",
    "    #print(len(_sel_inds), np.mean(_gratio[_li]), np.mean(_gratio[_hi]))\n",
    "    _lfr = np.nanmean(trans_pfs[_li, _gid])\n",
    "    _hfr = np.nanmean(trans_pfs[_hi, _gid])\n",
    "    abratio_ratio.append(_hfr / _lfr)\n",
    "\n",
    "abratio_fr_ratio = np.log(np.array(abratio_ratio)) / np.log(2)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(_single_col_width, _single_col_width),dpi=600)\n",
    "\n",
    "#ax.plot(np.arange(len(sel_A_gene_inds)), \n",
    "#        np.sort(np.log(abratio_ratio[sel_A_gene_inds])), \n",
    "#        'r.', label='A genes', markersize=3)\n",
    "#ax.plot(np.arange(len(sel_A_gene_inds),len(sel_A_gene_inds)+len(sel_B_gene_inds)), \n",
    "#         np.sort(np.log(abratio_ratio[sel_B_gene_inds])), \n",
    "#        'b.', label='B genes', markersize=3)\n",
    "\n",
    "ax.plot(np.sort(abratio_fr_ratio), \n",
    "        '.', color=[1,0.5,0], label='All genes', markersize=3)\n",
    "ax.plot(np.zeros(len(abratio_fr_ratio)), \n",
    "        'black', label='reference', linewidth=0.75)\n",
    "\n",
    "ax.tick_params('both', labelsize=_font_size, \n",
    "                width=_ticklabel_width, length=_ticklabel_size,\n",
    "                pad=1) # remove bottom ticklabels for ax1\n",
    "ax.set_xticks(np.arange(len(abratio_fr_ratio)))\n",
    "ax.set_xticklabels([])\n",
    "ax.set_ylabel('Log odds ratio', fontsize=_font_size, labelpad=1)\n",
    "ax.set_xlim([-1, len(abratio_fr_ratio)])\n",
    "ax.set_ylim([-1,2])\n",
    "\n",
    "[i[1].set_linewidth(_ticklabel_width) for i in ax.spines.items()]\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "ax.set_xlabel('Genes', fontsize=_font_size, labelpad=2)\n",
    "ax.set_ylabel('Log2 difference of firing rate', fontsize=_font_size, labelpad=1)\n",
    "ax.set_title(f\"high/low log(A/B) density ratio\", \n",
    "             fontsize=_font_size)\n",
    "\n",
    "#ax.legend(fontsize=_font_size, framealpha=1, loc='upper left')\n",
    "\n",
    "ax.text(len(abratio_fr_ratio), max(ax.get_ylim())/30, f'{np.round(np.mean(abratio_fr_ratio>0),2)*100:.0f}%', \n",
    "        fontsize=_font_size-1, verticalalignment='bottom', horizontalalignment='right',)\n",
    "ax.text(len(abratio_fr_ratio), -max(ax.get_ylim())/30, f'{np.round(np.mean(abratio_fr_ratio<=0),2)*100:.0f}%', \n",
    "        fontsize=_font_size-1, verticalalignment='top', horizontalalignment='right',)\n",
    "\n",
    "plt.gcf().subplots_adjust(left=0.2, bottom=0.15)\n",
    "\n",
    "plt.savefig(os.path.join(figure_folder, f\"Fig2I_firing-rate_diff_w_high-low_abratio_rep1.pdf\"), transparent=True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
